{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "import pandas as pd\n",
    "from utilities.preprocessing import parse_gaf,parse_kgml,get_go_term_description,kegg_symbols_and_names\n",
    "from neo4j import GraphDatabase\n",
    "import os\n",
    "import dask.dataframe as dd\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "\n",
    "\n",
    "kgml_path = 'data/KGML/hsa05012.xml'\n",
    "gaf_path='data/GAF/goa_human.gaf'\n",
    "\n",
    "\n",
    "def parse_kgml(file_path):\n",
    "    tree = ET.parse(file_path)\n",
    "    root = tree.getroot()\n",
    "    pathway_id = root.get('number')\n",
    "    pathway_name = root.get('title')\n",
    "    \n",
    "    entries = {\n",
    "        'gene': [],\n",
    "        'compound': []\n",
    "    }\n",
    "    interactions = []\n",
    "\n",
    "    for entry in root.findall('entry'):\n",
    "        entry_type = entry.get('type')\n",
    "        \n",
    "        if entry_type in ['gene', 'compound']:\n",
    "            entry_id = entry.get('id')\n",
    "            entry_names = entry.get('name').split()\n",
    "            \n",
    "            graphics = entry.find('graphics')\n",
    "            entry_symbols = graphics.get('name').split(', ') if graphics is not None else []\n",
    "            # Remove '...' from symbols if present\n",
    "            entry_symbols = [symbol.rstrip('...') for symbol in entry_symbols]\n",
    "\n",
    "            entries[entry_type].append({\n",
    "                f'{entry_type}_id': entry_id,\n",
    "                f'{entry_type}_names': entry_names,\n",
    "                f'{entry_type}_symbols': entry_symbols\n",
    "            })\n",
    "\n",
    "    for relation in root.findall('relation'):\n",
    "        entry1 = relation.get('entry1')\n",
    "        entry2 = relation.get('entry2')\n",
    "        relation_type = relation.get('type')\n",
    "        subtypes = [subtype.get('name') for subtype in relation.findall('subtype')]\n",
    "\n",
    "        interactions.append({\n",
    "            'entry1': entry1,\n",
    "            'entry2': entry2,\n",
    "            'type': relation_type,\n",
    "            'subtypes': subtypes\n",
    "        })\n",
    "\n",
    "    pathway_data = {\n",
    "        'pathway_id': pathway_id,\n",
    "        'pathway_name': pathway_name,\n",
    "        'genes': entries['gene'],\n",
    "        'compounds': entries['compound'],\n",
    "        'interactions': interactions\n",
    "    }\n",
    "\n",
    "    return pathway_data\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pathway_id': '05012',\n",
       " 'pathway_name': 'Parkinson disease',\n",
       " 'genes': [{'gene_id': '5',\n",
       "   'gene_names': ['hsa:120892'],\n",
       "   'gene_symbols': ['LRRK2', 'AURA17', 'DARDARIN', 'PARK8', 'RIPK7', 'ROCO2']},\n",
       "  {'gene_id': '11',\n",
       "   'gene_names': ['hsa:7317', 'hsa:7318'],\n",
       "   'gene_symbols': ['UBA1',\n",
       "    'A1S9',\n",
       "    'A1S9T',\n",
       "    'A1ST',\n",
       "    'AMCX1',\n",
       "    'CFAP124',\n",
       "    'GXP1',\n",
       "    'POC20',\n",
       "    'SMAX2',\n",
       "    'UBA1A',\n",
       "    'UBE1',\n",
       "    'UBE1X',\n",
       "    'VEXAS']},\n",
       "  {'gene_id': '12',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '13',\n",
       "   'gene_names': ['hsa:7332', 'hsa:9246'],\n",
       "   'gene_symbols': ['UBE2L3', 'E2-F1', 'L-UBC', 'UBCH7', 'UbcM4']},\n",
       "  {'gene_id': '14',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '15',\n",
       "   'gene_names': ['hsa:5071'],\n",
       "   'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       "  {'gene_id': '16',\n",
       "   'gene_names': ['hsa:5413'],\n",
       "   'gene_symbols': ['SEPTIN5',\n",
       "    'CDCREL',\n",
       "    'CDCREL-1',\n",
       "    'CDCREL1',\n",
       "    'H5',\n",
       "    'HCDCREL-1',\n",
       "    'PNUTL1',\n",
       "    'SEPT5',\n",
       "    'Septin-5']},\n",
       "  {'gene_id': '17',\n",
       "   'gene_names': ['hsa:9627'],\n",
       "   'gene_symbols': ['SNCAIP', 'SYPH1', 'Sph1']},\n",
       "  {'gene_id': '18',\n",
       "   'gene_names': ['hsa:2861'],\n",
       "   'gene_symbols': ['GPR37', 'EDNRBL', 'PAELR', 'hET(B)R-LP']},\n",
       "  {'gene_id': '19',\n",
       "   'gene_names': ['hsa:11315'],\n",
       "   'gene_symbols': ['PARK7', 'DJ-1', 'DJ1', 'GATD2', 'HEL-S-67p']},\n",
       "  {'gene_id': '20',\n",
       "   'gene_names': ['hsa:65018'],\n",
       "   'gene_symbols': ['PINK1', 'BRPK', 'PARK6']},\n",
       "  {'gene_id': '21',\n",
       "   'gene_names': ['hsa:6622'],\n",
       "   'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       "  {'gene_id': '22',\n",
       "   'gene_names': ['hsa:7345'],\n",
       "   'gene_symbols': ['UCHL1',\n",
       "    'HEL-117',\n",
       "    'HEL-S-53',\n",
       "    'NDGOA',\n",
       "    'PARK5',\n",
       "    'PGP_9.5',\n",
       "    'PGP9.5',\n",
       "    'PGP95',\n",
       "    'SPG79',\n",
       "    'SPG79A',\n",
       "    'UCHL-1',\n",
       "    'Uch-L1']},\n",
       "  {'gene_id': '25',\n",
       "   'gene_names': ['hsa:54205'],\n",
       "   'gene_symbols': ['CYCS', 'CYC', 'HCS', 'THC4']},\n",
       "  {'gene_id': '26',\n",
       "   'gene_names': ['hsa:100532726',\n",
       "    'hsa:107984365',\n",
       "    'hsa:126328',\n",
       "    'hsa:374291',\n",
       "    'hsa:4535',\n",
       "    'hsa:4536',\n",
       "    'hsa:4537',\n",
       "    'hsa:4538',\n",
       "    'hsa:4539',\n",
       "    'hsa:4540',\n",
       "    'hsa:4541',\n",
       "    'hsa:4694',\n",
       "    'hsa:4695',\n",
       "    'hsa:4696',\n",
       "    'hsa:4697',\n",
       "    'hsa:4698',\n",
       "    'hsa:4700',\n",
       "    'hsa:4701',\n",
       "    'hsa:4702',\n",
       "    'hsa:4704',\n",
       "    'hsa:4705',\n",
       "    'hsa:4706',\n",
       "    'hsa:4707',\n",
       "    'hsa:4708',\n",
       "    'hsa:4709',\n",
       "    'hsa:4710',\n",
       "    'hsa:4711',\n",
       "    'hsa:4712',\n",
       "    'hsa:4713',\n",
       "    'hsa:4714',\n",
       "    'hsa:4715',\n",
       "    'hsa:4716',\n",
       "    'hsa:4717',\n",
       "    'hsa:4718',\n",
       "    'hsa:4719',\n",
       "    'hsa:4720',\n",
       "    'hsa:4722',\n",
       "    'hsa:4723',\n",
       "    'hsa:4724',\n",
       "    'hsa:4725',\n",
       "    'hsa:4726',\n",
       "    'hsa:4728',\n",
       "    'hsa:4729',\n",
       "    'hsa:4731',\n",
       "    'hsa:51079',\n",
       "    'hsa:54539',\n",
       "    'hsa:55967',\n",
       "    'hsa:56901'],\n",
       "   'gene_symbols': ['NDUFC2-KCTD14']},\n",
       "  {'gene_id': '27',\n",
       "   'gene_names': ['hsa:6389', 'hsa:6390', 'hsa:6391', 'hsa:6392'],\n",
       "   'gene_symbols': ['SDHA',\n",
       "    'CMD1GG',\n",
       "    'FP',\n",
       "    'MC2DN1',\n",
       "    'NDAXOA',\n",
       "    'PGL5',\n",
       "    'PPGL5',\n",
       "    'SDH1',\n",
       "    'SDH2',\n",
       "    'SDHF']},\n",
       "  {'gene_id': '28',\n",
       "   'gene_names': ['hsa:10975',\n",
       "    'hsa:1537',\n",
       "    'hsa:27089',\n",
       "    'hsa:29796',\n",
       "    'hsa:440567',\n",
       "    'hsa:4519',\n",
       "    'hsa:7381',\n",
       "    'hsa:7384',\n",
       "    'hsa:7385',\n",
       "    'hsa:7386',\n",
       "    'hsa:7388'],\n",
       "   'gene_symbols': ['UQCR11', '0710008D09Rik', 'QCR10', 'UQCR']},\n",
       "  {'gene_id': '29',\n",
       "   'gene_names': ['hsa:125965',\n",
       "    'hsa:1327',\n",
       "    'hsa:1329',\n",
       "    'hsa:1337',\n",
       "    'hsa:1339',\n",
       "    'hsa:1340',\n",
       "    'hsa:1345',\n",
       "    'hsa:1346',\n",
       "    'hsa:1347',\n",
       "    'hsa:1349',\n",
       "    'hsa:1350',\n",
       "    'hsa:1351',\n",
       "    'hsa:170712',\n",
       "    'hsa:341947',\n",
       "    'hsa:4512',\n",
       "    'hsa:4513',\n",
       "    'hsa:4514',\n",
       "    'hsa:84701',\n",
       "    'hsa:9167',\n",
       "    'hsa:9377'],\n",
       "   'gene_symbols': ['COX6B2', 'COXVIB2', 'CT59']},\n",
       "  {'gene_id': '30',\n",
       "   'gene_names': ['hsa:10476',\n",
       "    'hsa:4508',\n",
       "    'hsa:4509',\n",
       "    'hsa:498',\n",
       "    'hsa:506',\n",
       "    'hsa:509',\n",
       "    'hsa:513',\n",
       "    'hsa:514',\n",
       "    'hsa:515',\n",
       "    'hsa:516',\n",
       "    'hsa:517',\n",
       "    'hsa:518',\n",
       "    'hsa:522',\n",
       "    'hsa:539'],\n",
       "   'gene_symbols': ['ATP5PD', 'APT5H', 'ATP5H', 'ATPQ']},\n",
       "  {'gene_id': '31',\n",
       "   'gene_names': ['hsa:27429'],\n",
       "   'gene_symbols': ['HTRA2', 'MGCA8', 'OMI', 'PARK13', 'PRSS25']},\n",
       "  {'gene_id': '32',\n",
       "   'gene_names': ['hsa:10105',\n",
       "    'hsa:291',\n",
       "    'hsa:292',\n",
       "    'hsa:293',\n",
       "    'hsa:7416',\n",
       "    'hsa:7417',\n",
       "    'hsa:7419',\n",
       "    'hsa:83447'],\n",
       "   'gene_symbols': ['PPIF', 'CYP3', 'CyP-M', 'Cyp-D', 'CypD']},\n",
       "  {'gene_id': '45',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '46',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '47',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '48',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '49',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '50',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '59',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '60',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '61',\n",
       "   'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "   'gene_symbols': ['RPS27A',\n",
       "    'CEP80',\n",
       "    'HEL112',\n",
       "    'S27A',\n",
       "    'UBA80',\n",
       "    'UBC',\n",
       "    'UBCEP1',\n",
       "    'UBCEP80',\n",
       "    'eS31']},\n",
       "  {'gene_id': '139',\n",
       "   'gene_names': ['hsa:118424', 'hsa:51465', 'hsa:7326', 'hsa:7327'],\n",
       "   'gene_symbols': ['UBE2J2', 'NCUBE-2', 'NCUBE2', 'PRO2121', 'UBC6']},\n",
       "  {'gene_id': '149',\n",
       "   'gene_names': ['hsa:7054'],\n",
       "   'gene_symbols': ['TH', 'DYT14', 'DYT5b', 'TYH']},\n",
       "  {'gene_id': '150',\n",
       "   'gene_names': ['hsa:6531'],\n",
       "   'gene_symbols': ['SLC6A3', 'DAT', 'DAT1', 'PKDYS', 'PKDYS1']},\n",
       "  {'gene_id': '152',\n",
       "   'gene_names': ['hsa:842'],\n",
       "   'gene_symbols': ['CASP9',\n",
       "    'APAF-3',\n",
       "    'APAF3',\n",
       "    'ICE-LAP6',\n",
       "    'MCH6',\n",
       "    'PPP1R56']},\n",
       "  {'gene_id': '153',\n",
       "   'gene_names': ['hsa:836'],\n",
       "   'gene_symbols': ['CASP3', 'CPP32', 'CPP32B', 'SCA-1']},\n",
       "  {'gene_id': '155',\n",
       "   'gene_names': ['hsa:317'],\n",
       "   'gene_symbols': ['APAF1', 'APAF-1', 'CED4']},\n",
       "  {'gene_id': '159',\n",
       "   'gene_names': ['hsa:6570', 'hsa:6571'],\n",
       "   'gene_symbols': ['SLC18A1', 'CGAT', 'VAT1', 'VMAT1']},\n",
       "  {'gene_id': '160',\n",
       "   'gene_names': ['hsa:6531'],\n",
       "   'gene_symbols': ['SLC6A3', 'DAT', 'DAT1', 'PKDYS', 'PKDYS1']},\n",
       "  {'gene_id': '168',\n",
       "   'gene_names': ['hsa:135'],\n",
       "   'gene_symbols': ['ADORA2A', 'A2aR', 'ADORA2', 'RDC8']},\n",
       "  {'gene_id': '169',\n",
       "   'gene_names': ['hsa:1813'],\n",
       "   'gene_symbols': ['DRD2', 'D2DR', 'D2R']},\n",
       "  {'gene_id': '170',\n",
       "   'gene_names': ['hsa:1812'],\n",
       "   'gene_symbols': ['DRD1', 'D1R', 'DADR', 'DRD1A']},\n",
       "  {'gene_id': '171',\n",
       "   'gene_names': ['hsa:2774', 'hsa:2778'],\n",
       "   'gene_symbols': ['GNAL', 'DYT25', 'HG1O']},\n",
       "  {'gene_id': '172',\n",
       "   'gene_names': ['hsa:2770', 'hsa:2771', 'hsa:2773'],\n",
       "   'gene_symbols': ['GNAI1', 'Gi', 'HG1B', 'NEDHISB']},\n",
       "  {'gene_id': '173',\n",
       "   'gene_names': ['hsa:2774', 'hsa:2778'],\n",
       "   'gene_symbols': ['GNAL', 'DYT25', 'HG1O']},\n",
       "  {'gene_id': '174',\n",
       "   'gene_names': ['hsa:111'],\n",
       "   'gene_symbols': ['ADCY5', 'AC5', 'DSKOD', 'FDFM']},\n",
       "  {'gene_id': '175',\n",
       "   'gene_names': ['hsa:111'],\n",
       "   'gene_symbols': ['ADCY5', 'AC5', 'DSKOD', 'FDFM']},\n",
       "  {'gene_id': '176',\n",
       "   'gene_names': ['hsa:5566', 'hsa:5567', 'hsa:5568'],\n",
       "   'gene_symbols': ['PRKACA', 'CAFD1', 'PKACA', 'PPNAD4']},\n",
       "  {'gene_id': '177',\n",
       "   'gene_names': ['hsa:5566', 'hsa:5567', 'hsa:5568'],\n",
       "   'gene_symbols': ['PRKACA', 'CAFD1', 'PKACA', 'PPNAD4']},\n",
       "  {'gene_id': '229',\n",
       "   'gene_names': ['hsa:6622'],\n",
       "   'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       "  {'gene_id': '232',\n",
       "   'gene_names': ['hsa:9451'],\n",
       "   'gene_symbols': ['EIF2AK3', 'PEK', 'PERK', 'WRS']},\n",
       "  {'gene_id': '233',\n",
       "   'gene_names': ['hsa:2081'],\n",
       "   'gene_symbols': ['ERN1', 'IRE1', 'IRE1P', 'IRE1a', 'hIRE1p']},\n",
       "  {'gene_id': '234',\n",
       "   'gene_names': ['hsa:22926'],\n",
       "   'gene_symbols': ['ATF6', 'ACHM7', 'ATF6A', 'ATP6alpha']},\n",
       "  {'gene_id': '236',\n",
       "   'gene_names': ['hsa:1965'],\n",
       "   'gene_symbols': ['EIF2S1',\n",
       "    'EIF-2',\n",
       "    'EIF-2A',\n",
       "    'EIF-2alpha',\n",
       "    'EIF2',\n",
       "    'EIF2A']},\n",
       "  {'gene_id': '237',\n",
       "   'gene_names': ['hsa:468'],\n",
       "   'gene_symbols': ['ATF4', 'CREB-2', 'CREB2', 'TAXREB67', 'TXREB']},\n",
       "  {'gene_id': '241',\n",
       "   'gene_names': ['hsa:7494'],\n",
       "   'gene_symbols': ['XBP1', 'TREB-5', 'TREB5', 'XBP-1', 'XBP2']},\n",
       "  {'gene_id': '243',\n",
       "   'gene_names': ['hsa:1649'],\n",
       "   'gene_symbols': ['DDIT3',\n",
       "    'AltDDIT3',\n",
       "    'C/EBPzeta',\n",
       "    'CEBPZ',\n",
       "    'CHOP',\n",
       "    'CHOP-10',\n",
       "    'CHOP10',\n",
       "    'GADD153']},\n",
       "  {'gene_id': '244',\n",
       "   'gene_names': ['hsa:1649'],\n",
       "   'gene_symbols': ['DDIT3',\n",
       "    'AltDDIT3',\n",
       "    'C/EBPzeta',\n",
       "    'CEBPZ',\n",
       "    'CHOP',\n",
       "    'CHOP-10',\n",
       "    'CHOP10',\n",
       "    'GADD153']},\n",
       "  {'gene_id': '248',\n",
       "   'gene_names': ['hsa:3309'],\n",
       "   'gene_symbols': ['HSPA5', 'BIP', 'GRP78', 'HEL-S-89n']},\n",
       "  {'gene_id': '255',\n",
       "   'gene_names': ['hsa:6622'],\n",
       "   'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       "  {'gene_id': '260',\n",
       "   'gene_names': ['hsa:143471',\n",
       "    'hsa:5682',\n",
       "    'hsa:5683',\n",
       "    'hsa:5684',\n",
       "    'hsa:5685',\n",
       "    'hsa:5686',\n",
       "    'hsa:5687',\n",
       "    'hsa:5688',\n",
       "    'hsa:5689',\n",
       "    'hsa:5690',\n",
       "    'hsa:5691',\n",
       "    'hsa:5692',\n",
       "    'hsa:5693',\n",
       "    'hsa:5694',\n",
       "    'hsa:5695'],\n",
       "   'gene_symbols': ['PSMA8', 'PSMA7L']},\n",
       "  {'gene_id': '261',\n",
       "   'gene_names': ['hsa:10213',\n",
       "    'hsa:11047',\n",
       "    'hsa:5700',\n",
       "    'hsa:5701',\n",
       "    'hsa:5702',\n",
       "    'hsa:5704',\n",
       "    'hsa:5705',\n",
       "    'hsa:5706',\n",
       "    'hsa:5707',\n",
       "    'hsa:5708',\n",
       "    'hsa:5709',\n",
       "    'hsa:5710',\n",
       "    'hsa:5713',\n",
       "    'hsa:5714',\n",
       "    'hsa:5715',\n",
       "    'hsa:5717',\n",
       "    'hsa:5718',\n",
       "    'hsa:5719',\n",
       "    'hsa:7979',\n",
       "    'hsa:9861'],\n",
       "   'gene_symbols': ['PSMD14', 'PAD1', 'POH1', 'RPN11']},\n",
       "  {'gene_id': '267',\n",
       "   'gene_names': ['hsa:3708', 'hsa:3709', 'hsa:3710'],\n",
       "   'gene_symbols': ['ITPR1',\n",
       "    'ACV',\n",
       "    'CLA4',\n",
       "    'INSP3R1',\n",
       "    'IP3R',\n",
       "    'IP3R1',\n",
       "    'PPP1R94',\n",
       "    'SCA15',\n",
       "    'SCA16',\n",
       "    'SCA29']},\n",
       "  {'gene_id': '271',\n",
       "   'gene_names': ['hsa:5335'],\n",
       "   'gene_symbols': ['PLCG1',\n",
       "    'IDAA',\n",
       "    'NCKAP3',\n",
       "    'PLC-II',\n",
       "    'PLC1',\n",
       "    'PLC148',\n",
       "    'PLCgamma1']},\n",
       "  {'gene_id': '272',\n",
       "   'gene_names': ['hsa:6622'],\n",
       "   'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       "  {'gene_id': '273',\n",
       "   'gene_names': ['hsa:5071'],\n",
       "   'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       "  {'gene_id': '278',\n",
       "   'gene_names': ['hsa:163688',\n",
       "    'hsa:51806',\n",
       "    'hsa:801',\n",
       "    'hsa:805',\n",
       "    'hsa:808',\n",
       "    'hsa:810',\n",
       "    'hsa:91860'],\n",
       "   'gene_symbols': ['CALML6', 'CAGLP']},\n",
       "  {'gene_id': '279',\n",
       "   'gene_names': ['hsa:815', 'hsa:816', 'hsa:817', 'hsa:818'],\n",
       "   'gene_symbols': ['CAMK2A',\n",
       "    'CAMKA',\n",
       "    'CaMKIINalpha',\n",
       "    'CaMKIIalpha',\n",
       "    'MRD53',\n",
       "    'MRT63']},\n",
       "  {'gene_id': '280',\n",
       "   'gene_names': ['hsa:4137'],\n",
       "   'gene_symbols': ['MAPT',\n",
       "    'DDPAC',\n",
       "    'FTDP-17',\n",
       "    'MAPTL',\n",
       "    'MSTD',\n",
       "    'MTBT1',\n",
       "    'MTBT2',\n",
       "    'PPND',\n",
       "    'PPP1R103',\n",
       "    'TAU',\n",
       "    'Tau-PHF6',\n",
       "    'tau-40']},\n",
       "  {'gene_id': '284',\n",
       "   'gene_names': ['hsa:6263'],\n",
       "   'gene_symbols': ['RYR3', 'CMYP20', 'RYR-3']},\n",
       "  {'gene_id': '292',\n",
       "   'gene_names': ['hsa:147700',\n",
       "    'hsa:3798',\n",
       "    'hsa:3799',\n",
       "    'hsa:3800',\n",
       "    'hsa:3831',\n",
       "    'hsa:64837',\n",
       "    'hsa:89953'],\n",
       "   'gene_symbols': ['KLC3', 'KLC2', 'KLC2L', 'KLCt', 'KNS2B']},\n",
       "  {'gene_id': '295',\n",
       "   'gene_names': ['hsa:10376',\n",
       "    'hsa:112714',\n",
       "    'hsa:113457',\n",
       "    'hsa:51807',\n",
       "    'hsa:7277',\n",
       "    'hsa:7278',\n",
       "    'hsa:7846',\n",
       "    'hsa:79861',\n",
       "    'hsa:84790'],\n",
       "   'gene_symbols': ['TUBA1B', 'K-ALPHA-1']},\n",
       "  {'gene_id': '296',\n",
       "   'gene_names': ['hsa:10381',\n",
       "    'hsa:10382',\n",
       "    'hsa:10383',\n",
       "    'hsa:124908013',\n",
       "    'hsa:124908015',\n",
       "    'hsa:124908041',\n",
       "    'hsa:203068',\n",
       "    'hsa:260334',\n",
       "    'hsa:347688',\n",
       "    'hsa:347733',\n",
       "    'hsa:7280',\n",
       "    'hsa:81027',\n",
       "    'hsa:84617'],\n",
       "   'gene_symbols': ['TUBB3',\n",
       "    'CDCBM',\n",
       "    'CDCBM1',\n",
       "    'CFEOM3',\n",
       "    'CFEOM3A',\n",
       "    'FEOM3',\n",
       "    'TUBB4',\n",
       "    'beta-4']},\n",
       "  {'gene_id': '300',\n",
       "   'gene_names': ['hsa:6622'],\n",
       "   'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       "  {'gene_id': '304',\n",
       "   'gene_names': ['hsa:65018'],\n",
       "   'gene_symbols': ['PINK1', 'BRPK', 'PARK6']},\n",
       "  {'gene_id': '307',\n",
       "   'gene_names': ['hsa:5071'],\n",
       "   'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       "  {'gene_id': '314',\n",
       "   'gene_names': ['hsa:65018'],\n",
       "   'gene_symbols': ['PINK1', 'BRPK', 'PARK6']},\n",
       "  {'gene_id': '315',\n",
       "   'gene_names': ['hsa:5071'],\n",
       "   'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       "  {'gene_id': '316',\n",
       "   'gene_names': ['hsa:581'],\n",
       "   'gene_symbols': ['BAX', 'BCL2L4']},\n",
       "  {'gene_id': '323',\n",
       "   'gene_names': ['hsa:6622'],\n",
       "   'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       "  {'gene_id': '326',\n",
       "   'gene_names': ['hsa:11315'],\n",
       "   'gene_symbols': ['PARK7', 'DJ-1', 'DJ1', 'GATD2', 'HEL-S-67p']},\n",
       "  {'gene_id': '327',\n",
       "   'gene_names': ['hsa:598'],\n",
       "   'gene_symbols': ['BCL2L1',\n",
       "    'BCL-XL/S',\n",
       "    'BCL2L',\n",
       "    'BCLX',\n",
       "    'Bcl-X',\n",
       "    'PPP1R52']},\n",
       "  {'gene_id': '332',\n",
       "   'gene_names': ['hsa:10131'],\n",
       "   'gene_symbols': ['TRAP1', 'HSP_75', 'HSP75', 'HSP90L', 'TRAP-1']},\n",
       "  {'gene_id': '335',\n",
       "   'gene_names': ['hsa:65018'],\n",
       "   'gene_symbols': ['PINK1', 'BRPK', 'PARK6']},\n",
       "  {'gene_id': '338',\n",
       "   'gene_names': ['hsa:7157'],\n",
       "   'gene_symbols': ['TP53', 'BCC7', 'BMFS5', 'LFS1', 'P53', 'TRP53']},\n",
       "  {'gene_id': '343',\n",
       "   'gene_names': ['hsa:1616'],\n",
       "   'gene_symbols': ['DAXX', 'BING2', 'DAP6', 'EAP1', 'SMIM40']},\n",
       "  {'gene_id': '344',\n",
       "   'gene_names': ['hsa:4217'],\n",
       "   'gene_symbols': ['MAP3K5', 'ASK1', 'MAPKKK5', 'MEKK5']},\n",
       "  {'gene_id': '345',\n",
       "   'gene_names': ['hsa:5599', 'hsa:5601', 'hsa:5602'],\n",
       "   'gene_symbols': ['MAPK8',\n",
       "    'JNK',\n",
       "    'JNK-46',\n",
       "    'JNK1',\n",
       "    'JNK1A2',\n",
       "    'JNK21B1/2',\n",
       "    'PRKM8',\n",
       "    'SAPK1',\n",
       "    'SAPK1c']},\n",
       "  {'gene_id': '347',\n",
       "   'gene_names': ['hsa:4780'],\n",
       "   'gene_symbols': ['NFE2L2', 'HEBP1', 'IMDDHH', 'NRF2', 'Nrf-2']},\n",
       "  {'gene_id': '361',\n",
       "   'gene_names': ['hsa:1843'],\n",
       "   'gene_symbols': ['DUSP1', 'CL100', 'HVH1', 'MKP-1', 'MKP1', 'PTPN10']},\n",
       "  {'gene_id': '362',\n",
       "   'gene_names': ['hsa:25828', 'hsa:7295'],\n",
       "   'gene_symbols': ['TXN2', 'COXPD29', 'MT-TRX', 'MTRX', 'TRX2', 'TXN']},\n",
       "  {'gene_id': '364',\n",
       "   'gene_names': ['hsa:6570', 'hsa:6571'],\n",
       "   'gene_symbols': ['SLC18A1', 'CGAT', 'VAT1', 'VMAT1']},\n",
       "  {'gene_id': '365',\n",
       "   'gene_names': ['hsa:6622'],\n",
       "   'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       "  {'gene_id': '368',\n",
       "   'gene_names': ['hsa:5071'],\n",
       "   'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       "  {'gene_id': '374',\n",
       "   'gene_names': ['hsa:4128', 'hsa:4129'],\n",
       "   'gene_symbols': ['MAOA', 'BRNRS', 'MAO-A']},\n",
       "  {'gene_id': '375',\n",
       "   'gene_names': ['hsa:5071'],\n",
       "   'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       "  {'gene_id': '382',\n",
       "   'gene_names': ['hsa:90550'],\n",
       "   'gene_symbols': ['MCU', 'C10orf42', 'CCDC109A', 'HsMCU']},\n",
       "  {'gene_id': '398',\n",
       "   'gene_names': ['hsa:55669', 'hsa:9927'],\n",
       "   'gene_symbols': ['MFN1', 'hfzo1', 'hfzo2']},\n",
       "  {'gene_id': '402',\n",
       "   'gene_names': ['hsa:6622'],\n",
       "   'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       "  {'gene_id': '711',\n",
       "   'gene_names': ['hsa:6531'],\n",
       "   'gene_symbols': ['SLC6A3', 'DAT', 'DAT1', 'PKDYS', 'PKDYS1']},\n",
       "  {'gene_id': '714',\n",
       "   'gene_names': ['hsa:9817'],\n",
       "   'gene_symbols': ['KEAP1', 'INrf2', 'KLHL19']},\n",
       "  {'gene_id': '715',\n",
       "   'gene_names': ['hsa:25828', 'hsa:7295'],\n",
       "   'gene_symbols': ['TXN2', 'COXPD29', 'MT-TRX', 'MTRX', 'TRX2', 'TXN']},\n",
       "  {'gene_id': '724',\n",
       "   'gene_names': ['hsa:201266',\n",
       "    'hsa:221074',\n",
       "    'hsa:23516',\n",
       "    'hsa:25800',\n",
       "    'hsa:27173',\n",
       "    'hsa:283375',\n",
       "    'hsa:29985',\n",
       "    'hsa:29986',\n",
       "    'hsa:4891',\n",
       "    'hsa:55334',\n",
       "    'hsa:55630',\n",
       "    'hsa:57181',\n",
       "    'hsa:64116',\n",
       "    'hsa:7922',\n",
       "    'hsa:91252'],\n",
       "   'gene_symbols': ['SLC39A11', 'C17orf26', 'ZIP-11', 'ZIP11']},\n",
       "  {'gene_id': '735',\n",
       "   'gene_names': ['hsa:6647'],\n",
       "   'gene_symbols': ['SOD1',\n",
       "    'ALS',\n",
       "    'ALS1',\n",
       "    'HEL-S-44',\n",
       "    'IPOA',\n",
       "    'SOD',\n",
       "    'STAHP',\n",
       "    'hSod1',\n",
       "    'homodimer']}],\n",
       " 'compounds': [{'compound_id': '23',\n",
       "   'compound_names': ['cpd:C04599'],\n",
       "   'compound_symbols': ['C04599']},\n",
       "  {'compound_id': '24',\n",
       "   'compound_names': ['cpd:C07593'],\n",
       "   'compound_symbols': ['C07593']},\n",
       "  {'compound_id': '80',\n",
       "   'compound_names': ['cpd:C00002'],\n",
       "   'compound_symbols': ['C00002']},\n",
       "  {'compound_id': '81',\n",
       "   'compound_names': ['cpd:C00008'],\n",
       "   'compound_symbols': ['C00008']},\n",
       "  {'compound_id': '82',\n",
       "   'compound_names': ['cpd:C00009'],\n",
       "   'compound_symbols': ['C00009']},\n",
       "  {'compound_id': '89',\n",
       "   'compound_names': ['cpd:C03758'],\n",
       "   'compound_symbols': ['C03758']},\n",
       "  {'compound_id': '98',\n",
       "   'compound_names': ['cpd:C00704', 'cpd:C00027'],\n",
       "   'compound_symbols': ['C00704']},\n",
       "  {'compound_id': '100',\n",
       "   'compound_names': ['cpd:C00002'],\n",
       "   'compound_symbols': ['C00002']},\n",
       "  {'compound_id': '101',\n",
       "   'compound_names': ['cpd:C00020'],\n",
       "   'compound_symbols': ['C00020']},\n",
       "  {'compound_id': '102',\n",
       "   'compound_names': ['cpd:C00013'],\n",
       "   'compound_symbols': ['C00013']},\n",
       "  {'compound_id': '104',\n",
       "   'compound_names': ['cpd:C11310'],\n",
       "   'compound_symbols': ['C11310']},\n",
       "  {'compound_id': '124',\n",
       "   'compound_names': ['cpd:C00002'],\n",
       "   'compound_symbols': ['C00002']},\n",
       "  {'compound_id': '125',\n",
       "   'compound_names': ['cpd:C00008'],\n",
       "   'compound_symbols': ['C00008']},\n",
       "  {'compound_id': '126',\n",
       "   'compound_names': ['cpd:C00009'],\n",
       "   'compound_symbols': ['C00009']},\n",
       "  {'compound_id': '128',\n",
       "   'compound_names': ['cpd:C00355'],\n",
       "   'compound_symbols': ['C00355']},\n",
       "  {'compound_id': '147',\n",
       "   'compound_names': ['cpd:C00082'],\n",
       "   'compound_symbols': ['C00082']},\n",
       "  {'compound_id': '189',\n",
       "   'compound_names': ['cpd:C03758'],\n",
       "   'compound_symbols': ['C03758']},\n",
       "  {'compound_id': '202',\n",
       "   'compound_names': ['cpd:C00575'],\n",
       "   'compound_symbols': ['C00575']},\n",
       "  {'compound_id': '203',\n",
       "   'compound_names': ['cpd:C00575'],\n",
       "   'compound_symbols': ['C00575']},\n",
       "  {'compound_id': '219',\n",
       "   'compound_names': ['cpd:C00212'],\n",
       "   'compound_symbols': ['C00212']},\n",
       "  {'compound_id': '264',\n",
       "   'compound_names': ['cpd:C01245'],\n",
       "   'compound_symbols': ['C01245']},\n",
       "  {'compound_id': '265',\n",
       "   'compound_names': ['cpd:C00076'],\n",
       "   'compound_symbols': ['C00076']},\n",
       "  {'compound_id': '266',\n",
       "   'compound_names': ['cpd:C00165'],\n",
       "   'compound_symbols': ['C00165']},\n",
       "  {'compound_id': '276',\n",
       "   'compound_names': ['cpd:C00076'],\n",
       "   'compound_symbols': ['C00076']},\n",
       "  {'compound_id': '303',\n",
       "   'compound_names': ['cpd:C15231'],\n",
       "   'compound_symbols': ['C15231']},\n",
       "  {'compound_id': '372',\n",
       "   'compound_names': ['cpd:C17755'],\n",
       "   'compound_symbols': ['C17755']},\n",
       "  {'compound_id': '383',\n",
       "   'compound_names': ['cpd:C00076'],\n",
       "   'compound_symbols': ['C00076']},\n",
       "  {'compound_id': '404',\n",
       "   'compound_names': ['cpd:C04043'],\n",
       "   'compound_symbols': ['C04043']},\n",
       "  {'compound_id': '706',\n",
       "   'compound_names': ['cpd:C14701'],\n",
       "   'compound_symbols': ['C14701']},\n",
       "  {'compound_id': '707',\n",
       "   'compound_names': ['cpd:C14701'],\n",
       "   'compound_symbols': ['C14701']},\n",
       "  {'compound_id': '708',\n",
       "   'compound_names': ['cpd:C14701'],\n",
       "   'compound_symbols': ['C14701']},\n",
       "  {'compound_id': '712',\n",
       "   'compound_names': ['cpd:C00704'],\n",
       "   'compound_symbols': ['C00704']},\n",
       "  {'compound_id': '713',\n",
       "   'compound_names': ['cpd:C00007'],\n",
       "   'compound_symbols': ['C00007']},\n",
       "  {'compound_id': '727',\n",
       "   'compound_names': ['cpd:C00027'],\n",
       "   'compound_symbols': ['C00027']},\n",
       "  {'compound_id': '728',\n",
       "   'compound_names': ['cpd:C16844'],\n",
       "   'compound_symbols': ['C16844']},\n",
       "  {'compound_id': '732',\n",
       "   'compound_names': ['cpd:C22381'],\n",
       "   'compound_symbols': ['C22381']},\n",
       "  {'compound_id': '742',\n",
       "   'compound_names': ['cpd:C00034'],\n",
       "   'compound_symbols': ['C00034']}],\n",
       " 'interactions': [{'entry1': '219',\n",
       "   'entry2': '168',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '168',\n",
       "   'entry2': '171',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '171',\n",
       "   'entry2': '174',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '174',\n",
       "   'entry2': '202',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '202',\n",
       "   'entry2': '176',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '276',\n",
       "   'entry2': '278',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '278',\n",
       "   'entry2': '279',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '233',\n",
       "   'entry2': '241',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '232',\n",
       "   'entry2': '236',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '236',\n",
       "   'entry2': '237',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '25',\n",
       "   'entry2': '152',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '155',\n",
       "   'entry2': '152',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '152',\n",
       "   'entry2': '153',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '279',\n",
       "   'entry2': '280',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation', 'phosphorylation']},\n",
       "  {'entry1': '21',\n",
       "   'entry2': '743',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '255',\n",
       "   'entry2': '248',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '248',\n",
       "   'entry2': '234',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '248',\n",
       "   'entry2': '233',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '248',\n",
       "   'entry2': '232',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '365',\n",
       "   'entry2': '364',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '323',\n",
       "   'entry2': '26',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '104',\n",
       "   'entry2': '26',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '24',\n",
       "   'entry2': '26',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '304',\n",
       "   'entry2': '26',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '303',\n",
       "   'entry2': '28',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '300',\n",
       "   'entry2': '292',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '31',\n",
       "   'entry2': '25',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '332',\n",
       "   'entry2': '25',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '744',\n",
       "   'entry2': '345',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '189',\n",
       "   'entry2': '169',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '169',\n",
       "   'entry2': '172',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '172',\n",
       "   'entry2': '174',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '189',\n",
       "   'entry2': '170',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '170',\n",
       "   'entry2': '173',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '173',\n",
       "   'entry2': '175',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '175',\n",
       "   'entry2': '203',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '203',\n",
       "   'entry2': '177',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '15',\n",
       "   'entry2': '13',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '15',\n",
       "   'entry2': '743',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '229',\n",
       "   'entry2': '149',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '375',\n",
       "   'entry2': '374',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '335',\n",
       "   'entry2': '31',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '315',\n",
       "   'entry2': '316',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '314',\n",
       "   'entry2': '315',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '326',\n",
       "   'entry2': '327',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '327',\n",
       "   'entry2': '316',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '20',\n",
       "   'entry2': '307',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '307',\n",
       "   'entry2': '398',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '19',\n",
       "   'entry2': '744',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '343',\n",
       "   'entry2': '19',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '343',\n",
       "   'entry2': '338',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '19',\n",
       "   'entry2': '347',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '234',\n",
       "   'entry2': '243',\n",
       "   'type': 'GErel',\n",
       "   'subtypes': ['expression']},\n",
       "  {'entry1': '237',\n",
       "   'entry2': '244',\n",
       "   'type': 'GErel',\n",
       "   'subtypes': ['expression']},\n",
       "  {'entry1': '338',\n",
       "   'entry2': '361',\n",
       "   'type': 'GErel',\n",
       "   'subtypes': ['expression']},\n",
       "  {'entry1': '347',\n",
       "   'entry2': '362',\n",
       "   'type': 'GErel',\n",
       "   'subtypes': ['expression']},\n",
       "  {'entry1': '15',\n",
       "   'entry2': '402',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '15',\n",
       "   'entry2': '18',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '15',\n",
       "   'entry2': '16',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '15',\n",
       "   'entry2': '17',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '11',\n",
       "   'entry2': '12',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['binding/association']},\n",
       "  {'entry1': '13',\n",
       "   'entry2': '14',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['binding/association']},\n",
       "  {'entry1': '59',\n",
       "   'entry2': '60',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['binding/association']},\n",
       "  {'entry1': '60',\n",
       "   'entry2': '61',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['binding/association']},\n",
       "  {'entry1': '45',\n",
       "   'entry2': '46',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['binding/association']},\n",
       "  {'entry1': '46',\n",
       "   'entry2': '47',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['binding/association']},\n",
       "  {'entry1': '273',\n",
       "   'entry2': '271',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '271',\n",
       "   'entry2': '264',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '264',\n",
       "   'entry2': '267',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '271',\n",
       "   'entry2': '266',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '267',\n",
       "   'entry2': '265',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '383',\n",
       "   'entry2': '32',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '32',\n",
       "   'entry2': '25',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '316',\n",
       "   'entry2': '25',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['activation']},\n",
       "  {'entry1': '295',\n",
       "   'entry2': '296',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['binding/association']},\n",
       "  {'entry1': '742',\n",
       "   'entry2': '27',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '732',\n",
       "   'entry2': '714',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '714',\n",
       "   'entry2': '347',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '732',\n",
       "   'entry2': '715',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '715',\n",
       "   'entry2': '744',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['inhibition']},\n",
       "  {'entry1': '19',\n",
       "   'entry2': '338',\n",
       "   'type': 'PPrel',\n",
       "   'subtypes': ['missing interaction']},\n",
       "  {'entry1': '706',\n",
       "   'entry2': '732',\n",
       "   'type': 'PCrel',\n",
       "   'subtypes': ['activation', 'indirect effect']}]}"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_data=parse_kgml(kgml_path)\n",
    "pathway_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'gene_id': '5',\n",
       "  'gene_names': ['hsa:120892'],\n",
       "  'gene_symbols': ['LRRK2', 'AURA17', 'DARDARIN', 'PARK8', 'RIPK7', 'ROCO2']},\n",
       " {'gene_id': '11',\n",
       "  'gene_names': ['hsa:7317', 'hsa:7318'],\n",
       "  'gene_symbols': ['UBA1',\n",
       "   'A1S9',\n",
       "   'A1S9T',\n",
       "   'A1ST',\n",
       "   'AMCX1',\n",
       "   'CFAP124',\n",
       "   'GXP1',\n",
       "   'POC20',\n",
       "   'SMAX2',\n",
       "   'UBA1A',\n",
       "   'UBE1',\n",
       "   'UBE1X',\n",
       "   'VEXAS']},\n",
       " {'gene_id': '12',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '13',\n",
       "  'gene_names': ['hsa:7332', 'hsa:9246'],\n",
       "  'gene_symbols': ['UBE2L3', 'E2-F1', 'L-UBC', 'UBCH7', 'UbcM4']},\n",
       " {'gene_id': '14',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '15',\n",
       "  'gene_names': ['hsa:5071'],\n",
       "  'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       " {'gene_id': '16',\n",
       "  'gene_names': ['hsa:5413'],\n",
       "  'gene_symbols': ['SEPTIN5',\n",
       "   'CDCREL',\n",
       "   'CDCREL-1',\n",
       "   'CDCREL1',\n",
       "   'H5',\n",
       "   'HCDCREL-1',\n",
       "   'PNUTL1',\n",
       "   'SEPT5',\n",
       "   'Septin-5']},\n",
       " {'gene_id': '17',\n",
       "  'gene_names': ['hsa:9627'],\n",
       "  'gene_symbols': ['SNCAIP', 'SYPH1', 'Sph1']},\n",
       " {'gene_id': '18',\n",
       "  'gene_names': ['hsa:2861'],\n",
       "  'gene_symbols': ['GPR37', 'EDNRBL', 'PAELR', 'hET(B)R-LP']},\n",
       " {'gene_id': '19',\n",
       "  'gene_names': ['hsa:11315'],\n",
       "  'gene_symbols': ['PARK7', 'DJ-1', 'DJ1', 'GATD2', 'HEL-S-67p']},\n",
       " {'gene_id': '20',\n",
       "  'gene_names': ['hsa:65018'],\n",
       "  'gene_symbols': ['PINK1', 'BRPK', 'PARK6']},\n",
       " {'gene_id': '21',\n",
       "  'gene_names': ['hsa:6622'],\n",
       "  'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       " {'gene_id': '22',\n",
       "  'gene_names': ['hsa:7345'],\n",
       "  'gene_symbols': ['UCHL1',\n",
       "   'HEL-117',\n",
       "   'HEL-S-53',\n",
       "   'NDGOA',\n",
       "   'PARK5',\n",
       "   'PGP_9.5',\n",
       "   'PGP9.5',\n",
       "   'PGP95',\n",
       "   'SPG79',\n",
       "   'SPG79A',\n",
       "   'UCHL-1',\n",
       "   'Uch-L1']},\n",
       " {'gene_id': '25',\n",
       "  'gene_names': ['hsa:54205'],\n",
       "  'gene_symbols': ['CYCS', 'CYC', 'HCS', 'THC4']},\n",
       " {'gene_id': '26',\n",
       "  'gene_names': ['hsa:100532726',\n",
       "   'hsa:107984365',\n",
       "   'hsa:126328',\n",
       "   'hsa:374291',\n",
       "   'hsa:4535',\n",
       "   'hsa:4536',\n",
       "   'hsa:4537',\n",
       "   'hsa:4538',\n",
       "   'hsa:4539',\n",
       "   'hsa:4540',\n",
       "   'hsa:4541',\n",
       "   'hsa:4694',\n",
       "   'hsa:4695',\n",
       "   'hsa:4696',\n",
       "   'hsa:4697',\n",
       "   'hsa:4698',\n",
       "   'hsa:4700',\n",
       "   'hsa:4701',\n",
       "   'hsa:4702',\n",
       "   'hsa:4704',\n",
       "   'hsa:4705',\n",
       "   'hsa:4706',\n",
       "   'hsa:4707',\n",
       "   'hsa:4708',\n",
       "   'hsa:4709',\n",
       "   'hsa:4710',\n",
       "   'hsa:4711',\n",
       "   'hsa:4712',\n",
       "   'hsa:4713',\n",
       "   'hsa:4714',\n",
       "   'hsa:4715',\n",
       "   'hsa:4716',\n",
       "   'hsa:4717',\n",
       "   'hsa:4718',\n",
       "   'hsa:4719',\n",
       "   'hsa:4720',\n",
       "   'hsa:4722',\n",
       "   'hsa:4723',\n",
       "   'hsa:4724',\n",
       "   'hsa:4725',\n",
       "   'hsa:4726',\n",
       "   'hsa:4728',\n",
       "   'hsa:4729',\n",
       "   'hsa:4731',\n",
       "   'hsa:51079',\n",
       "   'hsa:54539',\n",
       "   'hsa:55967',\n",
       "   'hsa:56901'],\n",
       "  'gene_symbols': ['NDUFC2-KCTD14']},\n",
       " {'gene_id': '27',\n",
       "  'gene_names': ['hsa:6389', 'hsa:6390', 'hsa:6391', 'hsa:6392'],\n",
       "  'gene_symbols': ['SDHA',\n",
       "   'CMD1GG',\n",
       "   'FP',\n",
       "   'MC2DN1',\n",
       "   'NDAXOA',\n",
       "   'PGL5',\n",
       "   'PPGL5',\n",
       "   'SDH1',\n",
       "   'SDH2',\n",
       "   'SDHF']},\n",
       " {'gene_id': '28',\n",
       "  'gene_names': ['hsa:10975',\n",
       "   'hsa:1537',\n",
       "   'hsa:27089',\n",
       "   'hsa:29796',\n",
       "   'hsa:440567',\n",
       "   'hsa:4519',\n",
       "   'hsa:7381',\n",
       "   'hsa:7384',\n",
       "   'hsa:7385',\n",
       "   'hsa:7386',\n",
       "   'hsa:7388'],\n",
       "  'gene_symbols': ['UQCR11', '0710008D09Rik', 'QCR10', 'UQCR']},\n",
       " {'gene_id': '29',\n",
       "  'gene_names': ['hsa:125965',\n",
       "   'hsa:1327',\n",
       "   'hsa:1329',\n",
       "   'hsa:1337',\n",
       "   'hsa:1339',\n",
       "   'hsa:1340',\n",
       "   'hsa:1345',\n",
       "   'hsa:1346',\n",
       "   'hsa:1347',\n",
       "   'hsa:1349',\n",
       "   'hsa:1350',\n",
       "   'hsa:1351',\n",
       "   'hsa:170712',\n",
       "   'hsa:341947',\n",
       "   'hsa:4512',\n",
       "   'hsa:4513',\n",
       "   'hsa:4514',\n",
       "   'hsa:84701',\n",
       "   'hsa:9167',\n",
       "   'hsa:9377'],\n",
       "  'gene_symbols': ['COX6B2', 'COXVIB2', 'CT59']},\n",
       " {'gene_id': '30',\n",
       "  'gene_names': ['hsa:10476',\n",
       "   'hsa:4508',\n",
       "   'hsa:4509',\n",
       "   'hsa:498',\n",
       "   'hsa:506',\n",
       "   'hsa:509',\n",
       "   'hsa:513',\n",
       "   'hsa:514',\n",
       "   'hsa:515',\n",
       "   'hsa:516',\n",
       "   'hsa:517',\n",
       "   'hsa:518',\n",
       "   'hsa:522',\n",
       "   'hsa:539'],\n",
       "  'gene_symbols': ['ATP5PD', 'APT5H', 'ATP5H', 'ATPQ']},\n",
       " {'gene_id': '31',\n",
       "  'gene_names': ['hsa:27429'],\n",
       "  'gene_symbols': ['HTRA2', 'MGCA8', 'OMI', 'PARK13', 'PRSS25']},\n",
       " {'gene_id': '32',\n",
       "  'gene_names': ['hsa:10105',\n",
       "   'hsa:291',\n",
       "   'hsa:292',\n",
       "   'hsa:293',\n",
       "   'hsa:7416',\n",
       "   'hsa:7417',\n",
       "   'hsa:7419',\n",
       "   'hsa:83447'],\n",
       "  'gene_symbols': ['PPIF', 'CYP3', 'CyP-M', 'Cyp-D', 'CypD']},\n",
       " {'gene_id': '45',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '46',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '47',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '48',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '49',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '50',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '59',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '60',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '61',\n",
       "  'gene_names': ['hsa:6233', 'hsa:7311', 'hsa:7314', 'hsa:7316'],\n",
       "  'gene_symbols': ['RPS27A',\n",
       "   'CEP80',\n",
       "   'HEL112',\n",
       "   'S27A',\n",
       "   'UBA80',\n",
       "   'UBC',\n",
       "   'UBCEP1',\n",
       "   'UBCEP80',\n",
       "   'eS31']},\n",
       " {'gene_id': '139',\n",
       "  'gene_names': ['hsa:118424', 'hsa:51465', 'hsa:7326', 'hsa:7327'],\n",
       "  'gene_symbols': ['UBE2J2', 'NCUBE-2', 'NCUBE2', 'PRO2121', 'UBC6']},\n",
       " {'gene_id': '149',\n",
       "  'gene_names': ['hsa:7054'],\n",
       "  'gene_symbols': ['TH', 'DYT14', 'DYT5b', 'TYH']},\n",
       " {'gene_id': '150',\n",
       "  'gene_names': ['hsa:6531'],\n",
       "  'gene_symbols': ['SLC6A3', 'DAT', 'DAT1', 'PKDYS', 'PKDYS1']},\n",
       " {'gene_id': '152',\n",
       "  'gene_names': ['hsa:842'],\n",
       "  'gene_symbols': ['CASP9', 'APAF-3', 'APAF3', 'ICE-LAP6', 'MCH6', 'PPP1R56']},\n",
       " {'gene_id': '153',\n",
       "  'gene_names': ['hsa:836'],\n",
       "  'gene_symbols': ['CASP3', 'CPP32', 'CPP32B', 'SCA-1']},\n",
       " {'gene_id': '155',\n",
       "  'gene_names': ['hsa:317'],\n",
       "  'gene_symbols': ['APAF1', 'APAF-1', 'CED4']},\n",
       " {'gene_id': '159',\n",
       "  'gene_names': ['hsa:6570', 'hsa:6571'],\n",
       "  'gene_symbols': ['SLC18A1', 'CGAT', 'VAT1', 'VMAT1']},\n",
       " {'gene_id': '160',\n",
       "  'gene_names': ['hsa:6531'],\n",
       "  'gene_symbols': ['SLC6A3', 'DAT', 'DAT1', 'PKDYS', 'PKDYS1']},\n",
       " {'gene_id': '168',\n",
       "  'gene_names': ['hsa:135'],\n",
       "  'gene_symbols': ['ADORA2A', 'A2aR', 'ADORA2', 'RDC8']},\n",
       " {'gene_id': '169',\n",
       "  'gene_names': ['hsa:1813'],\n",
       "  'gene_symbols': ['DRD2', 'D2DR', 'D2R']},\n",
       " {'gene_id': '170',\n",
       "  'gene_names': ['hsa:1812'],\n",
       "  'gene_symbols': ['DRD1', 'D1R', 'DADR', 'DRD1A']},\n",
       " {'gene_id': '171',\n",
       "  'gene_names': ['hsa:2774', 'hsa:2778'],\n",
       "  'gene_symbols': ['GNAL', 'DYT25', 'HG1O']},\n",
       " {'gene_id': '172',\n",
       "  'gene_names': ['hsa:2770', 'hsa:2771', 'hsa:2773'],\n",
       "  'gene_symbols': ['GNAI1', 'Gi', 'HG1B', 'NEDHISB']},\n",
       " {'gene_id': '173',\n",
       "  'gene_names': ['hsa:2774', 'hsa:2778'],\n",
       "  'gene_symbols': ['GNAL', 'DYT25', 'HG1O']},\n",
       " {'gene_id': '174',\n",
       "  'gene_names': ['hsa:111'],\n",
       "  'gene_symbols': ['ADCY5', 'AC5', 'DSKOD', 'FDFM']},\n",
       " {'gene_id': '175',\n",
       "  'gene_names': ['hsa:111'],\n",
       "  'gene_symbols': ['ADCY5', 'AC5', 'DSKOD', 'FDFM']},\n",
       " {'gene_id': '176',\n",
       "  'gene_names': ['hsa:5566', 'hsa:5567', 'hsa:5568'],\n",
       "  'gene_symbols': ['PRKACA', 'CAFD1', 'PKACA', 'PPNAD4']},\n",
       " {'gene_id': '177',\n",
       "  'gene_names': ['hsa:5566', 'hsa:5567', 'hsa:5568'],\n",
       "  'gene_symbols': ['PRKACA', 'CAFD1', 'PKACA', 'PPNAD4']},\n",
       " {'gene_id': '229',\n",
       "  'gene_names': ['hsa:6622'],\n",
       "  'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       " {'gene_id': '232',\n",
       "  'gene_names': ['hsa:9451'],\n",
       "  'gene_symbols': ['EIF2AK3', 'PEK', 'PERK', 'WRS']},\n",
       " {'gene_id': '233',\n",
       "  'gene_names': ['hsa:2081'],\n",
       "  'gene_symbols': ['ERN1', 'IRE1', 'IRE1P', 'IRE1a', 'hIRE1p']},\n",
       " {'gene_id': '234',\n",
       "  'gene_names': ['hsa:22926'],\n",
       "  'gene_symbols': ['ATF6', 'ACHM7', 'ATF6A', 'ATP6alpha']},\n",
       " {'gene_id': '236',\n",
       "  'gene_names': ['hsa:1965'],\n",
       "  'gene_symbols': ['EIF2S1',\n",
       "   'EIF-2',\n",
       "   'EIF-2A',\n",
       "   'EIF-2alpha',\n",
       "   'EIF2',\n",
       "   'EIF2A']},\n",
       " {'gene_id': '237',\n",
       "  'gene_names': ['hsa:468'],\n",
       "  'gene_symbols': ['ATF4', 'CREB-2', 'CREB2', 'TAXREB67', 'TXREB']},\n",
       " {'gene_id': '241',\n",
       "  'gene_names': ['hsa:7494'],\n",
       "  'gene_symbols': ['XBP1', 'TREB-5', 'TREB5', 'XBP-1', 'XBP2']},\n",
       " {'gene_id': '243',\n",
       "  'gene_names': ['hsa:1649'],\n",
       "  'gene_symbols': ['DDIT3',\n",
       "   'AltDDIT3',\n",
       "   'C/EBPzeta',\n",
       "   'CEBPZ',\n",
       "   'CHOP',\n",
       "   'CHOP-10',\n",
       "   'CHOP10',\n",
       "   'GADD153']},\n",
       " {'gene_id': '244',\n",
       "  'gene_names': ['hsa:1649'],\n",
       "  'gene_symbols': ['DDIT3',\n",
       "   'AltDDIT3',\n",
       "   'C/EBPzeta',\n",
       "   'CEBPZ',\n",
       "   'CHOP',\n",
       "   'CHOP-10',\n",
       "   'CHOP10',\n",
       "   'GADD153']},\n",
       " {'gene_id': '248',\n",
       "  'gene_names': ['hsa:3309'],\n",
       "  'gene_symbols': ['HSPA5', 'BIP', 'GRP78', 'HEL-S-89n']},\n",
       " {'gene_id': '255',\n",
       "  'gene_names': ['hsa:6622'],\n",
       "  'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       " {'gene_id': '260',\n",
       "  'gene_names': ['hsa:143471',\n",
       "   'hsa:5682',\n",
       "   'hsa:5683',\n",
       "   'hsa:5684',\n",
       "   'hsa:5685',\n",
       "   'hsa:5686',\n",
       "   'hsa:5687',\n",
       "   'hsa:5688',\n",
       "   'hsa:5689',\n",
       "   'hsa:5690',\n",
       "   'hsa:5691',\n",
       "   'hsa:5692',\n",
       "   'hsa:5693',\n",
       "   'hsa:5694',\n",
       "   'hsa:5695'],\n",
       "  'gene_symbols': ['PSMA8', 'PSMA7L']},\n",
       " {'gene_id': '261',\n",
       "  'gene_names': ['hsa:10213',\n",
       "   'hsa:11047',\n",
       "   'hsa:5700',\n",
       "   'hsa:5701',\n",
       "   'hsa:5702',\n",
       "   'hsa:5704',\n",
       "   'hsa:5705',\n",
       "   'hsa:5706',\n",
       "   'hsa:5707',\n",
       "   'hsa:5708',\n",
       "   'hsa:5709',\n",
       "   'hsa:5710',\n",
       "   'hsa:5713',\n",
       "   'hsa:5714',\n",
       "   'hsa:5715',\n",
       "   'hsa:5717',\n",
       "   'hsa:5718',\n",
       "   'hsa:5719',\n",
       "   'hsa:7979',\n",
       "   'hsa:9861'],\n",
       "  'gene_symbols': ['PSMD14', 'PAD1', 'POH1', 'RPN11']},\n",
       " {'gene_id': '267',\n",
       "  'gene_names': ['hsa:3708', 'hsa:3709', 'hsa:3710'],\n",
       "  'gene_symbols': ['ITPR1',\n",
       "   'ACV',\n",
       "   'CLA4',\n",
       "   'INSP3R1',\n",
       "   'IP3R',\n",
       "   'IP3R1',\n",
       "   'PPP1R94',\n",
       "   'SCA15',\n",
       "   'SCA16',\n",
       "   'SCA29']},\n",
       " {'gene_id': '271',\n",
       "  'gene_names': ['hsa:5335'],\n",
       "  'gene_symbols': ['PLCG1',\n",
       "   'IDAA',\n",
       "   'NCKAP3',\n",
       "   'PLC-II',\n",
       "   'PLC1',\n",
       "   'PLC148',\n",
       "   'PLCgamma1']},\n",
       " {'gene_id': '272',\n",
       "  'gene_names': ['hsa:6622'],\n",
       "  'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       " {'gene_id': '273',\n",
       "  'gene_names': ['hsa:5071'],\n",
       "  'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       " {'gene_id': '278',\n",
       "  'gene_names': ['hsa:163688',\n",
       "   'hsa:51806',\n",
       "   'hsa:801',\n",
       "   'hsa:805',\n",
       "   'hsa:808',\n",
       "   'hsa:810',\n",
       "   'hsa:91860'],\n",
       "  'gene_symbols': ['CALML6', 'CAGLP']},\n",
       " {'gene_id': '279',\n",
       "  'gene_names': ['hsa:815', 'hsa:816', 'hsa:817', 'hsa:818'],\n",
       "  'gene_symbols': ['CAMK2A',\n",
       "   'CAMKA',\n",
       "   'CaMKIINalpha',\n",
       "   'CaMKIIalpha',\n",
       "   'MRD53',\n",
       "   'MRT63']},\n",
       " {'gene_id': '280',\n",
       "  'gene_names': ['hsa:4137'],\n",
       "  'gene_symbols': ['MAPT',\n",
       "   'DDPAC',\n",
       "   'FTDP-17',\n",
       "   'MAPTL',\n",
       "   'MSTD',\n",
       "   'MTBT1',\n",
       "   'MTBT2',\n",
       "   'PPND',\n",
       "   'PPP1R103',\n",
       "   'TAU',\n",
       "   'Tau-PHF6',\n",
       "   'tau-40']},\n",
       " {'gene_id': '284',\n",
       "  'gene_names': ['hsa:6263'],\n",
       "  'gene_symbols': ['RYR3', 'CMYP20', 'RYR-3']},\n",
       " {'gene_id': '292',\n",
       "  'gene_names': ['hsa:147700',\n",
       "   'hsa:3798',\n",
       "   'hsa:3799',\n",
       "   'hsa:3800',\n",
       "   'hsa:3831',\n",
       "   'hsa:64837',\n",
       "   'hsa:89953'],\n",
       "  'gene_symbols': ['KLC3', 'KLC2', 'KLC2L', 'KLCt', 'KNS2B']},\n",
       " {'gene_id': '295',\n",
       "  'gene_names': ['hsa:10376',\n",
       "   'hsa:112714',\n",
       "   'hsa:113457',\n",
       "   'hsa:51807',\n",
       "   'hsa:7277',\n",
       "   'hsa:7278',\n",
       "   'hsa:7846',\n",
       "   'hsa:79861',\n",
       "   'hsa:84790'],\n",
       "  'gene_symbols': ['TUBA1B', 'K-ALPHA-1']},\n",
       " {'gene_id': '296',\n",
       "  'gene_names': ['hsa:10381',\n",
       "   'hsa:10382',\n",
       "   'hsa:10383',\n",
       "   'hsa:124908013',\n",
       "   'hsa:124908015',\n",
       "   'hsa:124908041',\n",
       "   'hsa:203068',\n",
       "   'hsa:260334',\n",
       "   'hsa:347688',\n",
       "   'hsa:347733',\n",
       "   'hsa:7280',\n",
       "   'hsa:81027',\n",
       "   'hsa:84617'],\n",
       "  'gene_symbols': ['TUBB3',\n",
       "   'CDCBM',\n",
       "   'CDCBM1',\n",
       "   'CFEOM3',\n",
       "   'CFEOM3A',\n",
       "   'FEOM3',\n",
       "   'TUBB4',\n",
       "   'beta-4']},\n",
       " {'gene_id': '300',\n",
       "  'gene_names': ['hsa:6622'],\n",
       "  'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       " {'gene_id': '304',\n",
       "  'gene_names': ['hsa:65018'],\n",
       "  'gene_symbols': ['PINK1', 'BRPK', 'PARK6']},\n",
       " {'gene_id': '307',\n",
       "  'gene_names': ['hsa:5071'],\n",
       "  'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       " {'gene_id': '314',\n",
       "  'gene_names': ['hsa:65018'],\n",
       "  'gene_symbols': ['PINK1', 'BRPK', 'PARK6']},\n",
       " {'gene_id': '315',\n",
       "  'gene_names': ['hsa:5071'],\n",
       "  'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       " {'gene_id': '316',\n",
       "  'gene_names': ['hsa:581'],\n",
       "  'gene_symbols': ['BAX', 'BCL2L4']},\n",
       " {'gene_id': '323',\n",
       "  'gene_names': ['hsa:6622'],\n",
       "  'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       " {'gene_id': '326',\n",
       "  'gene_names': ['hsa:11315'],\n",
       "  'gene_symbols': ['PARK7', 'DJ-1', 'DJ1', 'GATD2', 'HEL-S-67p']},\n",
       " {'gene_id': '327',\n",
       "  'gene_names': ['hsa:598'],\n",
       "  'gene_symbols': ['BCL2L1', 'BCL-XL/S', 'BCL2L', 'BCLX', 'Bcl-X', 'PPP1R52']},\n",
       " {'gene_id': '332',\n",
       "  'gene_names': ['hsa:10131'],\n",
       "  'gene_symbols': ['TRAP1', 'HSP_75', 'HSP75', 'HSP90L', 'TRAP-1']},\n",
       " {'gene_id': '335',\n",
       "  'gene_names': ['hsa:65018'],\n",
       "  'gene_symbols': ['PINK1', 'BRPK', 'PARK6']},\n",
       " {'gene_id': '338',\n",
       "  'gene_names': ['hsa:7157'],\n",
       "  'gene_symbols': ['TP53', 'BCC7', 'BMFS5', 'LFS1', 'P53', 'TRP53']},\n",
       " {'gene_id': '343',\n",
       "  'gene_names': ['hsa:1616'],\n",
       "  'gene_symbols': ['DAXX', 'BING2', 'DAP6', 'EAP1', 'SMIM40']},\n",
       " {'gene_id': '344',\n",
       "  'gene_names': ['hsa:4217'],\n",
       "  'gene_symbols': ['MAP3K5', 'ASK1', 'MAPKKK5', 'MEKK5']},\n",
       " {'gene_id': '345',\n",
       "  'gene_names': ['hsa:5599', 'hsa:5601', 'hsa:5602'],\n",
       "  'gene_symbols': ['MAPK8',\n",
       "   'JNK',\n",
       "   'JNK-46',\n",
       "   'JNK1',\n",
       "   'JNK1A2',\n",
       "   'JNK21B1/2',\n",
       "   'PRKM8',\n",
       "   'SAPK1',\n",
       "   'SAPK1c']},\n",
       " {'gene_id': '347',\n",
       "  'gene_names': ['hsa:4780'],\n",
       "  'gene_symbols': ['NFE2L2', 'HEBP1', 'IMDDHH', 'NRF2', 'Nrf-2']},\n",
       " {'gene_id': '361',\n",
       "  'gene_names': ['hsa:1843'],\n",
       "  'gene_symbols': ['DUSP1', 'CL100', 'HVH1', 'MKP-1', 'MKP1', 'PTPN10']},\n",
       " {'gene_id': '362',\n",
       "  'gene_names': ['hsa:25828', 'hsa:7295'],\n",
       "  'gene_symbols': ['TXN2', 'COXPD29', 'MT-TRX', 'MTRX', 'TRX2', 'TXN']},\n",
       " {'gene_id': '364',\n",
       "  'gene_names': ['hsa:6570', 'hsa:6571'],\n",
       "  'gene_symbols': ['SLC18A1', 'CGAT', 'VAT1', 'VMAT1']},\n",
       " {'gene_id': '365',\n",
       "  'gene_names': ['hsa:6622'],\n",
       "  'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       " {'gene_id': '368',\n",
       "  'gene_names': ['hsa:5071'],\n",
       "  'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       " {'gene_id': '374',\n",
       "  'gene_names': ['hsa:4128', 'hsa:4129'],\n",
       "  'gene_symbols': ['MAOA', 'BRNRS', 'MAO-A']},\n",
       " {'gene_id': '375',\n",
       "  'gene_names': ['hsa:5071'],\n",
       "  'gene_symbols': ['PRKN', 'AR-JP', 'LPRS2', 'PARK2', 'PDJ']},\n",
       " {'gene_id': '382',\n",
       "  'gene_names': ['hsa:90550'],\n",
       "  'gene_symbols': ['MCU', 'C10orf42', 'CCDC109A', 'HsMCU']},\n",
       " {'gene_id': '398',\n",
       "  'gene_names': ['hsa:55669', 'hsa:9927'],\n",
       "  'gene_symbols': ['MFN1', 'hfzo1', 'hfzo2']},\n",
       " {'gene_id': '402',\n",
       "  'gene_names': ['hsa:6622'],\n",
       "  'gene_symbols': ['SNCA', 'NACP', 'PARK1', 'PARK4', 'PD1']},\n",
       " {'gene_id': '711',\n",
       "  'gene_names': ['hsa:6531'],\n",
       "  'gene_symbols': ['SLC6A3', 'DAT', 'DAT1', 'PKDYS', 'PKDYS1']},\n",
       " {'gene_id': '714',\n",
       "  'gene_names': ['hsa:9817'],\n",
       "  'gene_symbols': ['KEAP1', 'INrf2', 'KLHL19']},\n",
       " {'gene_id': '715',\n",
       "  'gene_names': ['hsa:25828', 'hsa:7295'],\n",
       "  'gene_symbols': ['TXN2', 'COXPD29', 'MT-TRX', 'MTRX', 'TRX2', 'TXN']},\n",
       " {'gene_id': '724',\n",
       "  'gene_names': ['hsa:201266',\n",
       "   'hsa:221074',\n",
       "   'hsa:23516',\n",
       "   'hsa:25800',\n",
       "   'hsa:27173',\n",
       "   'hsa:283375',\n",
       "   'hsa:29985',\n",
       "   'hsa:29986',\n",
       "   'hsa:4891',\n",
       "   'hsa:55334',\n",
       "   'hsa:55630',\n",
       "   'hsa:57181',\n",
       "   'hsa:64116',\n",
       "   'hsa:7922',\n",
       "   'hsa:91252'],\n",
       "  'gene_symbols': ['SLC39A11', 'C17orf26', 'ZIP-11', 'ZIP11']},\n",
       " {'gene_id': '735',\n",
       "  'gene_names': ['hsa:6647'],\n",
       "  'gene_symbols': ['SOD1',\n",
       "   'ALS',\n",
       "   'ALS1',\n",
       "   'HEL-S-44',\n",
       "   'IPOA',\n",
       "   'SOD',\n",
       "   'STAHP',\n",
       "   'hSod1',\n",
       "   'homodimer']}]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_genes=pathway_data['genes']\n",
    "kegg_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbols</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>LRRK2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>AURA17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>DARDARIN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>PARK8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>RIPK7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>735</td>\n",
       "      <td>IPOA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>735</td>\n",
       "      <td>SOD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>735</td>\n",
       "      <td>STAHP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>735</td>\n",
       "      <td>hSod1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>735</td>\n",
       "      <td>homodimer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>557 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_id gene_symbols\n",
       "0         5        LRRK2\n",
       "0         5       AURA17\n",
       "0         5     DARDARIN\n",
       "0         5        PARK8\n",
       "0         5        RIPK7\n",
       "..      ...          ...\n",
       "102     735         IPOA\n",
       "102     735          SOD\n",
       "102     735        STAHP\n",
       "102     735        hSod1\n",
       "102     735    homodimer\n",
       "\n",
       "[557 rows x 2 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_df=pd.DataFrame(kegg_genes,columns=['gene_id','gene_symbols'])\n",
    "kegg_exploded=kegg_df.explode('gene_symbols')\n",
    "kegg_exploded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "column_names = [\n",
    "    'DB', 'DB_Object_ID', 'DB_Object_Symbol', 'Qualifier', 'GO_ID', \n",
    "    'DB:Reference', 'Evidence_Code', 'With_or_From', 'Aspect', \n",
    "    'DB_Object_Name', 'DB_Object_Synonym', 'DB_Object_Type', 'Taxon', \n",
    "    'Date', 'Assigned_By', 'Annotation_Extension', 'Gene_Product_Form_ID'\n",
    "]\n",
    "# Read the GAF file into a DataFrame, skipping comment lines\n",
    "gaf_df = pd.read_csv(gaf_path, sep='\\t', comment='!', header=None, names=column_names, dtype=str)\n",
    "gaf_df=gaf_df[gaf_df['DB_Object_Synonym'].notna()]\n",
    "gaf_df['DB_Object_Synonym'] = gaf_df['DB_Object_Synonym'].apply(lambda x: x.split('|') if pd.notna(x) else [])\n",
    "gaf_df['Taxon'] = gaf_df['Taxon'].apply(lambda x: x.split('|') if pd.notna(x) else [])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "53628     [SERF1A, SERF1B, FAM2A, FAM2B, SERF1, SERF1, S...\n",
       "53629     [SERF1A, SERF1B, FAM2A, FAM2B, SERF1, SERF1, S...\n",
       "53630     [SERF1A, SERF1B, FAM2A, FAM2B, SERF1, SERF1, S...\n",
       "53631     [SERF1A, SERF1B, FAM2A, FAM2B, SERF1, SERF1, S...\n",
       "53632     [SERF1A, SERF1B, FAM2A, FAM2B, SERF1, SERF1, S...\n",
       "                                ...                        \n",
       "701167     [MAGEA2, MAGEA2B, MAGE2, MAGE2, MAGEA2, MAGEA2A]\n",
       "701362    [CCL4L1, CCL4L2, CCL4L, CCL4L, LAG1, SCYA4L1, ...\n",
       "704881    [PABPC1L2A, PABPC1L2B, PABPC1L2, PABPC1L2, RBM...\n",
       "705150             [HSFY1, HSFY2, HSF2L, HSF2L, HSFY, HSFY]\n",
       "706480    [CCL4L1, CCL4L2, CCL4L, CCL4L, LAG1, SCYA4L1, ...\n",
       "Name: DB_Object_Synonym, Length: 221, dtype: object"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check if synonyms list has duplicates in each line-yes!\n",
    "gaf_df[gaf_df['DB_Object_Synonym'].apply(set).apply(len)!=gaf_df['DB_Object_Synonym'].apply(len)]['DB_Object_Synonym']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DB</th>\n",
       "      <th>DB_Object_ID</th>\n",
       "      <th>DB_Object_Symbol</th>\n",
       "      <th>Qualifier</th>\n",
       "      <th>GO_ID</th>\n",
       "      <th>DB:Reference</th>\n",
       "      <th>Evidence_Code</th>\n",
       "      <th>With_or_From</th>\n",
       "      <th>Aspect</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "      <th>DB_Object_Synonym</th>\n",
       "      <th>DB_Object_Type</th>\n",
       "      <th>Taxon</th>\n",
       "      <th>Date</th>\n",
       "      <th>Assigned_By</th>\n",
       "      <th>Annotation_Extension</th>\n",
       "      <th>Gene_Product_Form_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A024RBG1</td>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003723</td>\n",
       "      <td>GO_REF:0000043</td>\n",
       "      <td>IEA</td>\n",
       "      <td>UniProtKB-KW:KW-0694</td>\n",
       "      <td>F</td>\n",
       "      <td>Diphosphoinositol polyphosphate phosphohydrola...</td>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>protein</td>\n",
       "      <td>[taxon:9606]</td>\n",
       "      <td>20240408</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A024RBG1</td>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0046872</td>\n",
       "      <td>GO_REF:0000043</td>\n",
       "      <td>IEA</td>\n",
       "      <td>UniProtKB-KW:KW-0479</td>\n",
       "      <td>F</td>\n",
       "      <td>Diphosphoinositol polyphosphate phosphohydrola...</td>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>protein</td>\n",
       "      <td>[taxon:9606]</td>\n",
       "      <td>20240408</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A024RBG1</td>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>located_in</td>\n",
       "      <td>GO:0005829</td>\n",
       "      <td>GO_REF:0000052</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>Diphosphoinositol polyphosphate phosphohydrola...</td>\n",
       "      <td>NUDT4B</td>\n",
       "      <td>protein</td>\n",
       "      <td>[taxon:9606]</td>\n",
       "      <td>20230619</td>\n",
       "      <td>HPA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A075B6H7</td>\n",
       "      <td>IGKV3-7</td>\n",
       "      <td>involved_in</td>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>GO_REF:0000043</td>\n",
       "      <td>IEA</td>\n",
       "      <td>UniProtKB-KW:KW-1064</td>\n",
       "      <td>P</td>\n",
       "      <td>Probable non-functional immunoglobulin kappa v...</td>\n",
       "      <td>IGKV3-7</td>\n",
       "      <td>protein</td>\n",
       "      <td>[taxon:9606]</td>\n",
       "      <td>20240408</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UniProtKB</td>\n",
       "      <td>A0A075B6H7</td>\n",
       "      <td>IGKV3-7</td>\n",
       "      <td>located_in</td>\n",
       "      <td>GO:0005886</td>\n",
       "      <td>GO_REF:0000044</td>\n",
       "      <td>IEA</td>\n",
       "      <td>UniProtKB-SubCell:SL-0039</td>\n",
       "      <td>C</td>\n",
       "      <td>Probable non-functional immunoglobulin kappa v...</td>\n",
       "      <td>IGKV3-7</td>\n",
       "      <td>protein</td>\n",
       "      <td>[taxon:9606]</td>\n",
       "      <td>20240408</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          DB DB_Object_ID DB_Object_Symbol    Qualifier       GO_ID  \\\n",
       "0  UniProtKB   A0A024RBG1           NUDT4B      enables  GO:0003723   \n",
       "1  UniProtKB   A0A024RBG1           NUDT4B      enables  GO:0046872   \n",
       "2  UniProtKB   A0A024RBG1           NUDT4B   located_in  GO:0005829   \n",
       "3  UniProtKB   A0A075B6H7          IGKV3-7  involved_in  GO:0002250   \n",
       "4  UniProtKB   A0A075B6H7          IGKV3-7   located_in  GO:0005886   \n",
       "\n",
       "     DB:Reference Evidence_Code               With_or_From Aspect  \\\n",
       "0  GO_REF:0000043           IEA       UniProtKB-KW:KW-0694      F   \n",
       "1  GO_REF:0000043           IEA       UniProtKB-KW:KW-0479      F   \n",
       "2  GO_REF:0000052           IDA                        NaN      C   \n",
       "3  GO_REF:0000043           IEA       UniProtKB-KW:KW-1064      P   \n",
       "4  GO_REF:0000044           IEA  UniProtKB-SubCell:SL-0039      C   \n",
       "\n",
       "                                      DB_Object_Name DB_Object_Synonym  \\\n",
       "0  Diphosphoinositol polyphosphate phosphohydrola...            NUDT4B   \n",
       "1  Diphosphoinositol polyphosphate phosphohydrola...            NUDT4B   \n",
       "2  Diphosphoinositol polyphosphate phosphohydrola...            NUDT4B   \n",
       "3  Probable non-functional immunoglobulin kappa v...           IGKV3-7   \n",
       "4  Probable non-functional immunoglobulin kappa v...           IGKV3-7   \n",
       "\n",
       "  DB_Object_Type         Taxon      Date Assigned_By Annotation_Extension  \\\n",
       "0        protein  [taxon:9606]  20240408     UniProt                  NaN   \n",
       "1        protein  [taxon:9606]  20240408     UniProt                  NaN   \n",
       "2        protein  [taxon:9606]  20230619         HPA                  NaN   \n",
       "3        protein  [taxon:9606]  20240408     UniProt                  NaN   \n",
       "4        protein  [taxon:9606]  20240408     UniProt                  NaN   \n",
       "\n",
       "  Gene_Product_Form_ID  \n",
       "0                  NaN  \n",
       "1                  NaN  \n",
       "2                  NaN  \n",
       "3                  NaN  \n",
       "4                  NaN  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gaf_exploded=gaf_df.explode('DB_Object_Synonym')\n",
    "gaf_exploded.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Qualifier</th>\n",
       "      <th>GO_ID</th>\n",
       "      <th>Aspect</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "      <th>DB_Object_Synonym</th>\n",
       "      <th>DB_Object_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0034260</td>\n",
       "      <td>P</td>\n",
       "      <td>Leucine-rich repeat serine/threonine-protein k...</td>\n",
       "      <td>LRRK2</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0000149</td>\n",
       "      <td>F</td>\n",
       "      <td>Leucine-rich repeat serine/threonine-protein k...</td>\n",
       "      <td>LRRK2</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0000287</td>\n",
       "      <td>F</td>\n",
       "      <td>Leucine-rich repeat serine/threonine-protein k...</td>\n",
       "      <td>LRRK2</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003779</td>\n",
       "      <td>F</td>\n",
       "      <td>Leucine-rich repeat serine/threonine-protein k...</td>\n",
       "      <td>LRRK2</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>F</td>\n",
       "      <td>Leucine-rich repeat serine/threonine-protein k...</td>\n",
       "      <td>LRRK2</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60807</th>\n",
       "      <td>735</td>\n",
       "      <td>located_in</td>\n",
       "      <td>GO:0070062</td>\n",
       "      <td>C</td>\n",
       "      <td>Insulin-like growth factor-binding protein com...</td>\n",
       "      <td>ALS</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60809</th>\n",
       "      <td>735</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0042567</td>\n",
       "      <td>C</td>\n",
       "      <td>Insulin-like growth factor-binding protein com...</td>\n",
       "      <td>ALS</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60810</th>\n",
       "      <td>735</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005520</td>\n",
       "      <td>F</td>\n",
       "      <td>Insulin-like growth factor-binding protein com...</td>\n",
       "      <td>ALS</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60811</th>\n",
       "      <td>735</td>\n",
       "      <td>is_active_in</td>\n",
       "      <td>GO:0031012</td>\n",
       "      <td>C</td>\n",
       "      <td>Insulin-like growth factor-binding protein com...</td>\n",
       "      <td>ALS</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60812</th>\n",
       "      <td>735</td>\n",
       "      <td>is_active_in</td>\n",
       "      <td>GO:0005615</td>\n",
       "      <td>C</td>\n",
       "      <td>Insulin-like growth factor-binding protein com...</td>\n",
       "      <td>ALS</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14374 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id                   Qualifier       GO_ID Aspect  \\\n",
       "0           5  acts_upstream_of_or_within  GO:0034260      P   \n",
       "1           5                     enables  GO:0000149      F   \n",
       "2           5                     enables  GO:0000287      F   \n",
       "3           5                     enables  GO:0003779      F   \n",
       "4           5                     enables  GO:0003924      F   \n",
       "...       ...                         ...         ...    ...   \n",
       "60807     735                  located_in  GO:0070062      C   \n",
       "60809     735                     part_of  GO:0042567      C   \n",
       "60810     735                     enables  GO:0005520      F   \n",
       "60811     735                is_active_in  GO:0031012      C   \n",
       "60812     735                is_active_in  GO:0005615      C   \n",
       "\n",
       "                                          DB_Object_Name DB_Object_Synonym  \\\n",
       "0      Leucine-rich repeat serine/threonine-protein k...             LRRK2   \n",
       "1      Leucine-rich repeat serine/threonine-protein k...             LRRK2   \n",
       "2      Leucine-rich repeat serine/threonine-protein k...             LRRK2   \n",
       "3      Leucine-rich repeat serine/threonine-protein k...             LRRK2   \n",
       "4      Leucine-rich repeat serine/threonine-protein k...             LRRK2   \n",
       "...                                                  ...               ...   \n",
       "60807  Insulin-like growth factor-binding protein com...               ALS   \n",
       "60809  Insulin-like growth factor-binding protein com...               ALS   \n",
       "60810  Insulin-like growth factor-binding protein com...               ALS   \n",
       "60811  Insulin-like growth factor-binding protein com...               ALS   \n",
       "60812  Insulin-like growth factor-binding protein com...               ALS   \n",
       "\n",
       "      DB_Object_Type  \n",
       "0            protein  \n",
       "1            protein  \n",
       "2            protein  \n",
       "3            protein  \n",
       "4            protein  \n",
       "...              ...  \n",
       "60807        protein  \n",
       "60809        protein  \n",
       "60810        protein  \n",
       "60811        protein  \n",
       "60812        protein  \n",
       "\n",
       "[14374 rows x 7 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge kegg and gaf on common synonyms\n",
    "merged=pd.merge(kegg_exploded,gaf_exploded,left_on='gene_symbols', right_on='DB_Object_Synonym', how='inner')[['gene_id','Qualifier','GO_ID','Aspect','DB_Object_Name','DB_Object_Synonym','DB_Object_Type']].drop_duplicates()\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ADORA2A': 'adenosine A2a receptor',\n",
       " 'GNAL': 'G protein subunit alpha L',\n",
       " 'GNAS': 'GNAS complex locus',\n",
       " 'DRD2': 'dopamine receptor D2',\n",
       " 'GNAI1': 'G protein subunit alpha i1',\n",
       " 'GNAI3': 'G protein subunit alpha i3',\n",
       " 'GNAI2': 'G protein subunit alpha i2',\n",
       " 'ADCY5': 'adenylate cyclase 5',\n",
       " 'PRKACA': 'protein kinase cAMP-activated catalytic subunit alpha',\n",
       " 'PRKACB': 'protein kinase cAMP-activated catalytic subunit beta',\n",
       " 'PRKACG': 'protein kinase cAMP-activated catalytic subunit gamma',\n",
       " 'DRD1': 'dopamine receptor D1',\n",
       " 'UBA1': 'ubiquitin like modifier activating enzyme 1',\n",
       " 'UBA7': 'ubiquitin like modifier activating enzyme 7',\n",
       " 'UBA52': 'ubiquitin A-52 residue ribosomal protein fusion product 1',\n",
       " 'RPS27A': 'ribosomal protein S27a',\n",
       " 'UBB': 'ubiquitin B',\n",
       " 'UBC': 'ubiquitin C',\n",
       " 'UBE2L3': 'ubiquitin conjugating enzyme E2 L3',\n",
       " 'UBE2L6': 'ubiquitin conjugating enzyme E2 L6',\n",
       " 'UBE2J2': 'ubiquitin conjugating enzyme E2 J2',\n",
       " 'UBE2J1': 'ubiquitin conjugating enzyme E2 J1',\n",
       " 'UBE2G2': 'ubiquitin conjugating enzyme E2 G2',\n",
       " 'UBE2G1': 'ubiquitin conjugating enzyme E2 G1',\n",
       " 'PRKN': 'parkin RBR E3 ubiquitin protein ligase',\n",
       " 'SNCA': 'synuclein alpha',\n",
       " 'GPR37': 'G protein-coupled receptor 37',\n",
       " 'SEPTIN5': 'septin 5',\n",
       " 'SNCAIP': 'synuclein alpha interacting protein',\n",
       " 'PSMA6': 'proteasome 20S subunit alpha 6',\n",
       " 'PSMA2': 'proteasome 20S subunit alpha 2',\n",
       " 'PSMA4': 'proteasome 20S subunit alpha 4',\n",
       " 'PSMA7': 'proteasome 20S subunit alpha 7',\n",
       " 'PSMA8': 'proteasome 20S subunit alpha 8',\n",
       " 'PSMA5': 'proteasome 20S subunit alpha 5',\n",
       " 'PSMA1': 'proteasome 20S subunit alpha 1',\n",
       " 'PSMA3': 'proteasome 20S subunit alpha 3',\n",
       " 'PSMB6': 'proteasome 20S subunit beta 6',\n",
       " 'PSMB7': 'proteasome 20S subunit beta 7',\n",
       " 'PSMB3': 'proteasome 20S subunit beta 3',\n",
       " 'PSMB2': 'proteasome 20S subunit beta 2',\n",
       " 'PSMB5': 'proteasome 20S subunit beta 5',\n",
       " 'PSMB1': 'proteasome 20S subunit beta 1',\n",
       " 'PSMB4': 'proteasome 20S subunit beta 4',\n",
       " 'PSMC2': 'proteasome 26S subunit, ATPase 2',\n",
       " 'PSMC1': 'proteasome 26S subunit, ATPase 1',\n",
       " 'PSMC4': 'proteasome 26S subunit, ATPase 4',\n",
       " 'PSMC6': 'proteasome 26S subunit, ATPase 6',\n",
       " 'PSMC3': 'proteasome 26S subunit, ATPase 3',\n",
       " 'PSMC5': 'proteasome 26S subunit, ATPase 5',\n",
       " 'PSMD2': 'proteasome 26S subunit ubiquitin receptor, non-ATPase 2',\n",
       " 'PSMD1': 'proteasome 26S subunit, non-ATPase 1',\n",
       " 'PSMD3': 'proteasome 26S subunit, non-ATPase 3',\n",
       " 'PSMD9': 'proteasome 26S subunit, non-ATPase 9',\n",
       " 'PSMD12': 'proteasome 26S subunit, non-ATPase 12',\n",
       " 'PSMD11': 'proteasome 26S subunit, non-ATPase 11',\n",
       " 'PSMD6': 'proteasome 26S subunit, non-ATPase 6',\n",
       " 'PSMD7': 'proteasome 26S subunit, non-ATPase 7',\n",
       " 'PSMD13': 'proteasome 26S subunit, non-ATPase 13',\n",
       " 'PSMD4': 'proteasome 26S subunit ubiquitin receptor, non-ATPase 4',\n",
       " 'PSMD14': 'proteasome 26S subunit, non-ATPase 14',\n",
       " 'PSMD8': 'proteasome 26S subunit, non-ATPase 8',\n",
       " 'ADRM1': 'ADRM1 26S proteasome ubiquitin receptor',\n",
       " 'SEM1': 'SEM1 26S proteasome subunit',\n",
       " 'UCHL1': 'ubiquitin C-terminal hydrolase L1',\n",
       " 'CALML3': 'calmodulin like 3',\n",
       " 'CALM2': 'calmodulin 2',\n",
       " 'CALM3': 'calmodulin 3',\n",
       " 'CALM1': 'calmodulin 1',\n",
       " 'CALML6': 'calmodulin like 6',\n",
       " 'CALML5': 'calmodulin like 5',\n",
       " 'CALML4': 'calmodulin like 4',\n",
       " 'CAMK2A': 'calcium/calmodulin dependent protein kinase II alpha',\n",
       " 'CAMK2D': 'calcium/calmodulin dependent protein kinase II delta',\n",
       " 'CAMK2B': 'calcium/calmodulin dependent protein kinase II beta',\n",
       " 'CAMK2G': 'calcium/calmodulin dependent protein kinase II gamma',\n",
       " 'MAPT': 'microtubule associated protein tau',\n",
       " 'RYR3': 'ryanodine receptor 3',\n",
       " 'ITPR1': 'inositol 1,4,5-trisphosphate receptor type 1',\n",
       " 'ITPR2': 'inositol 1,4,5-trisphosphate receptor type 2',\n",
       " 'ITPR3': 'inositol 1,4,5-trisphosphate receptor type 3',\n",
       " 'PLCG1': 'phospholipase C gamma 1',\n",
       " 'HSPA5': 'heat shock protein family A (Hsp70) member 5',\n",
       " 'ATF6': 'activating transcription factor 6',\n",
       " 'DDIT3': 'DNA damage inducible transcript 3',\n",
       " 'ERN1': 'endoplasmic reticulum to nucleus signaling 1',\n",
       " 'XBP1': 'X-box binding protein 1',\n",
       " 'EIF2AK3': 'eukaryotic translation initiation factor 2 alpha kinase 3',\n",
       " 'EIF2S1': 'eukaryotic translation initiation factor 2 subunit alpha',\n",
       " 'ATF4': 'activating transcription factor 4',\n",
       " 'TH': 'tyrosine hydroxylase',\n",
       " 'MAOB': 'monoamine oxidase B',\n",
       " 'MAOA': 'monoamine oxidase A',\n",
       " 'SLC6A3': 'solute carrier family 6 member 3',\n",
       " 'SLC18A1': 'solute carrier family 18 member A1',\n",
       " 'SLC18A2': 'solute carrier family 18 member A2',\n",
       " 'SLC11A2': 'solute carrier family 11 member 2',\n",
       " 'SLC39A2': 'solute carrier family 39 member 2',\n",
       " 'SLC39A3': 'solute carrier family 39 member 3',\n",
       " 'SLC39A1': 'solute carrier family 39 member 1',\n",
       " 'SLC39A4': 'solute carrier family 39 member 4',\n",
       " 'SLC39A5': 'solute carrier family 39 member 5',\n",
       " 'SLC39A6': 'solute carrier family 39 member 6',\n",
       " 'SLC39A7': 'solute carrier family 39 member 7',\n",
       " 'SLC39A8': 'solute carrier family 39 member 8',\n",
       " 'SLC39A9': 'solute carrier family 39 member 9',\n",
       " 'SLC39A10': 'solute carrier family 39 member 10',\n",
       " 'SLC39A11': 'solute carrier family 39 member 11',\n",
       " 'SLC39A12': 'solute carrier family 39 member 12',\n",
       " 'SLC39A13': 'solute carrier family 39 member 13',\n",
       " 'SLC39A14': 'solute carrier family 39 member 14',\n",
       " 'ND1': 'NADH dehydrogenase subunit 1',\n",
       " 'ND2': 'NADH dehydrogenase subunit 2',\n",
       " 'ND3': 'NADH dehydrogenase subunit 3',\n",
       " 'ND4': 'NADH dehydrogenase subunit 4',\n",
       " 'ND4L': 'NADH dehydrogenase subunit 4L',\n",
       " 'ND5': 'NADH dehydrogenase subunit 5',\n",
       " 'ND6': 'NADH dehydrogenase subunit 6',\n",
       " 'NDUFV1': 'NADH:ubiquinone oxidoreductase core subunit V1',\n",
       " 'NDUFV2': 'NADH:ubiquinone oxidoreductase core subunit V2',\n",
       " 'NDUFV3': 'NADH:ubiquinone oxidoreductase subunit V3',\n",
       " 'NDUFA1': 'NADH:ubiquinone oxidoreductase subunit A1',\n",
       " 'NDUFA2': 'NADH:ubiquinone oxidoreductase subunit A2',\n",
       " 'NDUFA3': 'NADH:ubiquinone oxidoreductase subunit A3',\n",
       " 'NDUFA4': 'NDUFA4 mitochondrial complex associated',\n",
       " 'NDUFA4L2': 'NDUFA4 mitochondrial complex associated like 2',\n",
       " 'NDUFA5': 'NADH:ubiquinone oxidoreductase subunit A5',\n",
       " 'NDUFA6': 'NADH:ubiquinone oxidoreductase subunit A6',\n",
       " 'NDUFA7': 'NADH:ubiquinone oxidoreductase subunit A7',\n",
       " 'NDUFA8': 'NADH:ubiquinone oxidoreductase subunit A8',\n",
       " 'NDUFA9': 'NADH:ubiquinone oxidoreductase subunit A9',\n",
       " 'NDUFA10': 'NADH:ubiquinone oxidoreductase subunit A10',\n",
       " 'NDUFAB1': 'NADH:ubiquinone oxidoreductase subunit AB1',\n",
       " 'NDUFA11': 'NADH:ubiquinone oxidoreductase subunit A11',\n",
       " 'NDUFA12': 'NADH:ubiquinone oxidoreductase subunit A12',\n",
       " 'NDUFA13': 'NADH:ubiquinone oxidoreductase subunit A13',\n",
       " 'NDUFB1': 'NADH:ubiquinone oxidoreductase subunit B1',\n",
       " 'NDUFB2': 'NADH:ubiquinone oxidoreductase subunit B2',\n",
       " 'NDUFB3': 'NADH:ubiquinone oxidoreductase subunit B3',\n",
       " 'NDUFB4': 'NADH:ubiquinone oxidoreductase subunit B4',\n",
       " 'NDUFB5': 'NADH:ubiquinone oxidoreductase subunit B5',\n",
       " 'NDUFB6': 'NADH:ubiquinone oxidoreductase subunit B6',\n",
       " 'NDUFB7': 'NADH:ubiquinone oxidoreductase subunit B7',\n",
       " 'NDUFB8': 'NADH:ubiquinone oxidoreductase subunit B8',\n",
       " 'NDUFB9': 'NADH:ubiquinone oxidoreductase subunit B9',\n",
       " 'NDUFB10': 'NADH:ubiquinone oxidoreductase subunit B10',\n",
       " 'NDUFB11': 'NADH:ubiquinone oxidoreductase subunit B11',\n",
       " 'NDUFS1': 'NADH:ubiquinone oxidoreductase core subunit S1',\n",
       " 'NDUFS2': 'NADH:ubiquinone oxidoreductase core subunit S2',\n",
       " 'NDUFS3': 'NADH:ubiquinone oxidoreductase core subunit S3',\n",
       " 'NDUFS4': 'NADH:ubiquinone oxidoreductase subunit S4',\n",
       " 'NDUFS5': 'NADH:ubiquinone oxidoreductase subunit S5',\n",
       " 'NDUFS6': 'NADH:ubiquinone oxidoreductase subunit S6',\n",
       " 'NDUFS7': 'NADH:ubiquinone oxidoreductase core subunit S7',\n",
       " 'NDUFS8': 'NADH:ubiquinone oxidoreductase core subunit S8',\n",
       " 'NDUFC1': 'NADH:ubiquinone oxidoreductase subunit C1',\n",
       " 'NDUFC2': 'NADH:ubiquinone oxidoreductase subunit C2',\n",
       " 'NDUFC2-KCTD14': 'NDUFC2-KCTD14 readthrough',\n",
       " 'SDHA': 'succinate dehydrogenase complex flavoprotein subunit A',\n",
       " 'SDHB': 'succinate dehydrogenase complex iron sulfur subunit B',\n",
       " 'SDHC': 'succinate dehydrogenase complex subunit C',\n",
       " 'SDHD': 'succinate dehydrogenase complex subunit D',\n",
       " 'UQCRFS1': 'ubiquinol-cytochrome c reductase, Rieske iron-sulfur polypeptide 1',\n",
       " 'CYTB': 'cytochrome b',\n",
       " 'CYC1': 'cytochrome c1',\n",
       " 'UQCRC1': 'ubiquinol-cytochrome c reductase core protein 1',\n",
       " 'UQCRC2': 'ubiquinol-cytochrome c reductase core protein 2',\n",
       " 'UQCRH': 'ubiquinol-cytochrome c reductase hinge protein',\n",
       " 'UQCRHL': 'ubiquinol-cytochrome c reductase hinge protein like',\n",
       " 'UQCRB': 'ubiquinol-cytochrome c reductase binding protein',\n",
       " 'UQCRQ': 'ubiquinol-cytochrome c reductase complex III subunit VII',\n",
       " 'UQCR10': 'ubiquinol-cytochrome c reductase, complex III subunit X',\n",
       " 'UQCR11': 'ubiquinol-cytochrome c reductase, complex III subunit XI',\n",
       " 'COX3': 'cytochrome c oxidase subunit III',\n",
       " 'COX1': 'cytochrome c oxidase subunit I',\n",
       " 'COX2': 'cytochrome c oxidase subunit II',\n",
       " 'COX4I2': 'cytochrome c oxidase subunit 4I2',\n",
       " 'COX4I1': 'cytochrome c oxidase subunit 4I1',\n",
       " 'COX5A': 'cytochrome c oxidase subunit 5A',\n",
       " 'COX5B': 'cytochrome c oxidase subunit 5B',\n",
       " 'COX6A1': 'cytochrome c oxidase subunit 6A1',\n",
       " 'COX6A2': 'cytochrome c oxidase subunit 6A2',\n",
       " 'COX6B1': 'cytochrome c oxidase subunit 6B1',\n",
       " 'COX6B2': 'cytochrome c oxidase subunit 6B2',\n",
       " 'COX6C': 'cytochrome c oxidase subunit 6C',\n",
       " 'COX7A1': 'cytochrome c oxidase subunit 7A1',\n",
       " 'COX7A2': 'cytochrome c oxidase subunit 7A2',\n",
       " 'COX7A2L': 'cytochrome c oxidase subunit 7A2 like',\n",
       " 'COX7B': 'cytochrome c oxidase subunit 7B',\n",
       " 'COX7B2': 'cytochrome c oxidase subunit 7B2',\n",
       " 'COX7C': 'cytochrome c oxidase subunit 7C',\n",
       " 'COX8C': 'cytochrome c oxidase subunit 8C',\n",
       " 'COX8A': 'cytochrome c oxidase subunit 8A',\n",
       " 'ATP5F1A': 'ATP synthase F1 subunit alpha',\n",
       " 'ATP5F1B': 'ATP synthase F1 subunit beta',\n",
       " 'ATP5F1C': 'ATP synthase F1 subunit gamma',\n",
       " 'ATP5F1D': 'ATP synthase F1 subunit delta',\n",
       " 'ATP5F1E': 'ATP synthase F1 subunit epsilon',\n",
       " 'ATP6': 'ATP synthase F0 subunit 6',\n",
       " 'ATP5PB': 'ATP synthase peripheral stalk-membrane subunit b',\n",
       " 'ATP5MC1': 'ATP synthase membrane subunit c locus 1',\n",
       " 'ATP5MC2': 'ATP synthase membrane subunit c locus 2',\n",
       " 'ATP5MC3': 'ATP synthase membrane subunit c locus 3',\n",
       " 'ATP5PD': 'ATP synthase peripheral stalk subunit d',\n",
       " 'ATP5PO': 'ATP synthase peripheral stalk subunit OSCP',\n",
       " 'ATP5PF': 'ATP synthase peripheral stalk subunit F6',\n",
       " 'ATP8': 'ATP synthase F0 subunit 8',\n",
       " 'PINK1': 'PTEN induced kinase 1',\n",
       " 'MCU': 'mitochondrial calcium uniporter',\n",
       " 'LRRK2': 'leucine rich repeat kinase 2',\n",
       " 'HTRA2': 'HtrA serine peptidase 2',\n",
       " 'TRAP1': 'TNF receptor associated protein 1',\n",
       " 'VDAC1': 'voltage dependent anion channel 1',\n",
       " 'VDAC2': 'voltage dependent anion channel 2',\n",
       " 'VDAC3': 'voltage dependent anion channel 3',\n",
       " 'SLC25A4': 'solute carrier family 25 member 4',\n",
       " 'SLC25A5': 'solute carrier family 25 member 5',\n",
       " 'SLC25A6': 'solute carrier family 25 member 6',\n",
       " 'SLC25A31': 'solute carrier family 25 member 31',\n",
       " 'PPIF': 'peptidylprolyl isomerase F',\n",
       " 'CYCS': 'cytochrome c, somatic',\n",
       " 'APAF1': 'apoptotic peptidase activating factor 1',\n",
       " 'CASP9': 'caspase 9',\n",
       " 'CASP3': 'caspase 3',\n",
       " 'PARK7': 'Parkinsonism associated deglycase',\n",
       " 'BCL2L1': 'BCL2 like 1',\n",
       " 'BAX': 'BCL2 associated X, apoptosis regulator',\n",
       " 'MFN1': 'mitofusin 1',\n",
       " 'MFN2': 'mitofusin 2',\n",
       " 'KIF5A': 'kinesin family member 5A',\n",
       " 'KIF5B': 'kinesin family member 5B',\n",
       " 'KIF5C': 'kinesin family member 5C',\n",
       " 'KLC3': 'kinesin light chain 3',\n",
       " 'KLC1': 'kinesin light chain 1',\n",
       " 'KLC2': 'kinesin light chain 2',\n",
       " 'KLC4': 'kinesin light chain 4',\n",
       " 'TUBA1B': 'tubulin alpha 1b',\n",
       " 'TUBA4A': 'tubulin alpha 4a',\n",
       " 'TUBA3C': 'tubulin alpha 3c',\n",
       " 'TUBA1A': 'tubulin alpha 1a',\n",
       " 'TUBA1C': 'tubulin alpha 1c',\n",
       " 'TUBA8': 'tubulin alpha 8',\n",
       " 'TUBA3E': 'tubulin alpha 3e',\n",
       " 'TUBA3D': 'tubulin alpha 3d',\n",
       " 'TUBAL3': 'tubulin alpha like 3',\n",
       " 'TUBB6': 'tubulin beta 6 class V',\n",
       " 'TUBB': 'tubulin beta class I',\n",
       " 'TUBB1': 'tubulin beta 1 class VI',\n",
       " 'TUBB2A': 'tubulin beta 2A class IIa',\n",
       " 'TUBB3': 'tubulin beta 3 class III',\n",
       " 'TUBB4A': 'tubulin beta 4A class IVa',\n",
       " 'TUBB8': 'tubulin beta 8 class VIII',\n",
       " 'TUBB2B': 'tubulin beta 2B class IIb',\n",
       " 'TUBB4B': 'tubulin beta 4B class IVb',\n",
       " 'TUBB8B': 'tubulin beta 8B',\n",
       " 'TP53': 'tumor protein p53',\n",
       " 'DUSP1': 'dual specificity phosphatase 1',\n",
       " 'NFE2L2': 'NFE2 like bZIP transcription factor 2',\n",
       " 'TXN': 'thioredoxin',\n",
       " 'TXN2': 'thioredoxin 2',\n",
       " 'DAXX': 'death domain associated protein',\n",
       " 'MAP3K5': 'mitogen-activated protein kinase kinase kinase 5',\n",
       " 'MAPK8': 'mitogen-activated protein kinase 8',\n",
       " 'MAPK10': 'mitogen-activated protein kinase 10',\n",
       " 'MAPK9': 'mitogen-activated protein kinase 9',\n",
       " 'SOD1': 'superoxide dismutase 1',\n",
       " 'KEAP1': 'kelch like ECH associated protein 1'}"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_names=kegg_symbols_and_names('https://www.kegg.jp/entry/pathway+hsa05012')\n",
    "kegg_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>DB_Object_Synonym</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>LRRK2</td>\n",
       "      <td>Leucine-rich repeat serine/threonine-protein k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>5</td>\n",
       "      <td>PARK8</td>\n",
       "      <td>Leucine-rich repeat serine/threonine-protein k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1024</th>\n",
       "      <td>11</td>\n",
       "      <td>UBA1</td>\n",
       "      <td>Ubiquitin-like modifier-activating enzyme 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1075</th>\n",
       "      <td>11</td>\n",
       "      <td>A1S9T</td>\n",
       "      <td>Ubiquitin-like modifier-activating enzyme 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1126</th>\n",
       "      <td>11</td>\n",
       "      <td>UBE1</td>\n",
       "      <td>Ubiquitin-like modifier-activating enzyme 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60546</th>\n",
       "      <td>724</td>\n",
       "      <td>SLC39A11</td>\n",
       "      <td>Zinc transporter ZIP11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60557</th>\n",
       "      <td>724</td>\n",
       "      <td>C17orf26</td>\n",
       "      <td>Zinc transporter ZIP11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60568</th>\n",
       "      <td>724</td>\n",
       "      <td>ZIP11</td>\n",
       "      <td>Zinc transporter ZIP11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60579</th>\n",
       "      <td>735</td>\n",
       "      <td>SOD1</td>\n",
       "      <td>Superoxide dismutase [Cu-Zn]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60796</th>\n",
       "      <td>735</td>\n",
       "      <td>ALS</td>\n",
       "      <td>Insulin-like growth factor-binding protein com...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>278 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id DB_Object_Synonym  \\\n",
       "0           5             LRRK2   \n",
       "512         5             PARK8   \n",
       "1024       11              UBA1   \n",
       "1075       11             A1S9T   \n",
       "1126       11              UBE1   \n",
       "...       ...               ...   \n",
       "60546     724          SLC39A11   \n",
       "60557     724          C17orf26   \n",
       "60568     724             ZIP11   \n",
       "60579     735              SOD1   \n",
       "60796     735               ALS   \n",
       "\n",
       "                                          DB_Object_Name  \n",
       "0      Leucine-rich repeat serine/threonine-protein k...  \n",
       "512    Leucine-rich repeat serine/threonine-protein k...  \n",
       "1024         Ubiquitin-like modifier-activating enzyme 1  \n",
       "1075         Ubiquitin-like modifier-activating enzyme 1  \n",
       "1126         Ubiquitin-like modifier-activating enzyme 1  \n",
       "...                                                  ...  \n",
       "60546                             Zinc transporter ZIP11  \n",
       "60557                             Zinc transporter ZIP11  \n",
       "60568                             Zinc transporter ZIP11  \n",
       "60579                       Superoxide dismutase [Cu-Zn]  \n",
       "60796  Insulin-like growth factor-binding protein com...  \n",
       "\n",
       "[278 rows x 3 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aux=merged[['gene_id','DB_Object_Synonym','DB_Object_Name']].drop_duplicates()\n",
    "aux"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>DB_Object_Synonym</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RPS27A</td>\n",
       "      <td>ribosomal protein S27a</td>\n",
       "      <td>12</td>\n",
       "      <td>RPS27A</td>\n",
       "      <td>Ubiquitin-ribosomal protein eS31 fusion protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>UBC</td>\n",
       "      <td>ubiquitin C</td>\n",
       "      <td>12</td>\n",
       "      <td>UBC</td>\n",
       "      <td>Polyubiquitin-C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene_Symbol               Gene_Name gene_id DB_Object_Synonym  \\\n",
       "13      RPS27A  ribosomal protein S27a      12            RPS27A   \n",
       "24         UBC             ubiquitin C      12               UBC   \n",
       "\n",
       "                                     DB_Object_Name  \n",
       "13  Ubiquitin-ribosomal protein eS31 fusion protein  \n",
       "24                                  Polyubiquitin-C  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#example of gene with multiple names and symbols both from KEGG and GO\n",
    "from_kegg=pd.DataFrame(list(kegg_names.items()), columns=['Gene_Symbol', 'Gene_Name'])\n",
    "kegg_go=pd.merge(from_kegg,aux,left_on='Gene_Symbol',right_on='DB_Object_Synonym').drop_duplicates()\n",
    "kegg_go[kegg_go['gene_id']=='12']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Qualifier</th>\n",
       "      <th>GO_ID</th>\n",
       "      <th>Aspect</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "      <th>DB_Object_Synonym</th>\n",
       "      <th>DB_Object_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0034260</td>\n",
       "      <td>P</td>\n",
       "      <td>Leucine-rich repeat serine/threonine-protein k...</td>\n",
       "      <td>LRRK2</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1024</th>\n",
       "      <td>11</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0000792</td>\n",
       "      <td>C</td>\n",
       "      <td>Ubiquitin-like modifier-activating enzyme 1</td>\n",
       "      <td>UBA1</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>12</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003723</td>\n",
       "      <td>F</td>\n",
       "      <td>Ubiquitin-ribosomal protein eS31 fusion protein</td>\n",
       "      <td>RPS27A</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2343</th>\n",
       "      <td>12</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003723</td>\n",
       "      <td>F</td>\n",
       "      <td>Polyubiquitin-C</td>\n",
       "      <td>UBC</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3563</th>\n",
       "      <td>13</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003713</td>\n",
       "      <td>F</td>\n",
       "      <td>Ubiquitin-conjugating enzyme E2 L3</td>\n",
       "      <td>UBE2L3</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60406</th>\n",
       "      <td>715</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>F</td>\n",
       "      <td>Histone-lysine N-methyltransferase 2B</td>\n",
       "      <td>TRX2</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60453</th>\n",
       "      <td>715</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003723</td>\n",
       "      <td>F</td>\n",
       "      <td>Thioredoxin</td>\n",
       "      <td>TXN</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60546</th>\n",
       "      <td>724</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005375</td>\n",
       "      <td>F</td>\n",
       "      <td>Zinc transporter ZIP11</td>\n",
       "      <td>SLC39A11</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60579</th>\n",
       "      <td>735</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0043410</td>\n",
       "      <td>P</td>\n",
       "      <td>Superoxide dismutase [Cu-Zn]</td>\n",
       "      <td>SOD1</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60796</th>\n",
       "      <td>735</td>\n",
       "      <td>involved_in</td>\n",
       "      <td>GO:0007155</td>\n",
       "      <td>P</td>\n",
       "      <td>Insulin-like growth factor-binding protein com...</td>\n",
       "      <td>ALS</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>155 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id                   Qualifier       GO_ID Aspect  \\\n",
       "0           5  acts_upstream_of_or_within  GO:0034260      P   \n",
       "1024       11                     part_of  GO:0000792      C   \n",
       "1177       12                     enables  GO:0003723      F   \n",
       "2343       12                     enables  GO:0003723      F   \n",
       "3563       13                     enables  GO:0003713      F   \n",
       "...       ...                         ...         ...    ...   \n",
       "60406     715                     enables  GO:0005515      F   \n",
       "60453     715                     enables  GO:0003723      F   \n",
       "60546     724                     enables  GO:0005375      F   \n",
       "60579     735  acts_upstream_of_or_within  GO:0043410      P   \n",
       "60796     735                 involved_in  GO:0007155      P   \n",
       "\n",
       "                                          DB_Object_Name DB_Object_Synonym  \\\n",
       "0      Leucine-rich repeat serine/threonine-protein k...             LRRK2   \n",
       "1024         Ubiquitin-like modifier-activating enzyme 1              UBA1   \n",
       "1177     Ubiquitin-ribosomal protein eS31 fusion protein            RPS27A   \n",
       "2343                                     Polyubiquitin-C               UBC   \n",
       "3563                  Ubiquitin-conjugating enzyme E2 L3            UBE2L3   \n",
       "...                                                  ...               ...   \n",
       "60406              Histone-lysine N-methyltransferase 2B              TRX2   \n",
       "60453                                        Thioredoxin               TXN   \n",
       "60546                             Zinc transporter ZIP11          SLC39A11   \n",
       "60579                       Superoxide dismutase [Cu-Zn]              SOD1   \n",
       "60796  Insulin-like growth factor-binding protein com...               ALS   \n",
       "\n",
       "      DB_Object_Type  \n",
       "0            protein  \n",
       "1024         protein  \n",
       "1177         protein  \n",
       "2343         protein  \n",
       "3563         protein  \n",
       "...              ...  \n",
       "60406        protein  \n",
       "60453        protein  \n",
       "60546        protein  \n",
       "60579        protein  \n",
       "60796        protein  \n",
       "\n",
       "[155 rows x 7 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#example from GO\n",
    "ids=merged[['gene_id','DB_Object_Name']].drop_duplicates()['DB_Object_Name'].index\n",
    "merged.loc[ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>Leucine-rich repeat serine/threonine-protein k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1024</th>\n",
       "      <td>11</td>\n",
       "      <td>Ubiquitin-like modifier-activating enzyme 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>12</td>\n",
       "      <td>Ubiquitin-ribosomal protein eS31 fusion protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2343</th>\n",
       "      <td>12</td>\n",
       "      <td>Polyubiquitin-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3563</th>\n",
       "      <td>13</td>\n",
       "      <td>Ubiquitin-conjugating enzyme E2 L3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60406</th>\n",
       "      <td>715</td>\n",
       "      <td>Histone-lysine N-methyltransferase 2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60453</th>\n",
       "      <td>715</td>\n",
       "      <td>Thioredoxin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60546</th>\n",
       "      <td>724</td>\n",
       "      <td>Zinc transporter ZIP11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60579</th>\n",
       "      <td>735</td>\n",
       "      <td>Superoxide dismutase [Cu-Zn]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60796</th>\n",
       "      <td>735</td>\n",
       "      <td>Insulin-like growth factor-binding protein com...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>155 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id                                     DB_Object_Name\n",
       "0           5  Leucine-rich repeat serine/threonine-protein k...\n",
       "1024       11        Ubiquitin-like modifier-activating enzyme 1\n",
       "1177       12    Ubiquitin-ribosomal protein eS31 fusion protein\n",
       "2343       12                                    Polyubiquitin-C\n",
       "3563       13                 Ubiquitin-conjugating enzyme E2 L3\n",
       "...       ...                                                ...\n",
       "60406     715              Histone-lysine N-methyltransferase 2B\n",
       "60453     715                                        Thioredoxin\n",
       "60546     724                             Zinc transporter ZIP11\n",
       "60579     735                       Superoxide dismutase [Cu-Zn]\n",
       "60796     735  Insulin-like growth factor-binding protein com...\n",
       "\n",
       "[155 rows x 2 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#gene_ids map to different Names \n",
    "merged[['gene_id','DB_Object_Name']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "      <th>DB_Object_Synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>[Ubiquitin-like modifier-activating enzyme 1]</td>\n",
       "      <td>[UBA1, UBE1, A1S9T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>[Polyubiquitin-C, Ubiquitin-ribosomal protein ...</td>\n",
       "      <td>[UBCEP1, UBA80, UBC, RPS27A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>[Ubiquitin-conjugating enzyme E2 L3]</td>\n",
       "      <td>[UBCH7, UBE2L3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>139</td>\n",
       "      <td>[Ubiquitin-conjugating enzyme E2 J2]</td>\n",
       "      <td>[NCUBE2, UBE2J2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14</td>\n",
       "      <td>[Polyubiquitin-C, Ubiquitin-ribosomal protein ...</td>\n",
       "      <td>[UBCEP1, UBA80, UBC, RPS27A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>711</td>\n",
       "      <td>[Sodium-dependent dopamine transporter]</td>\n",
       "      <td>[DAT1, SLC6A3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>714</td>\n",
       "      <td>[Kelch-like ECH-associated protein 1]</td>\n",
       "      <td>[KLHL19, KEAP1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>715</td>\n",
       "      <td>[Thioredoxin, Thioredoxin, mitochondrial, Hist...</td>\n",
       "      <td>[TRX2, TXN, TXN2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>724</td>\n",
       "      <td>[Zinc transporter ZIP11]</td>\n",
       "      <td>[ZIP11, C17orf26, SLC39A11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>735</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_id                                     DB_Object_Name  \\\n",
       "0        11      [Ubiquitin-like modifier-activating enzyme 1]   \n",
       "1        12  [Polyubiquitin-C, Ubiquitin-ribosomal protein ...   \n",
       "2        13               [Ubiquitin-conjugating enzyme E2 L3]   \n",
       "3       139               [Ubiquitin-conjugating enzyme E2 J2]   \n",
       "4        14  [Polyubiquitin-C, Ubiquitin-ribosomal protein ...   \n",
       "..      ...                                                ...   \n",
       "98      711            [Sodium-dependent dopamine transporter]   \n",
       "99      714              [Kelch-like ECH-associated protein 1]   \n",
       "100     715  [Thioredoxin, Thioredoxin, mitochondrial, Hist...   \n",
       "101     724                           [Zinc transporter ZIP11]   \n",
       "102     735  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "\n",
       "                DB_Object_Synonym  \n",
       "0             [UBA1, UBE1, A1S9T]  \n",
       "1    [UBCEP1, UBA80, UBC, RPS27A]  \n",
       "2                 [UBCH7, UBE2L3]  \n",
       "3                [NCUBE2, UBE2J2]  \n",
       "4    [UBCEP1, UBA80, UBC, RPS27A]  \n",
       "..                            ...  \n",
       "98                 [DAT1, SLC6A3]  \n",
       "99                [KLHL19, KEAP1]  \n",
       "100             [TRX2, TXN, TXN2]  \n",
       "101   [ZIP11, C17orf26, SLC39A11]  \n",
       "102                   [SOD1, ALS]  \n",
       "\n",
       "[103 rows x 3 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#let's aggregate all synonyms and names by gene_id\n",
    "agg_names_synonyms=merged[['gene_id','DB_Object_Name','DB_Object_Synonym']].groupby('gene_id').agg(lambda x:list(set(x))).reset_index()\n",
    "agg_names_synonyms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Qualifier</th>\n",
       "      <th>GO_ID</th>\n",
       "      <th>Aspect</th>\n",
       "      <th>DB_Object_Type</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "      <th>DB_Object_Synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0034260</td>\n",
       "      <td>P</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0000149</td>\n",
       "      <td>F</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0000287</td>\n",
       "      <td>F</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003779</td>\n",
       "      <td>F</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>F</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14369</th>\n",
       "      <td>735</td>\n",
       "      <td>located_in</td>\n",
       "      <td>GO:0070062</td>\n",
       "      <td>C</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14370</th>\n",
       "      <td>735</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0042567</td>\n",
       "      <td>C</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14371</th>\n",
       "      <td>735</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005520</td>\n",
       "      <td>F</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14372</th>\n",
       "      <td>735</td>\n",
       "      <td>is_active_in</td>\n",
       "      <td>GO:0031012</td>\n",
       "      <td>C</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14373</th>\n",
       "      <td>735</td>\n",
       "      <td>is_active_in</td>\n",
       "      <td>GO:0005615</td>\n",
       "      <td>C</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14374 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id                   Qualifier       GO_ID Aspect DB_Object_Type  \\\n",
       "0           5  acts_upstream_of_or_within  GO:0034260      P        protein   \n",
       "1           5                     enables  GO:0000149      F        protein   \n",
       "2           5                     enables  GO:0000287      F        protein   \n",
       "3           5                     enables  GO:0003779      F        protein   \n",
       "4           5                     enables  GO:0003924      F        protein   \n",
       "...       ...                         ...         ...    ...            ...   \n",
       "14369     735                  located_in  GO:0070062      C        protein   \n",
       "14370     735                     part_of  GO:0042567      C        protein   \n",
       "14371     735                     enables  GO:0005520      F        protein   \n",
       "14372     735                is_active_in  GO:0031012      C        protein   \n",
       "14373     735                is_active_in  GO:0005615      C        protein   \n",
       "\n",
       "                                          DB_Object_Name DB_Object_Synonym  \n",
       "0      [Leucine-rich repeat serine/threonine-protein ...    [LRRK2, PARK8]  \n",
       "1      [Leucine-rich repeat serine/threonine-protein ...    [LRRK2, PARK8]  \n",
       "2      [Leucine-rich repeat serine/threonine-protein ...    [LRRK2, PARK8]  \n",
       "3      [Leucine-rich repeat serine/threonine-protein ...    [LRRK2, PARK8]  \n",
       "4      [Leucine-rich repeat serine/threonine-protein ...    [LRRK2, PARK8]  \n",
       "...                                                  ...               ...  \n",
       "14369  [Superoxide dismutase [Cu-Zn], Insulin-like gr...       [SOD1, ALS]  \n",
       "14370  [Superoxide dismutase [Cu-Zn], Insulin-like gr...       [SOD1, ALS]  \n",
       "14371  [Superoxide dismutase [Cu-Zn], Insulin-like gr...       [SOD1, ALS]  \n",
       "14372  [Superoxide dismutase [Cu-Zn], Insulin-like gr...       [SOD1, ALS]  \n",
       "14373  [Superoxide dismutase [Cu-Zn], Insulin-like gr...       [SOD1, ALS]  \n",
       "\n",
       "[14374 rows x 7 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge merged with agg_names_synonym so that all gene_id entries have consistent names and symbols\n",
    "merged=pd.merge(merged[['gene_id','Qualifier','GO_ID','Aspect','DB_Object_Type']],agg_names_synonyms)\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_cores = os.cpu_count()\n",
    "num_cores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [],
   "source": [
    "def go_id_description(df:pd.DataFrame)->pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Process a DataFrame to get GO term descriptions.\n",
    "    \n",
    "    Args:\n",
    "        df (pd.DataFrame): DataFrame with at least one column 'GO_ID'.\n",
    "    \n",
    "    Returns:\n",
    "        pd.DataFrame: DataFrame with 'GO_ID', 'GO_label', and 'GO_definition'.\n",
    "    \"\"\"\n",
    "    if not pd.api.types.is_string_dtype(df['GO_ID']):\n",
    "        raise ValueError(\"'GO_ID' column must be of type string.\")\n",
    "    \n",
    "    go_df=df[['GO_ID']].drop_duplicates()\n",
    "\n",
    "    # Convert the Pandas DataFrame to a Dask DataFrame\n",
    "    num_cores = os.cpu_count()\n",
    "    ddf = dd.from_pandas(go_df, npartitions=num_cores)\n",
    "\n",
    "    # Define a function to apply to each partition\n",
    "    def apply_go_term_description(partition):\n",
    "        results = [get_go_term_description(go_id) for go_id in partition['GO_ID']]\n",
    "        result_df = pd.DataFrame(results)\n",
    "        return pd.concat([partition.reset_index(drop=True), result_df], axis=1)\n",
    "    \n",
    "    # define metadata structure for Dask computation\n",
    "    meta = pd.DataFrame({\n",
    "        'GO_ID': pd.Series(dtype='str'),\n",
    "        'GO_label': pd.Series(dtype='str'),\n",
    "        'GO_definition': pd.Series(dtype='str')\n",
    "    })\n",
    "\n",
    "    go_ddf = ddf.map_partitions(apply_go_term_description, meta=meta)\n",
    "    \n",
    "    return go_ddf.compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_ID</th>\n",
       "      <th>GO_label</th>\n",
       "      <th>GO_definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0034260</td>\n",
       "      <td>negative regulation of GTPase activity</td>\n",
       "      <td>Any process that stops or reduces the rate of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000149</td>\n",
       "      <td>SNARE binding</td>\n",
       "      <td>Binding to a SNARE (soluble N-ethylmaleimide-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000287</td>\n",
       "      <td>magnesium ion binding</td>\n",
       "      <td>Binding to a magnesium (Mg) ion.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0003779</td>\n",
       "      <td>actin binding</td>\n",
       "      <td>Binding to monomeric or multimeric forms of ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GTPase activity</td>\n",
       "      <td>Catalysis of the reaction: GTP + H2O = GDP + H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>GO:0099610</td>\n",
       "      <td>action potential initiation</td>\n",
       "      <td>The initiating cycle of an action potential. I...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>GO:0005777</td>\n",
       "      <td>peroxisome</td>\n",
       "      <td>A small organelle enclosed by a single membran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>GO:0042567</td>\n",
       "      <td>insulin-like growth factor ternary complex</td>\n",
       "      <td>A complex of three proteins, which in animals ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>GO:0005520</td>\n",
       "      <td>insulin-like growth factor binding</td>\n",
       "      <td>Binding to an insulin-like growth factor, any ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>GO:0031012</td>\n",
       "      <td>extracellular matrix</td>\n",
       "      <td>A structure lying external to one or more cell...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1824 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GO_ID                                    GO_label  \\\n",
       "0    GO:0034260      negative regulation of GTPase activity   \n",
       "1    GO:0000149                               SNARE binding   \n",
       "2    GO:0000287                       magnesium ion binding   \n",
       "3    GO:0003779                               actin binding   \n",
       "4    GO:0003924                             GTPase activity   \n",
       "..          ...                                         ...   \n",
       "223  GO:0099610                 action potential initiation   \n",
       "224  GO:0005777                                  peroxisome   \n",
       "225  GO:0042567  insulin-like growth factor ternary complex   \n",
       "226  GO:0005520          insulin-like growth factor binding   \n",
       "227  GO:0031012                        extracellular matrix   \n",
       "\n",
       "                                         GO_definition  \n",
       "0    Any process that stops or reduces the rate of ...  \n",
       "1    Binding to a SNARE (soluble N-ethylmaleimide-s...  \n",
       "2                     Binding to a magnesium (Mg) ion.  \n",
       "3    Binding to monomeric or multimeric forms of ac...  \n",
       "4    Catalysis of the reaction: GTP + H2O = GDP + H...  \n",
       "..                                                 ...  \n",
       "223  The initiating cycle of an action potential. I...  \n",
       "224  A small organelle enclosed by a single membran...  \n",
       "225  A complex of three proteins, which in animals ...  \n",
       "226  Binding to an insulin-like growth factor, any ...  \n",
       "227  A structure lying external to one or more cell...  \n",
       "\n",
       "[1824 rows x 3 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_df=merged[['GO_ID']].drop_duplicates()\n",
    "\n",
    "# Convert the Pandas DataFrame to a Dask DataFrame\n",
    "ddf = dd.from_pandas(go_df, npartitions=num_cores)\n",
    "\n",
    "# Define a function to apply to each partition\n",
    "def apply_go_term_description(partition):\n",
    "    results = [get_go_term_description(go_id) for go_id in partition['GO_ID']]\n",
    "    result_df = pd.DataFrame(results)\n",
    "    return pd.concat([partition.reset_index(drop=True), result_df], axis=1)\n",
    "\n",
    "\n",
    "meta = pd.DataFrame({\n",
    "    'GO_ID': pd.Series(dtype='str'),\n",
    "    'GO_label': pd.Series(dtype='str'),\n",
    "    'GO_definition': pd.Series(dtype='str')\n",
    "})\n",
    "\n",
    "go_ddf = ddf.map_partitions(apply_go_term_description, meta=meta)\n",
    "\n",
    "# Compute the result\n",
    "go_df = go_ddf.compute()\n",
    "go_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "aspect_dict={\"P\":\"Biological Process\", \"F\":\"Molecular Function\", \"C\":\"Cellular Component\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Qualifier</th>\n",
       "      <th>GO_ID</th>\n",
       "      <th>Aspect</th>\n",
       "      <th>DB_Object_Type</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "      <th>DB_Object_Synonym</th>\n",
       "      <th>GO_label</th>\n",
       "      <th>GO_definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0034260</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "      <td>negative regulation of GTPase activity</td>\n",
       "      <td>Any process that stops or reduces the rate of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0000149</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "      <td>SNARE binding</td>\n",
       "      <td>Binding to a SNARE (soluble N-ethylmaleimide-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0000287</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "      <td>magnesium ion binding</td>\n",
       "      <td>Binding to a magnesium (Mg) ion.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003779</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "      <td>actin binding</td>\n",
       "      <td>Binding to monomeric or multimeric forms of ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "      <td>GTPase activity</td>\n",
       "      <td>Catalysis of the reaction: GTP + H2O = GDP + H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14369</th>\n",
       "      <td>735</td>\n",
       "      <td>located_in</td>\n",
       "      <td>GO:0070062</td>\n",
       "      <td>Cellular Component</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "      <td>extracellular exosome</td>\n",
       "      <td>A vesicle that is released into the extracellu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14370</th>\n",
       "      <td>735</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0042567</td>\n",
       "      <td>Cellular Component</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "      <td>insulin-like growth factor ternary complex</td>\n",
       "      <td>A complex of three proteins, which in animals ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14371</th>\n",
       "      <td>735</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005520</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "      <td>insulin-like growth factor binding</td>\n",
       "      <td>Binding to an insulin-like growth factor, any ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14372</th>\n",
       "      <td>735</td>\n",
       "      <td>is_active_in</td>\n",
       "      <td>GO:0031012</td>\n",
       "      <td>Cellular Component</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "      <td>extracellular matrix</td>\n",
       "      <td>A structure lying external to one or more cell...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14373</th>\n",
       "      <td>735</td>\n",
       "      <td>is_active_in</td>\n",
       "      <td>GO:0005615</td>\n",
       "      <td>Cellular Component</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "      <td>extracellular space</td>\n",
       "      <td>That part of a multicellular organism outside ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14374 rows  9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id                   Qualifier       GO_ID              Aspect  \\\n",
       "0           5  acts_upstream_of_or_within  GO:0034260  Biological Process   \n",
       "1           5                     enables  GO:0000149  Molecular Function   \n",
       "2           5                     enables  GO:0000287  Molecular Function   \n",
       "3           5                     enables  GO:0003779  Molecular Function   \n",
       "4           5                     enables  GO:0003924  Molecular Function   \n",
       "...       ...                         ...         ...                 ...   \n",
       "14369     735                  located_in  GO:0070062  Cellular Component   \n",
       "14370     735                     part_of  GO:0042567  Cellular Component   \n",
       "14371     735                     enables  GO:0005520  Molecular Function   \n",
       "14372     735                is_active_in  GO:0031012  Cellular Component   \n",
       "14373     735                is_active_in  GO:0005615  Cellular Component   \n",
       "\n",
       "      DB_Object_Type                                     DB_Object_Name  \\\n",
       "0            protein  [Leucine-rich repeat serine/threonine-protein ...   \n",
       "1            protein  [Leucine-rich repeat serine/threonine-protein ...   \n",
       "2            protein  [Leucine-rich repeat serine/threonine-protein ...   \n",
       "3            protein  [Leucine-rich repeat serine/threonine-protein ...   \n",
       "4            protein  [Leucine-rich repeat serine/threonine-protein ...   \n",
       "...              ...                                                ...   \n",
       "14369        protein  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "14370        protein  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "14371        protein  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "14372        protein  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "14373        protein  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "\n",
       "      DB_Object_Synonym                                    GO_label  \\\n",
       "0        [LRRK2, PARK8]      negative regulation of GTPase activity   \n",
       "1        [LRRK2, PARK8]                               SNARE binding   \n",
       "2        [LRRK2, PARK8]                       magnesium ion binding   \n",
       "3        [LRRK2, PARK8]                               actin binding   \n",
       "4        [LRRK2, PARK8]                             GTPase activity   \n",
       "...                 ...                                         ...   \n",
       "14369       [SOD1, ALS]                       extracellular exosome   \n",
       "14370       [SOD1, ALS]  insulin-like growth factor ternary complex   \n",
       "14371       [SOD1, ALS]          insulin-like growth factor binding   \n",
       "14372       [SOD1, ALS]                        extracellular matrix   \n",
       "14373       [SOD1, ALS]                         extracellular space   \n",
       "\n",
       "                                           GO_definition  \n",
       "0      Any process that stops or reduces the rate of ...  \n",
       "1      Binding to a SNARE (soluble N-ethylmaleimide-s...  \n",
       "2                       Binding to a magnesium (Mg) ion.  \n",
       "3      Binding to monomeric or multimeric forms of ac...  \n",
       "4      Catalysis of the reaction: GTP + H2O = GDP + H...  \n",
       "...                                                  ...  \n",
       "14369  A vesicle that is released into the extracellu...  \n",
       "14370  A complex of three proteins, which in animals ...  \n",
       "14371  Binding to an insulin-like growth factor, any ...  \n",
       "14372  A structure lying external to one or more cell...  \n",
       "14373  That part of a multicellular organism outside ...  \n",
       "\n",
       "[14374 rows x 9 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged=pd.merge(merged,go_df,on='GO_ID')\n",
    "merged=merged.replace({'Aspect':aspect_dict})\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {},
   "outputs": [],
   "source": [
    "#LOAD config.yaml\n",
    "def kegg_go_integration(kgml_path: str, gaf_path: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Integrates KEGG and GO data.\n",
    "\n",
    "    Args:\n",
    "        kgml_path (str): Path to the KEGG KGML file.\n",
    "        gaf_path (str): Path to the GO GAF file.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Merged DataFrame with integrated KEGG and GO data.\n",
    "    \"\"\"\n",
    "    # Parse KEGG data\n",
    "    kegg_genes=parse_kgml(kgml_path)['genes']\n",
    "    kegg_df=pd.DataFrame(kegg_genes,columns=['gene_id','gene_symbols'])\n",
    "    kegg_exploded_df=kegg_df.explode('gene_symbols')\n",
    "\n",
    "    # Parse GAF data\n",
    "    gaf_df=parse_gaf(gaf_path)[['Qualifier','GO_ID','Aspect','DB_Object_Name','DB_Object_Synonym','DB_Object_Type']]\n",
    "    gaf_exploded_df=gaf_df.explode('DB_Object_Synonym')\n",
    "\n",
    "    # Merge KEGG and GAF on common synonyms\n",
    "    merged_df=pd.merge(kegg_exploded_df,gaf_exploded_df,left_on='gene_symbols',right_on='DB_Object_Synonym',how='inner').drop_duplicates()\n",
    "\n",
    "    # Aggregate all synonyms and names by gene_id\n",
    "    agg_names_synonyms_df=merged_df[['gene_id','DB_Object_Name','DB_Object_Synonym']].groupby('gene_id').agg(lambda x:list(set(x))).reset_index()\n",
    "\n",
    "    # Merge merged_df with agg_names_synonyms_df so that all gene_id entries have consistent names and symbols\n",
    "    merged_df=pd.merge(merged_df[['gene_id','Qualifier','GO_ID','Aspect','DB_Object_Type']],agg_names_synonyms_df)\n",
    "\n",
    "    #Get dataframe with GO_IDs and their descriptions\n",
    "    go_df=go_id_description(merged_df)\n",
    "    \n",
    "    # Merge GO descriptions and replace aspect values\n",
    "    merged_df=pd.merge(merged_df,go_df,on='GO_ID').replace({'Aspect':aspect_dict})\n",
    "\n",
    "    return merged_df\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Qualifier</th>\n",
       "      <th>GO_ID</th>\n",
       "      <th>Aspect</th>\n",
       "      <th>DB_Object_Type</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "      <th>DB_Object_Synonym</th>\n",
       "      <th>GO_label</th>\n",
       "      <th>GO_definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0034260</td>\n",
       "      <td>Biological Process</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "      <td>negative regulation of GTPase activity</td>\n",
       "      <td>Any process that stops or reduces the rate of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0000149</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "      <td>SNARE binding</td>\n",
       "      <td>Binding to a SNARE (soluble N-ethylmaleimide-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0000287</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "      <td>magnesium ion binding</td>\n",
       "      <td>Binding to a magnesium (Mg) ion.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003779</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "      <td>actin binding</td>\n",
       "      <td>Binding to monomeric or multimeric forms of ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Leucine-rich repeat serine/threonine-protein ...</td>\n",
       "      <td>[LRRK2, PARK8]</td>\n",
       "      <td>GTPase activity</td>\n",
       "      <td>Catalysis of the reaction: GTP + H2O = GDP + H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14369</th>\n",
       "      <td>735</td>\n",
       "      <td>located_in</td>\n",
       "      <td>GO:0070062</td>\n",
       "      <td>Cellular Component</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "      <td>extracellular exosome</td>\n",
       "      <td>A vesicle that is released into the extracellu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14370</th>\n",
       "      <td>735</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0042567</td>\n",
       "      <td>Cellular Component</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "      <td>insulin-like growth factor ternary complex</td>\n",
       "      <td>A complex of three proteins, which in animals ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14371</th>\n",
       "      <td>735</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005520</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "      <td>insulin-like growth factor binding</td>\n",
       "      <td>Binding to an insulin-like growth factor, any ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14372</th>\n",
       "      <td>735</td>\n",
       "      <td>is_active_in</td>\n",
       "      <td>GO:0031012</td>\n",
       "      <td>Cellular Component</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "      <td>extracellular matrix</td>\n",
       "      <td>A structure lying external to one or more cell...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14373</th>\n",
       "      <td>735</td>\n",
       "      <td>is_active_in</td>\n",
       "      <td>GO:0005615</td>\n",
       "      <td>Cellular Component</td>\n",
       "      <td>protein</td>\n",
       "      <td>[Superoxide dismutase [Cu-Zn], Insulin-like gr...</td>\n",
       "      <td>[SOD1, ALS]</td>\n",
       "      <td>extracellular space</td>\n",
       "      <td>That part of a multicellular organism outside ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14374 rows  9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id                   Qualifier       GO_ID              Aspect  \\\n",
       "0           5  acts_upstream_of_or_within  GO:0034260  Biological Process   \n",
       "1           5                     enables  GO:0000149  Molecular Function   \n",
       "2           5                     enables  GO:0000287  Molecular Function   \n",
       "3           5                     enables  GO:0003779  Molecular Function   \n",
       "4           5                     enables  GO:0003924  Molecular Function   \n",
       "...       ...                         ...         ...                 ...   \n",
       "14369     735                  located_in  GO:0070062  Cellular Component   \n",
       "14370     735                     part_of  GO:0042567  Cellular Component   \n",
       "14371     735                     enables  GO:0005520  Molecular Function   \n",
       "14372     735                is_active_in  GO:0031012  Cellular Component   \n",
       "14373     735                is_active_in  GO:0005615  Cellular Component   \n",
       "\n",
       "      DB_Object_Type                                     DB_Object_Name  \\\n",
       "0            protein  [Leucine-rich repeat serine/threonine-protein ...   \n",
       "1            protein  [Leucine-rich repeat serine/threonine-protein ...   \n",
       "2            protein  [Leucine-rich repeat serine/threonine-protein ...   \n",
       "3            protein  [Leucine-rich repeat serine/threonine-protein ...   \n",
       "4            protein  [Leucine-rich repeat serine/threonine-protein ...   \n",
       "...              ...                                                ...   \n",
       "14369        protein  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "14370        protein  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "14371        protein  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "14372        protein  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "14373        protein  [Superoxide dismutase [Cu-Zn], Insulin-like gr...   \n",
       "\n",
       "      DB_Object_Synonym                                    GO_label  \\\n",
       "0        [LRRK2, PARK8]      negative regulation of GTPase activity   \n",
       "1        [LRRK2, PARK8]                               SNARE binding   \n",
       "2        [LRRK2, PARK8]                       magnesium ion binding   \n",
       "3        [LRRK2, PARK8]                               actin binding   \n",
       "4        [LRRK2, PARK8]                             GTPase activity   \n",
       "...                 ...                                         ...   \n",
       "14369       [SOD1, ALS]                       extracellular exosome   \n",
       "14370       [SOD1, ALS]  insulin-like growth factor ternary complex   \n",
       "14371       [SOD1, ALS]          insulin-like growth factor binding   \n",
       "14372       [SOD1, ALS]                        extracellular matrix   \n",
       "14373       [SOD1, ALS]                         extracellular space   \n",
       "\n",
       "                                           GO_definition  \n",
       "0      Any process that stops or reduces the rate of ...  \n",
       "1      Binding to a SNARE (soluble N-ethylmaleimide-s...  \n",
       "2                       Binding to a magnesium (Mg) ion.  \n",
       "3      Binding to monomeric or multimeric forms of ac...  \n",
       "4      Catalysis of the reaction: GTP + H2O = GDP + H...  \n",
       "...                                                  ...  \n",
       "14369  A vesicle that is released into the extracellu...  \n",
       "14370  A complex of three proteins, which in animals ...  \n",
       "14371  Binding to an insulin-like growth factor, any ...  \n",
       "14372  A structure lying external to one or more cell...  \n",
       "14373  That part of a multicellular organism outside ...  \n",
       "\n",
       "[14374 rows x 9 columns]"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged2=kegg_go_integration(kgml_path,gaf_path)\n",
    "merged2.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Names in KEGG and GO don't match (examples in 4930)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'phosphoinositide-3-kinase regulatory subunit 1' in set(gaf_df['DB_Object_Name'].apply(lambda x:x.lower()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'protein kinase C zeta'.lower() in set(gaf_df['DB_Object_Name'].apply(lambda x:x.lower()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23361, 42944, 19583)"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "symbols=set(gaf_new['DB_Object_Symbol'])\n",
    "synonyms=set(gaf_new['DB_Object_Synonym'].explode())\n",
    "len(synonyms-symbols),len(synonyms),len(symbols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DB_Object_ID\n",
       "1               19640\n",
       "2                   8\n",
       "3                   1\n",
       "8                   1\n",
       "12                  1\n",
       "15                  1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=gaf_new[['DB_Object_ID','DB_Object_Name']]\n",
    "df.drop_duplicates().groupby('DB_Object_Name').count().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "uri = os.getenv(\"NEO4J_URI\")\n",
    "user = os.getenv(\"NEO4J_USER\")\n",
    "password = os.getenv(\"NEO4J_PASSWORD\")\n",
    "api_key = os.getenv(\"OPENAI_API_KEY\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/1597374991.py:92: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_disease_node, {\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/1597374991.py:99: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_gene_node, gene,disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/1597374991.py:100: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_go_node, gene)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/1597374991.py:101: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_disease_association, gene, disease_id, evidence=\"from KGML\")\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/1597374991.py:102: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_go_association, gene,disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/1597374991.py:105: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_gene_interaction, interaction,disease_id)\n"
     ]
    }
   ],
   "source": [
    "class KGMLGAFImporter:\n",
    "    def __init__(self, driver):\n",
    "        self.driver = driver\n",
    "\n",
    "    def close(self):\n",
    "        self.driver.close()\n",
    "\n",
    "    def create_gene_node(self, tx, gene,disease_id):\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MERGE (g:Gene {unique_id: $unique_id})\n",
    "            SET g.synonyms = $synonyms\n",
    "            SET g.names= $names\n",
    "            \"\"\",\n",
    "            unique_id=f\"{disease_id}_{gene['gene_id']}\",\n",
    "            names=gene['DB_Object_Name'],\n",
    "            synonyms=gene['DB_Object_Synonym'],\n",
    "            type='gene'\n",
    "        )\n",
    "\n",
    "    def create_disease_node(self, tx, disease):\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MERGE (d:Disease {disease_id: $disease_id}) \n",
    "            SET d.name = $name\n",
    "            \"\"\",\n",
    "            disease_id=disease['disease_id'], \n",
    "            name=disease['name']\n",
    "        )\n",
    "    \n",
    "    def create_go_node(self,tx,gene):\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MERGE (a:GO_Annotation {qualifier:$qualifier, GO_ID:$GO_ID})\n",
    "            SET a.aspect=$aspect,\n",
    "                a.object_type=$object_type,\n",
    "                a.name=$GO_label,\n",
    "                a.definition=$GO_definition \n",
    "            \"\"\",\n",
    "            qualifier=gene['Qualifier'],\n",
    "            GO_ID=gene['GO_ID'],\n",
    "            aspect=gene['Aspect'],\n",
    "            object_type=gene['DB_Object_Type'],\t\n",
    "            GO_label=gene['GO_label'],\n",
    "            GO_definition=gene['GO_definition']\n",
    "        )\n",
    "\n",
    "    def create_gene_interaction(self, tx, interaction,disease_id):\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MATCH (g1:Gene {unique_id: $entry_1})\n",
    "            MATCH (g2:Gene {unique_id: $entry_2})\n",
    "            MERGE (g1)-[r:INTERACTS_WITH {type: $interaction_type, subtypes: $subtypes}]->(g2)\n",
    "            \"\"\",\n",
    "            entry_1=f\"{disease_id}_{interaction['entry1']}\",\n",
    "            entry_2=f\"{disease_id}_{interaction['entry2']}\",\n",
    "            interaction_type=interaction['type'],\n",
    "            subtypes=interaction['subtypes']\n",
    "        )\n",
    "\n",
    "    def create_disease_association(self, tx, gene, disease_id, evidence):\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MATCH (g:Gene {unique_id: $unique_id})\n",
    "            MATCH (d:Disease {disease_id: $disease_id})\n",
    "            MERGE (g)-[r:ASSOCIATED_WITH {evidence: $evidence}]->(d)\n",
    "            \"\"\",\n",
    "            unique_id=f\"{disease_id}_{gene['gene_id']}\",\n",
    "            disease_id=disease_id,\n",
    "            evidence=evidence\n",
    "        )\n",
    "    \n",
    "    def create_go_association(self,tx,gene,disease_id):\n",
    "            tx.run(\n",
    "            \"\"\"\n",
    "            MATCH (g:Gene {unique_id: $unique_id})\n",
    "            MATCH (a:GO_Annotation {qualifier:$qualifier, GO_ID:$GO_ID})\n",
    "            MERGE (g)-[r:HAS_GO_ANNOTATION]->(a)\n",
    "            \"\"\",\n",
    "            unique_id=f\"{disease_id}_{gene['gene_id']}\",\n",
    "            qualifier=gene['Qualifier'],\n",
    "            GO_ID=gene['GO_ID']\n",
    "           )\n",
    "\n",
    "    def import_data(self, kgml_files):\n",
    "        with self.driver.session() as session:\n",
    "            for kgml_file_path in kgml_files:\n",
    "                pathway_data = parse_kgml(kgml_file_path)\n",
    "                disease_id = pathway_data['pathway_id']\n",
    "                disease_name = pathway_data['pathway_name']\n",
    "                # Create disease node\n",
    "                session.write_transaction(self.create_disease_node, {\n",
    "                    'disease_id': disease_id,\n",
    "                    'name': disease_name\n",
    "                })\n",
    "                merged=kegg_go_integration(kgml_file_path,gaf_path)\n",
    "                # Create gene nodes and their associations with the disease\n",
    "                for _,df in merged.groupby('gene_id'):\n",
    "                    for _, gene in df.iterrows():\n",
    "                        session.write_transaction(self.create_gene_node, gene,disease_id)\n",
    "                        session.write_transaction(self.create_go_node, gene)\n",
    "                        session.write_transaction(self.create_disease_association, gene, disease_id, evidence=\"from KGML\")\n",
    "                        session.write_transaction(self.create_go_association, gene,disease_id)\n",
    "                # Create interactions between genes\n",
    "                for interaction in pathway_data['interactions']:\n",
    "                    session.write_transaction(self.create_gene_interaction, interaction,disease_id)\n",
    "\n",
    "\n",
    "# Connect to the Neo4j database\n",
    "driver = GraphDatabase.driver(uri, auth=(user, password))\n",
    "\n",
    "def create_constraints(driver):\n",
    "    with driver.session() as session:\n",
    "        session.run(\"CREATE CONSTRAINT FOR (d:Disease) REQUIRE d.disease_id IS UNIQUE;\")\n",
    "\n",
    "kgml_files = ['data/KGML/hsa05012.xml']\n",
    "\n",
    "# Create constraints and import data\n",
    "importer = KGMLGAFImporter(driver)\n",
    "importer.import_data(kgml_files)\n",
    "importer.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:292: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_disease_node, {\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:303: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_gene_node, gene.to_dict(), disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:304: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_go_node, gene.to_dict())\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:305: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_disease_association, gene.to_dict(), disease_id, evidence=\"from KGML\")\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:306: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_go_association, gene.to_dict(), disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:310: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_gene_interaction, interaction, disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:292: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_disease_node, {\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:303: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_gene_node, gene.to_dict(), disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:304: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_go_node, gene.to_dict())\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:305: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_disease_association, gene.to_dict(), disease_id, evidence=\"from KGML\")\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:306: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_go_association, gene.to_dict(), disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:310: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_gene_interaction, interaction, disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:292: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_disease_node, {\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:303: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_gene_node, gene.to_dict(), disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:304: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_go_node, gene.to_dict())\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:305: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_disease_association, gene.to_dict(), disease_id, evidence=\"from KGML\")\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:306: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_go_association, gene.to_dict(), disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:310: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_gene_interaction, interaction, disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:292: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_disease_node, {\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:303: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_gene_node, gene.to_dict(), disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:304: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_go_node, gene.to_dict())\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:305: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_disease_association, gene.to_dict(), disease_id, evidence=\"from KGML\")\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:306: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_go_association, gene.to_dict(), disease_id)\n",
      "/var/folders/x6/p2mqc4q53sz1h1z68cz92v840000gn/T/ipykernel_2673/2056716640.py:310: DeprecationWarning: write_transaction has been renamed to execute_write\n",
      "  session.write_transaction(self.create_gene_interaction, interaction, disease_id)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from neo4j import GraphDatabase\n",
    "from utilities.preprocessing import parse_kgml\n",
    "import pandas as pd\n",
    "from pydantic import BaseModel, Field, ValidationError, field_validator\n",
    "from typing import List, Optional, Dict, Any\n",
    "from neo4j import GraphDatabase,Transaction\n",
    "\n",
    "\n",
    "class GeneData(BaseModel):\n",
    "    \"\"\"\n",
    "    Pydantic model for Gene data with fields validated and default values handled.\n",
    "\n",
    "    Attributes:\n",
    "        gene_id: Unique identifier for the gene.\n",
    "        Qualifier: Qualifier for the gene annotation.\n",
    "        GO_ID: Gene Ontology ID.\n",
    "        Aspect: Aspect of the gene ontology.\n",
    "        DB_Object_Type: Type of the database object.\n",
    "        DB_Object_Name: List of names for the database object.\n",
    "        DB_Object_Synonym: List of synonyms for the database object.\n",
    "        GO_label: Label for the Gene Ontology term.\n",
    "        GO_definition: Definition for the Gene Ontology term.\n",
    "    \"\"\"\n",
    "    gene_id: str\n",
    "    Qualifier: Optional[str]\n",
    "    GO_ID: Optional[str]\n",
    "    Aspect: Optional[str]\n",
    "    DB_Object_Type: Optional[str]\n",
    "    DB_Object_Name: Optional[List[str]] = Field(default_factory=list)\n",
    "    DB_Object_Synonym: Optional[List[str]] = Field(default_factory=list)\n",
    "    GO_label: Optional[str]\n",
    "    GO_definition: Optional[str]\n",
    "\n",
    "    @field_validator('*', mode='before')\n",
    "    def handle_nan(cls, v: Any) -> Any:\n",
    "        \"\"\"\n",
    "        Handle NaN values by converting them to None.\n",
    "        If the value is a list, it iterates over the elements and replaces NaN with None.\n",
    "        \n",
    "        Args:\n",
    "            v: The value to check and potentially modify.\n",
    "        \n",
    "        Returns:\n",
    "            The modified value with NaN values handled.\n",
    "        \"\"\"\n",
    "        if isinstance(v, list):\n",
    "            return [None if pd.isna(i) else i for i in v]\n",
    "        if pd.isna(v):\n",
    "            return None\n",
    "        return v\n",
    "\n",
    "    @field_validator('DB_Object_Name', 'DB_Object_Synonym', mode='before')\n",
    "    def convert_to_list(cls, v:Optional[Any]) -> List[Any]:\n",
    "        \"\"\"\n",
    "        Convert None values to an empty list for specific fields.\n",
    "        \n",
    "        Args:\n",
    "            v: The value to check and potentially modify.\n",
    "        \n",
    "        Returns:\n",
    "            An empty list if the value is None, otherwise the original value.\n",
    "        \"\"\"\n",
    "        if v is None:\n",
    "            return []\n",
    "        return v\n",
    "\n",
    "\n",
    "class Disease(BaseModel):\n",
    "    \"\"\"\n",
    "    Pydantic model for Disease data with required fields.\n",
    "\n",
    "    Attributes:\n",
    "        disease_id: Unique identifier for the disease.\n",
    "        name: Name of the disease.\n",
    "    \"\"\"\n",
    "    disease_id: str\n",
    "    name: str\n",
    "\n",
    "\n",
    "class GeneInteraction(BaseModel):\n",
    "    \"\"\"\n",
    "    Pydantic model for Gene Interaction data with fields validated and default values handled.\n",
    "\n",
    "    Attributes:\n",
    "        entry1: Identifier for the first gene in the interaction.\n",
    "        entry2: Identifier for the second gene in the interaction.\n",
    "        type: Type of interaction.\n",
    "        subtypes: List of subtypes for the interaction.\n",
    "    \"\"\"\n",
    "    entry1: str\n",
    "    entry2: str\n",
    "    type: str\n",
    "    subtypes: Optional[List[str]] = Field(default_factory=list)\n",
    "\n",
    "\n",
    "class KGMLGAFImporter:\n",
    "    \"\"\"\n",
    "    Class to import data from KGML and GAF files into a Neo4j database.\n",
    "\n",
    "    Attributes:\n",
    "        driver: Neo4j driver for database connection.\n",
    "    \"\"\"\n",
    "    def __init__(self, driver: GraphDatabase.driver):\n",
    "        \"\"\"\n",
    "        Initialize the importer with a Neo4j driver.\n",
    "\n",
    "        Args:\n",
    "            driver: Neo4j driver for database connection.\n",
    "        \"\"\"\n",
    "        self.driver = driver\n",
    "\n",
    "    def close(self):\n",
    "        \"\"\"\n",
    "        Close the Neo4j driver connection.\n",
    "        \"\"\"\n",
    "        self.driver.close()\n",
    "\n",
    "    def create_gene_node(self, tx: Transaction, gene: Dict[str, Any], disease_id: str):\n",
    "        \"\"\"\n",
    "        Create a Gene node in the Neo4j database.\n",
    "\n",
    "        Args:\n",
    "            tx: Neo4j transaction object.\n",
    "            gene: Dictionary containing gene data.\n",
    "            disease_id: ID of the associated disease.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            gene_model = GeneData(**gene)\n",
    "        except ValidationError as e:\n",
    "            print(f\"Validation error for gene {gene['gene_id']}: {e}\")\n",
    "            return\n",
    "\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MERGE (g:Gene {unique_id: $unique_id})\n",
    "            SET g.synonyms = $synonyms,\n",
    "                g.names = $names\n",
    "            \"\"\",\n",
    "            unique_id=f\"{disease_id}_{gene_model.gene_id}\",\n",
    "            names=gene_model.DB_Object_Name,\n",
    "            synonyms=gene_model.DB_Object_Synonym,\n",
    "        )\n",
    "\n",
    "    def create_disease_node(self, tx: Transaction, disease: Dict[str, Any]):\n",
    "        \"\"\"\n",
    "        Create a Disease node in the Neo4j database.\n",
    "\n",
    "        Args:\n",
    "            tx: Neo4j transaction object.\n",
    "            disease: Dictionary containing disease data.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            disease_model = Disease(**disease)\n",
    "        except ValidationError as e:\n",
    "            print(f\"Validation error for disease {disease['disease_id']}: {e}\")\n",
    "            return\n",
    "\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MERGE (d:Disease {disease_id: $disease_id})\n",
    "            SET d.name = $name\n",
    "            \"\"\",\n",
    "            disease_id=disease_model.disease_id,\n",
    "            name=disease_model.name\n",
    "        )\n",
    "\n",
    "    def create_go_node(self, tx: Transaction, gene: Dict[str, Any]):\n",
    "        \"\"\"\n",
    "        Create a GO Annotation node in the Neo4j database.\n",
    "\n",
    "        Args:\n",
    "            tx: Neo4j transaction object.\n",
    "            gene: Dictionary containing gene data.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            gene_model = GeneData(**gene)\n",
    "        except ValidationError as e:\n",
    "            print(f\"Validation error for gene {gene['gene_id']}: {e}\")\n",
    "            return\n",
    "\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MERGE (a:GO_Annotation {qualifier: $qualifier, GO_ID: $GO_ID})\n",
    "            SET a.aspect = $aspect,\n",
    "                a.object_type = $object_type,\n",
    "                a.name = $GO_label,\n",
    "                a.definition = $GO_definition\n",
    "            \"\"\",\n",
    "            qualifier=gene_model.Qualifier,\n",
    "            GO_ID=gene_model.GO_ID,\n",
    "            aspect=gene_model.Aspect,\n",
    "            object_type=gene_model.DB_Object_Type,\n",
    "            GO_label=gene_model.GO_label,\n",
    "            GO_definition=gene_model.GO_definition\n",
    "        )\n",
    "\n",
    "    def create_gene_interaction(self, tx: Transaction, interaction: Dict[str, Any], disease_id: str):\n",
    "        \"\"\"\n",
    "        Create an INTERACTS_WITH relationship between Gene nodes in the Neo4j database.\n",
    "\n",
    "        Args:\n",
    "            tx: Neo4j transaction object.\n",
    "            interaction: Dictionary containing interaction data.\n",
    "            disease_id: ID of the associated disease.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            interaction_model = GeneInteraction(**interaction)\n",
    "        except ValidationError as e:\n",
    "            print(f\"Validation error for interaction {interaction['entry1']} -> {interaction['entry2']}: {e}\")\n",
    "            return\n",
    "\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MATCH (g1:Gene {unique_id: $entry_1})\n",
    "            MATCH (g2:Gene {unique_id: $entry_2})\n",
    "            MERGE (g1)-[r:INTERACTS_WITH {type: $interaction_type, subtypes: $subtypes}]->(g2)\n",
    "            \"\"\",\n",
    "            entry_1=f\"{disease_id}_{interaction_model.entry1}\",\n",
    "            entry_2=f\"{disease_id}_{interaction_model.entry2}\",\n",
    "            interaction_type=interaction_model.type,\n",
    "            subtypes=interaction_model.subtypes\n",
    "        )\n",
    "\n",
    "    def create_disease_association(self, tx: Transaction, gene: Dict[str, Any], disease_id: str, evidence: str):\n",
    "        \"\"\"\n",
    "        Create an ASSOCIATED_WITH relationship between Gene and Disease nodes in the Neo4j database.\n",
    "\n",
    "        Args:\n",
    "            tx: Neo4j transaction object.\n",
    "            gene: Dictionary containing gene data.\n",
    "            disease_id: ID of the associated disease.\n",
    "            evidence: Evidence for the association.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            gene_model = GeneData(**gene)\n",
    "        except ValidationError as e:\n",
    "            print(f\"Validation error for gene {gene['gene_id']}: {e}\")\n",
    "            return\n",
    "\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MATCH (g:Gene {unique_id: $unique_id})\n",
    "            MATCH (d:Disease {disease_id: $disease_id})\n",
    "            MERGE (g)-[r:ASSOCIATED_WITH {evidence: $evidence}]->(d)\n",
    "            \"\"\",\n",
    "            unique_id=f\"{disease_id}_{gene_model.gene_id}\",\n",
    "            disease_id=disease_id,\n",
    "            evidence=evidence\n",
    "        )\n",
    "\n",
    "    def create_go_association(self, tx: Transaction, gene: Dict[str, Any], disease_id: str):\n",
    "        \"\"\"\n",
    "        Create a HAS_GO_ANNOTATION relationship between Gene and GO Annotation nodes in the Neo4j database.\n",
    "\n",
    "        Args:\n",
    "            tx: Neo4j transaction object.\n",
    "            gene: Dictionary containing gene data.\n",
    "            disease_id: ID of the associated disease.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            gene_model = GeneData(**gene)\n",
    "        except ValidationError as e:\n",
    "            print(f\"Validation error for gene {gene['gene_id']}: {e}\")\n",
    "            return\n",
    "\n",
    "        tx.run(\n",
    "            \"\"\"\n",
    "            MATCH (g:Gene {unique_id: $unique_id})\n",
    "            MATCH (a:GO_Annotation {qualifier: $qualifier, GO_ID: $GO_ID})\n",
    "            MERGE (g)-[r:HAS_GO_ANNOTATION]->(a)\n",
    "            \"\"\",\n",
    "            unique_id=f\"{disease_id}_{gene_model.gene_id}\",\n",
    "            qualifier=gene_model.Qualifier,\n",
    "            GO_ID=gene_model.GO_ID\n",
    "        )\n",
    "\n",
    "    def import_data(self, kgml_files: List[str]):\n",
    "        \"\"\"\n",
    "        Import data from KGML files and a merged DataFrame into the Neo4j database.\n",
    "\n",
    "        Args:\n",
    "            kgml_files: List of file paths to KGML files.\n",
    "            merged: DataFrame containing merged data.\n",
    "        \"\"\"\n",
    "        with self.driver.session() as session:\n",
    "            for kgml_file_path in kgml_files:\n",
    "                pathway_data = parse_kgml(kgml_file_path)\n",
    "                disease_id = pathway_data['pathway_id']\n",
    "                disease_name = pathway_data['pathway_name']\n",
    "                \n",
    "                # Create disease node\n",
    "                session.write_transaction(self.create_disease_node, {\n",
    "                    'disease_id': disease_id,\n",
    "                    'name': disease_name\n",
    "                })\n",
    "\n",
    "                #Create dataframe integrating KGML genes and associated GO terms\n",
    "                merged=kegg_go_integration(kgml_file_path,gaf_path)\n",
    "\n",
    "                # Create gene nodes and their associations with the disease\n",
    "                for _, df in merged.groupby('gene_id'):\n",
    "                    for _, gene in df.iterrows():\n",
    "                        session.write_transaction(self.create_gene_node, gene.to_dict(), disease_id)\n",
    "                        session.write_transaction(self.create_go_node, gene.to_dict())\n",
    "                        session.write_transaction(self.create_disease_association, gene.to_dict(), disease_id, evidence=\"from KGML\")\n",
    "                        session.write_transaction(self.create_go_association, gene.to_dict(), disease_id)\n",
    "                        \n",
    "                # Create interactions between genes\n",
    "                for interaction in pathway_data['interactions']:\n",
    "                    session.write_transaction(self.create_gene_interaction, interaction, disease_id)\n",
    "\n",
    "\n",
    "\n",
    "# Connect to the Neo4j database\n",
    "driver = GraphDatabase.driver(uri, auth=(user, password))\n",
    "\n",
    "def create_constraints(driver):\n",
    "    with driver.session() as session:\n",
    "        session.run(\"CREATE CONSTRAINT FOR (d:Disease) REQUIRE d.disease_id IS UNIQUE;\")\n",
    "\n",
    "kgml_files = [\n",
    "              'data/KGML/hsa04930.xml',\n",
    "              'data/KGML/hsa05010.xml',\n",
    "              'data/KGML/hsa05012.xml',\n",
    "              'data/KGML/hsa05210.xml'\n",
    "              ]\n",
    "\n",
    "# Create constraints and import data\n",
    "importer = KGMLGAFImporter(driver)\n",
    "importer.import_data(kgml_files)\n",
    "importer.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.vectorstores.neo4j_vector import Neo4jVector\n",
    "from langchain.embeddings.openai import OpenAIEmbeddings\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.FeatureDeprecationWarning} {category: DEPRECATION} {title: This feature is deprecated and will be removed in future versions.} {description: The query used a deprecated procedure. ('db.create.setVectorProperty' has been replaced by 'db.create.setNodeVectorProperty')} {position: line: 1, column: 75, offset: 74} for query: \"UNWIND $data AS row MATCH (n:`GO_Annotation`) WHERE elementId(n) = row.id CALL db.create.setVectorProperty(n, 'embedding', row.embedding) YIELD node RETURN count(*)\"\n",
      "Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.FeatureDeprecationWarning} {category: DEPRECATION} {title: This feature is deprecated and will be removed in future versions.} {description: The query used a deprecated procedure. ('db.create.setVectorProperty' has been replaced by 'db.create.setNodeVectorProperty')} {position: line: 1, column: 75, offset: 74} for query: \"UNWIND $data AS row MATCH (n:`GO_Annotation`) WHERE elementId(n) = row.id CALL db.create.setVectorProperty(n, 'embedding', row.embedding) YIELD node RETURN count(*)\"\n"
     ]
    }
   ],
   "source": [
    "vector_index = Neo4jVector.from_existing_graph(\n",
    "    OpenAIEmbeddings(api_key=api_key),\n",
    "    url=uri,\n",
    "    username=user,\n",
    "    password=password,\n",
    "    index_name='go_ids',\n",
    "    node_label=\"GO_Annotation\",\n",
    "    text_node_properties=['qualifier', 'name', 'definition','aspect'],\n",
    "    embedding_node_property='embedding',\n",
    ")\n",
    "response = vector_index.similarity_search(\n",
    "    \"(Spermatogenesis-associated protein 2) inhibition of (Kinesin light chain 3)\",k=7,\n",
    "    #filter={'GO_ID':{\"$in\":[\"GO:0018\",\"001634\"]}}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(metadata={'object_type': 'protein', 'GO_ID': 'GO:0016938'}, page_content='\\nqualifier: part_of\\nname: kinesin I complex\\ndefinition: A complex of two kinesin heavy chains and two kinesin light chains.\\naspect: Cellular Component'),\n",
       " Document(metadata={'object_type': 'protein', 'GO_ID': 'GO:0005871'}, page_content='\\nqualifier: part_of\\nname: kinesin complex\\ndefinition: Any complex that includes a dimer of molecules from the kinesin superfamily, a group of related proteins that contain an extended region of predicted alpha-helical coiled coil in the main chain that likely produces dimerization. The native complexes of several kinesin family members have also been shown to contain additional peptides, often designated light chains as all of the noncatalytic subunits that are currently known are smaller than the chain that contains the motor unit. Kinesin complexes generally possess a force-generating enzymatic activity, or motor, which converts the free energy of the gamma phosphate bond of ATP into mechanical work.\\naspect: Cellular Component'),\n",
       " Document(metadata={'object_type': 'protein', 'GO_ID': 'GO:0004861'}, page_content='\\nqualifier: enables\\nname: cyclin-dependent protein serine/threonine kinase inhibitor activity\\ndefinition: Binds to and stops, prevents or reduces the activity of a cyclin-dependent protein serine/threonine kinase.\\naspect: Molecular Function'),\n",
       " Document(metadata={'object_type': 'protein', 'GO_ID': 'GO:1990111'}, page_content='\\nqualifier: part_of\\nname: spermatoproteasome complex\\ndefinition: A proteasome specifically found in mammalian testis. Contains the proteasome activator PA200 in the regulatory particle, and beta1i, beta2i, beta5i and/or alpha4s in the core (20S) subunit. Beta1i, beta2i and beta5i are inducible catalytic subunits, closely related to beta1, beta2 and beta5. Alpha4s is a sperm-specific 20S subunit, but unlike other alternative 20S subunits alpha4s lies in the outer alpha-ring and lacks catalytic activity.\\naspect: Cellular Component'),\n",
       " Document(metadata={'object_type': 'protein', 'GO_ID': 'GO:0090266'}, page_content='\\nqualifier: involved_in\\nname: regulation of mitotic cell cycle spindle assembly checkpoint\\ndefinition: Any process that modulates the rate, frequency, or extent of the mitotic cell cycle spindle assembly checkpoint, a cell cycle checkpoint that delays the metaphase/anaphase transition of a mitotic nuclear division until the spindle is correctly assembled and chromosomes are attached to the spindle.\\naspect: Biological Process'),\n",
       " Document(metadata={'object_type': 'protein', 'GO_ID': 'GO:0031115'}, page_content='\\nqualifier: involved_in\\nname: negative regulation of microtubule polymerization\\ndefinition: Any process that stops, prevents, or reduces the frequency, rate or extent of microtubule polymerization.\\naspect: Biological Process'),\n",
       " Document(metadata={'object_type': 'protein', 'GO_ID': 'GO:0060236'}, page_content='\\nqualifier: involved_in\\nname: regulation of mitotic spindle organization\\ndefinition: Any process that modulates the rate, frequency or extent of the assembly, arrangement of constituent parts, or disassembly of the microtubule spindle during a mitotic cell cycle.\\naspect: Biological Process')]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"I don't have the specific GO_ID for the interaction between Spermatogenesis-associated protein 2 and Kinesin light chain 3.\""
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain.chains import RetrievalQA\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "vector_qa = RetrievalQA.from_chain_type(\n",
    "    llm=ChatOpenAI(api_key=api_key),\n",
    "    chain_type=\"stuff\",\n",
    "    retriever=vector_index.as_retriever()\n",
    ")\n",
    "\n",
    "vector_qa.run(\n",
    "    \"Which GO_ID corresponding to (Spermatogenesis-associated protein 2) describes best the following interaction:(Spermatogenesis-associated protein 2) inhibition of (Kinesin light chain 3)\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** BASE CLASS AND DERIVED CLASSES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml\n",
    "from langchain_community.graphs.neo4j_graph import Neo4jGraph\n",
    "from neo4j.exceptions import CypherSyntaxError\n",
    "from langchain.vectorstores.neo4j_vector import Neo4jVector\n",
    "from langchain.embeddings.openai import OpenAIEmbeddings\n",
    "import textwrap\n",
    "from openai import OpenAI\n",
    "import os\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "\n",
    "uri = os.getenv(\"NEO4J_URI\")\n",
    "user = os.getenv(\"NEO4J_USER\")\n",
    "password = os.getenv(\"NEO4J_PASSWORD\")\n",
    "api_key = os.getenv(\"OPENAI_API_KEY\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "class BaseAgent:\n",
    "    def __init__(self, uri: str, user: str, password: str, api_key: str, config_path: str):\n",
    "        \"\"\"\n",
    "        Initialize the BaseAgent with Neo4jGraph, OpenAI API key, and configuration.\n",
    "\n",
    "        Args:\n",
    "            uri (str): URI for the Neo4j database.\n",
    "            user (str): Username for the Neo4j database.\n",
    "            password (str): Password for the Neo4j database.\n",
    "            api_key (str): API key for OpenAI.\n",
    "            config_path (str): Path to the configuration YAML file.\n",
    "        \"\"\"\n",
    "        self.kg = Neo4jGraph(url=uri, username=user, password=password)\n",
    "        self.api_key = api_key\n",
    "        self.config = self.load_config(config_path)\n",
    "        self.schema = textwrap.fill(self.kg.schema, 60)\n",
    "\n",
    "    def load_config(self, path: str) -> dict:\n",
    "        \"\"\"\n",
    "        Load configuration from a YAML file.\n",
    "\n",
    "        Args:\n",
    "            path (str): Path to the YAML configuration file.\n",
    "\n",
    "        Returns:\n",
    "            dict: Configuration dictionary.\n",
    "        \"\"\"\n",
    "        with open(path, 'r') as file:\n",
    "            return yaml.safe_load(file)\n",
    "\n",
    "    def generate(self, prompt_template: str, temperature: float = 0, **kwargs) -> str:\n",
    "        \"\"\"\n",
    "        Generate a response using the OpenAI API.\n",
    "\n",
    "        Args:\n",
    "            prompt_template (str): Template for the prompt to be sent to OpenAI.\n",
    "            temperature (float, optional): Sampling temperature for OpenAI completion. Defaults to 0.\n",
    "            **kwargs: Additional keyword arguments for formatting the prompt.\n",
    "\n",
    "        Returns:\n",
    "            str: Generated response from OpenAI.\n",
    "        \"\"\"\n",
    "        prompt = prompt_template.format(**kwargs)\n",
    "        messages = [\n",
    "            {\"role\": \"system\", \"content\": prompt},\n",
    "            {\"role\": \"user\", \"content\": kwargs.get('question', '')},\n",
    "        ]\n",
    "\n",
    "        client = OpenAI(api_key=self.api_key)\n",
    "\n",
    "        completions = client.chat.completions.create(\n",
    "            model=\"gpt-4o\",\n",
    "            temperature=temperature,\n",
    "            messages=messages,\n",
    "            max_tokens=500\n",
    "        )\n",
    "        return completions.choices[0].message.content\n",
    "\n",
    "    def run_cypher_query(self, cypher_statement: str, retry: bool = True) -> dict:\n",
    "        \"\"\"\n",
    "        Execute a Cypher query against the Neo4j database.\n",
    "\n",
    "        Args:\n",
    "            cypher_statement (str): Cypher query to be executed.\n",
    "            retry (bool, optional): Whether to retry the query in case of a Cypher syntax error. Defaults to True.\n",
    "\n",
    "        Returns:\n",
    "            dict: Result of the Cypher query.\n",
    "\n",
    "        Raises:\n",
    "            CypherSyntaxError: If the Cypher query has a syntax error and retry is False.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            print(cypher_statement)\n",
    "            return self.kg.query(cypher_statement)\n",
    "        except CypherSyntaxError as e:\n",
    "            if retry:\n",
    "                print(\"Retrying due to Cypher syntax error...\")\n",
    "                return self.generate_cypher_retry(cypher_statement, str(e))\n",
    "            else:\n",
    "                raise e\n",
    "\n",
    "    def generate_cypher_retry(self, cypher_statement: str, error_message: str) -> str:\n",
    "        \"\"\"\n",
    "        Generate a corrected Cypher query in case of a syntax error.\n",
    "\n",
    "        Args:\n",
    "            cypher_statement (str): Original Cypher query that caused the syntax error.\n",
    "            error_message (str): Error message from the Cypher syntax error.\n",
    "\n",
    "        Returns:\n",
    "            str: Corrected Cypher query.\n",
    "        \"\"\"\n",
    "        retry_prompt = self.current_prompts['retry_prompt']\n",
    "        return self.generate(retry_prompt, cypher_statement=cypher_statement, error_message=error_message)\n",
    "    \n",
    "    def set_prompt_context(self, context: str):\n",
    "        \"\"\"\n",
    "        Set the current prompt context based on the configuration.\n",
    "\n",
    "        Args:\n",
    "            context (str): Context key to set the current prompts.\n",
    "\n",
    "        Raises:\n",
    "            ValueError: If the context is not found in the configuration.\n",
    "        \"\"\"\n",
    "        if context in self.config['prompts']:\n",
    "            self.current_prompts = self.config['prompts'][context]\n",
    "        else:\n",
    "            raise ValueError(f\"Context '{context}' not found in configuration.\")\n",
    " \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MATCH (g:Gene)-[:ASSOCIATED_WITH]->(d:Disease)\n",
      "WHERE 'Insulin' IN g.names OR 'Insulin' IN g.synonyms\n",
      "RETURN g.names as gene_names, g.synonyms as gene_synonyms, d.name as disease_name, d.disease_id as KEGG_pathway\n",
      "[{'gene_names': ['Insulin'], 'gene_synonyms': ['INS'], 'disease_name': 'Alzheimer disease', 'KEGG_pathway': '05010'}, {'gene_names': ['Insulin'], 'gene_synonyms': ['INS'], 'disease_name': 'Type II diabetes mellitus', 'KEGG_pathway': '04930'}, {'gene_names': ['Insulin'], 'gene_synonyms': ['INS'], 'disease_name': 'Type II diabetes mellitus', 'KEGG_pathway': '04930'}]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"Insulin, also known as 'INS', is associated with Alzheimer disease and Type II diabetes mellitus. It appears in pathway 05010 of the KEGG database for Alzheimer disease and in pathway 04930 for Type II diabetes mellitus.\""
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class DiseaseAssociation(BaseAgent):\n",
    "    def __init__(self, uri: str, user: str, password: str, api_key: str, config_path: str):\n",
    "        \"\"\"\n",
    "        Initialize the DiseaseAssociation with Neo4jGraph, OpenAI API key, and configuration.\n",
    "\n",
    "        Args:\n",
    "            uri (str): URI for the Neo4j database.\n",
    "            user (str): Username for the Neo4j database.\n",
    "            password (str): Password for the Neo4j database.\n",
    "            api_key (str): API key for OpenAI.\n",
    "            config_path (str): Path to the configuration YAML file.\n",
    "        \"\"\"        \n",
    "        super().__init__(uri, user, password, api_key, config_path)\n",
    "        \n",
    "    def generate_response(self, question: str) -> str:\n",
    "        \"\"\"\n",
    "        Generate a response for a given question about disease associations.\n",
    "\n",
    "        Args:\n",
    "            question (str): The question to be answered.\n",
    "\n",
    "        Returns:\n",
    "            str: The generated response from OpenAI.\n",
    "        \"\"\"\n",
    "        # Generate initial Cypher statement\n",
    "        self.set_prompt_context('disease_association')\n",
    "        initial_prompt = self.current_prompts['initial_prompt']\n",
    "        cypher_statement = self.generate(initial_prompt, schema=self.schema, question=question)\n",
    "        \n",
    "        # Execute Cypher query and get results\n",
    "        cypher_result = self.run_cypher_query(cypher_statement)\n",
    "        print(cypher_result)\n",
    "\n",
    "        # Generate final response\n",
    "        final_prompt = self.current_prompts['final_prompt']\n",
    "        final_response = self.generate(final_prompt, question=question, information=cypher_result)\n",
    "        \n",
    "        return final_response\n",
    "\n",
    "# Usage example\n",
    "custom_query = DiseaseAssociation(\n",
    "    uri=uri,\n",
    "    user=user,\n",
    "    password=password,\n",
    "    api_key=api_key,\n",
    "    config_path='config.yaml'\n",
    ")\n",
    "custom_query.generate_response(\"What diseases is Insulin associated with?\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TESTING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Any, Dict, List\n",
    "from utilities.preprocessing import dict_to_frozenset\n",
    "\n",
    "\n",
    "\n",
    "class DownstreamInteraction(BaseAgent):\n",
    "    vector_index = None\n",
    "\n",
    "    def __init__(self, uri: str, user: str, password: str, api_key: str, config_path: str):\n",
    "        \"\"\"\n",
    "        Initialize the DownstreamInteraction with Neo4jGraph, OpenAI API key, and configuration.\n",
    "\n",
    "        Args:\n",
    "            uri (str): URI for the Neo4j database.\n",
    "            user (str): Username for the Neo4j database.\n",
    "            password (str): Password for the Neo4j database.\n",
    "            api_key (str): API key for OpenAI.\n",
    "            config_path (str): Path to the configuration YAML file.\n",
    "        \"\"\"\n",
    "        super().__init__(uri, user, password, api_key, config_path)\n",
    "        if DownstreamInteraction.vector_index is None:\n",
    "            DownstreamInteraction.vector_index = Neo4jVector.from_existing_graph(\n",
    "                OpenAIEmbeddings(api_key=api_key),\n",
    "                url=uri,\n",
    "                username=user,\n",
    "                password=password,\n",
    "                index_name='go_ids',\n",
    "                node_label=\"GO_Annotation\",\n",
    "                text_node_properties=['qualifier', 'name', 'definition','aspect'],\n",
    "                embedding_node_property='embedding',\n",
    "            )\n",
    "\n",
    "    def get_go_ids(self, unique_id: str) -> List[str]:\n",
    "        \"\"\"\n",
    "        Retrieve GO IDs associated with a gene unique ID.\n",
    "\n",
    "        Args:\n",
    "            unique_id (str): Unique ID of the gene.\n",
    "\n",
    "        Returns:\n",
    "            List[str]: List of GO IDs associated with the gene.\n",
    "        \"\"\"\n",
    "        go_result = self.kg.query(\n",
    "            \"\"\"\n",
    "            MATCH (g: Gene {unique_id:$unique_id})-[:HAS_GO_ANNOTATION]->(a: GO_Annotation)\n",
    "            RETURN collect(a.GO_ID) as GO_ID\n",
    "            \"\"\",\n",
    "            params={'unique_id': unique_id} \n",
    "        )\n",
    "        go_list = go_result[0]['GO_ID']\n",
    "        return go_list\n",
    "\n",
    "    def perform_similarity_search(self, interaction: Dict[str, Any], go_list: List[str]) -> str:\n",
    "        \"\"\"\n",
    "        Perform a similarity search using the interaction details and GO IDs.\n",
    "\n",
    "        Args:\n",
    "            interaction (Dict[str, Any]): Interaction details.\n",
    "            go_list (List[str]): List of GO IDs.\n",
    "\n",
    "        Returns:\n",
    "            str: Description of the interaction from the similarity search.\n",
    "        \"\"\"\n",
    "        start_node_names=interaction['start'].get('names')\n",
    "        end_node_names=interaction['end'].get('names')\n",
    "        subtype=interaction['subtypes']\n",
    "        interaction_type=self.config['interaction_type_dict'].get(interaction['type'])\n",
    "\n",
    "        #form crude search query consisting of start and end node gene names, as well as (sub)type of the interaction connecting them (eg activation, inhibition etc).\n",
    "        search_query = f\"{start_node_names}, {subtype}, {end_node_names}, {interaction_type}\"\n",
    "\n",
    "        #perform similarity search based on the search_query- filter for GO_IDs relevant to the first node.\n",
    "        response = DownstreamInteraction.vector_index.similarity_search(\n",
    "            search_query, k=1,\n",
    "            filter={'GO_ID': {\"$in\": go_list}}\n",
    "        )\n",
    "        return response[0].page_content\n",
    "\n",
    "    def process_interaction(self, interaction: Dict[str, Any]) -> str:\n",
    "        \"\"\"\n",
    "        Process an interaction to generate a descriptive response.\n",
    "\n",
    "        Args:\n",
    "            interaction (Dict[str, Any]): Interaction details.\n",
    "\n",
    "        Returns:\n",
    "            str: Generated response describing the interaction.\n",
    "        \"\"\"\n",
    "        unique_id = interaction['start'].get('unique_id')\n",
    "        go_list = self.get_go_ids(unique_id)\n",
    "        interaction_description = self.perform_similarity_search(interaction,go_list)\n",
    "        final_prompt = self.current_prompts['final_prompt']\n",
    "        start_node_names=[interaction['start'].get('names')]\n",
    "        end_node_names=[interaction['end'].get('names')]\n",
    "        question=f\"Please describe the interaction of {start_node_names} with {end_node_names} using the description {interaction_description}.\"\n",
    "        final_response = self.generate(\n",
    "            final_prompt,\n",
    "            temperature=0.5,\n",
    "            question=question\n",
    "        )\n",
    "        return final_response\n",
    "\n",
    "    def generate_response(self, question: str) -> List[List[str]]:\n",
    "        \"\"\"\n",
    "        Generate a response for a given question about downstream interactions.\n",
    "\n",
    "        Args:\n",
    "            question (str): The question to be answered.\n",
    "\n",
    "        Returns:\n",
    "            List[List[str]]: List of all distinct downstream interaction paths from the specified gene node.\n",
    "        \"\"\"\n",
    "        # Set the context for downstream interaction prompts\n",
    "        self.set_prompt_context('downstream_interaction')\n",
    "        \n",
    "        # Generate initial Cypher statement\n",
    "        initial_prompt = self.current_prompts['initial_prompt']\n",
    "        cypher_statement = self.generate(initial_prompt, schema=self.schema, question=question)\n",
    "        \n",
    "        # Execute Cypher query and get results\n",
    "        cypher_result = self.run_cypher_query(cypher_statement)\n",
    "        \n",
    "        processed_interactions = {}\n",
    "        all_paths_list = []\n",
    "        for path in cypher_result[0]['interactions']:\n",
    "            path_list=[]\n",
    "            for interaction in path:\n",
    "                interaction_frozenset = dict_to_frozenset(interaction)\n",
    "        \n",
    "                # Check if this interaction has already been processed\n",
    "                if interaction_frozenset in processed_interactions:\n",
    "                    # Retrieve the result from the dictionary and append it to path_description\n",
    "                    path_list.append(processed_interactions[interaction_frozenset])\n",
    "                    continue\n",
    "                interaction_result=self.process_interaction(interaction)\n",
    "                path_list.append(interaction_result)\n",
    "                processed_interactions[interaction_frozenset] = interaction_result\n",
    "            all_paths_list.append(path_list)\n",
    "\n",
    "        if not all_paths_list:\n",
    "            return \"I could not find the answer in the database. Please try again.\"\n",
    "        \n",
    "        print(\"The following are the distinct downstream interaction paths from the specified gene node:\")\n",
    "        return all_paths_list\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Received notification from DBMS server: {severity: WARNING} {code: Neo.ClientNotification.Statement.FeatureDeprecationWarning} {category: DEPRECATION} {title: This feature is deprecated and will be removed in future versions.} {description: The query used a deprecated procedure. ('db.create.setVectorProperty' has been replaced by 'db.create.setNodeVectorProperty')} {position: line: 1, column: 75, offset: 74} for query: \"UNWIND $data AS row MATCH (n:`GO_Annotation`) WHERE elementId(n) = row.id CALL db.create.setVectorProperty(n, 'embedding', row.embedding) YIELD node RETURN count(*)\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MATCH (start: Gene)\n",
      "WHERE ('PARK7' IN start.names) or ('PARK7' IN start.synonyms)\n",
      "CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \n",
      "YIELD path\n",
      "WITH path\n",
      "WHERE NOT EXISTS {\n",
      "    MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\n",
      "    WHERE lastNode = last(nodes(path))\n",
      "}\n",
      "WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\n",
      "RETURN collect(relationships) AS interactions\n",
      "The following are the distinct downstream interaction paths from the specified gene node:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[['Parkinson disease protein 7 interacts with Nuclear factor erythroid 2-related factor 2 (also known as Heme-binding protein 1) via enabling protein binding. This interaction involves the binding of Parkinson disease protein 7 to another protein.',\n",
       "  'Nuclear factor erythroid 2-related factor 2 (also known as Heme-binding protein 1) interacts with Thioredoxin (also known as Thioredoxin, mitochondrial, Histone-lysine N-methyltransferase 2B) via enabling RNA polymerase II-specific DNA-binding transcription factor binding. This interaction involves binding to a sequence-specific DNA binding RNA polymerase II transcription factor, which selectively and non-covalently interacts with a specific DNA sequence to modulate transcription. This interaction falls under the aspect of Molecular Function.'],\n",
       " ['Parkinson disease protein 7 interacts with Cellular tumor antigen p53 via enabling protein binding. This interaction involves the binding of Parkinson disease protein 7 to Cellular tumor antigen p53, which is categorized under the Molecular Function aspect.',\n",
       "  'Cellular tumor antigen p53 interacts with Dual specificity protein phosphatase 1 and is involved in the DNA damage response, specifically in signal transduction by the p53 class mediator, resulting in the transcription of the p21 class mediator. This biological process is a cascade of events initiated by the cell cycle regulator phosphoprotein p53, leading to the induction of p21 transcription in response to DNA damage.'],\n",
       " ['Parkinson disease protein 7 interacts with Bcl-2-like protein 1 via enabling protein binding, which means that Parkinson disease protein 7 binds to Bcl-2-like protein 1. This interaction falls under the aspect of Molecular Function.',\n",
       "  'Bcl-2-like protein 1 interacts with Apoptosis regulator BAX as part of the Bcl-2 family protein complex. This complex consists of members of the Bcl-2 family of anti- and proapoptotic regulators, which respond to intracellular stress cues like DNA damage or cytokine deprivation. These proteins interact with each other to determine whether the caspase proteolytic cascade should be activated.',\n",
       "  'Apoptosis regulator BAX interacts with Cytochrome c as part of the Bcl-2 family protein complex. This complex consists of members of the Bcl-2 family of anti- and proapoptotic regulators, which respond to various forms of intracellular stress, such as DNA damage or cytokine deprivation. These proteins interact with opposing family members to determine whether or not the caspase proteolytic cascade should be unleashed.',\n",
       "  'Cytochrome c interacts with Caspase-9 as part of the apoptosome, a multisubunit protein complex involved in the signaling phase of the apoptotic process. In mammals, the apoptosome typically consists of seven Apaf-1 subunits bound to cytochrome c and caspase-9, and similar complexes promoting apoptosis are formed from homologous gene products in other eukaryotic organisms.',\n",
       "  'Caspase-9 interacts with Caspase-3 as part of a caspase complex. This complex is a protein assembly that includes one or more cysteine-type endopeptidases, also known as caspases, which provide the complex with peptidase activity specific for the hydrolysis of aspartyl bonds. These complexes are involved in processes such as apoptosis or inflammation.']]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "downstream_query = DownstreamInteraction(\n",
    "    uri=uri,\n",
    "    user=user,\n",
    "    password=password,\n",
    "    api_key=api_key,\n",
    "    config_path='config.yaml'\n",
    ")\n",
    "\n",
    "downstream_query.generate_response('What are the downstream interactions of PARK7?')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.agents import AgentExecutor, create_react_agent\n",
    "from langchain.tools import Tool\n",
    "from langchain import hub\n",
    "from langchain_core.runnables.history import RunnableWithMessageHistory\n",
    "from langchain_community.chat_message_histories import Neo4jChatMessageHistory\n",
    "from openai import OpenAI\n",
    "from langchain.agents import AgentType\n",
    "\n",
    "from langchain.agents import initialize_agent\n",
    "\n",
    "tools = [\n",
    "    Tool.from_function(\n",
    "        name=\"Disease Association\",\n",
    "        description=\"For when you need to ask questions like 'What disease is gene X associated with?'. The question will be a string. Return a string.\",\n",
    "        func=DiseaseAssociation(uri,user,password,api_key,config_path='config.yaml').generate_response\n",
    "    ),\n",
    "\n",
    "    Tool.from_function(\n",
    "        name=\"Downstream Interaction\",\n",
    "        description=\"For answering text based questions like 'What are the downstream interactions of gene X in disease pathway Y?' The question will be a String. Return a String\",\n",
    "        func=DownstreamInteraction(uri,user,password,api_key,config_path='config.yaml').generate_response\n",
    "    )\n",
    "]\n",
    "\n",
    "llm = ChatOpenAI(temperature=0, model=\"gpt-4o\")\n",
    "\n",
    "agent= initialize_agent(\n",
    "    tools, \n",
    "    llm, \n",
    "    agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION,\n",
    "    handle_parsing_errors=True,\n",
    "    verbose = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n",
      "\u001b[32;1m\u001b[1;3mTo determine the disease associated with the gene PARK7 and provide associated evidence, I will use the Disease Association tool.\n",
      "\n",
      "Action: Disease Association\n",
      "Action Input: What disease is gene PARK7 associated with? Provide associated evidence.\u001b[0mMATCH (g:Gene)-[r:ASSOCIATED_WITH]->(d:Disease)\n",
      "WHERE 'PARK7' IN g.synonyms\n",
      "RETURN g.names as gene_names, g.synonyms as gene_synonyms, d.name as disease_name, d.disease_id as KEGG_pathway, r.evidence as association_evidence\n",
      "[{'gene_names': ['Parkinson disease protein 7'], 'gene_synonyms': ['PARK7'], 'disease_name': 'Parkinson disease', 'KEGG_pathway': '05012', 'association_evidence': 'from KGML'}, {'gene_names': ['Parkinson disease protein 7'], 'gene_synonyms': ['PARK7'], 'disease_name': 'Parkinson disease', 'KEGG_pathway': '05012', 'association_evidence': 'from KGML'}]\n",
      "\n",
      "Observation: \u001b[36;1m\u001b[1;3mGene PARK7, also known as 'Parkinson disease protein 7', is associated with Parkinson disease. This association is supported by evidence from the KEGG pathway 05012, as indicated in the KGML.\u001b[0m\n",
      "Thought:\u001b[32;1m\u001b[1;3mI now know the final answer.\n",
      "\n",
      "Final Answer: Gene PARK7, also known as 'Parkinson disease protein 7', is associated with Parkinson disease. This association is supported by evidence from the KEGG pathway 05012, as indicated in the KGML.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'input': 'What disease is PARK7 associated with??Provide associated evidence',\n",
       " 'output': \"Gene PARK7, also known as 'Parkinson disease protein 7', is associated with Parkinson disease. This association is supported by evidence from the KEGG pathway 05012, as indicated in the KGML.\"}"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "agent('What disease is PARK7 associated with?? Provide associated evidence')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n",
      "\u001b[32;1m\u001b[1;3mTo answer the question about the downstream interactions of PARK7, I will use the Downstream Interaction tool.\n",
      "\n",
      "Action: Downstream Interaction\n",
      "Action Input: What are the downstream interactions of PARK7?\u001b[0mMATCH (start: Gene)\n",
      "WHERE 'PARK7' IN start.synonyms\n",
      "CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \n",
      "YIELD path\n",
      "WITH path\n",
      "WHERE NOT EXISTS {\n",
      "    MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\n",
      "    WHERE lastNode = last(nodes(path))\n",
      "}\n",
      "WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\n",
      "RETURN collect(relationships) AS interactions\n",
      "The following are the distinct downstream interaction paths from the specified gene node:\n",
      "\n",
      "Observation: \u001b[33;1m\u001b[1;3m[['Parkinson disease protein 7 interacts with Nuclear factor erythroid 2-related factor 2 (also known as Heme-binding protein 1) via enabling protein binding. This interaction involves the binding to a protein, which is a molecular function.', 'Nuclear factor erythroid 2-related factor 2 (also known as Heme-binding protein 1) interacts with Thioredoxin (also known as Thioredoxin, mitochondrial, Histone-lysine N-methyltransferase 2B) via enabling RNA polymerase II-specific DNA-binding transcription factor binding. This interaction involves binding to a sequence-specific DNA binding RNA polymerase II transcription factor, which are factors that interact selectively and non-covalently with a specific DNA sequence to modulate transcription. This falls under the aspect of Molecular Function.'], ['Parkinson disease protein 7 interacts with Cellular tumor antigen p53 via enabling protein binding. This interaction involves the binding of Parkinson disease protein 7 to Cellular tumor antigen p53.', \"'Cellular tumor antigen p53' interacts with 'Dual specificity protein phosphatase 1' in a process involved in the DNA damage response, specifically signal transduction by the p53 class mediator resulting in the transcription of the p21 class mediator. This means that 'Cellular tumor antigen p53' plays a crucial role in a cascade of processes that are triggered by the detection of DNA damage, leading to the induction of p21 or its equivalents, with 'Dual specificity protein phosphatase 1' being part of this pathway.\"], [\"'Parkinson disease protein 7' interacts with 'Bcl-2-like protein 1' via enabling protein binding, which means it binds to a protein. This interaction falls under the aspect of Molecular Function.\", 'Bcl-2-like protein 1 interacts with Apoptosis regulator BAX as part of the Bcl-2 family protein complex. This complex consists of members of the Bcl-2 family of anti- and proapoptotic regulators. Bcl-2 proteins respond to cues from various forms of intracellular stress, such as DNA damage or cytokine deprivation, and interact with opposing family members to determine whether or not the caspase proteolytic cascade should be unleashed.', 'Apoptosis regulator BAX interacts with Cytochrome c as part of the Bcl-2 family protein complex. This complex consists of members of the Bcl-2 family of anti- and proapoptotic regulators. Bcl-2 proteins respond to various forms of intracellular stress, such as DNA damage or cytokine deprivation, and interact with opposing family members to determine whether or not the caspase proteolytic cascade should be unleashed.', 'Cytochrome c interacts with Caspase-9 as part of the apoptosome, a multisubunit protein complex involved in the signaling phase of the apoptotic process. In mammals, this complex typically consists of seven Apaf-1 subunits bound to cytochrome c and caspase-9, and similar complexes are formed from homologous gene products in other eukaryotic organisms.', 'Caspase-9 interacts with Caspase-3 as part of a caspase complex. This complex is a protein assembly that includes one or more cysteine-type endopeptidases, known as caspases, which have peptidase activity specific for hydrolyzing aspartyl bonds. These complexes are typically involved in processes such as apoptosis or inflammation.']]\u001b[0m\n",
      "Thought:\u001b[32;1m\u001b[1;3mI now have the information on the downstream interactions of PARK7. Here are the detailed interactions:\n",
      "\n",
      "1. **PARK7 (Parkinson disease protein 7) interacts with Nuclear factor erythroid 2-related factor 2 (also known as Heme-binding protein 1) via enabling protein binding.**\n",
      "   - This interaction involves the binding to a protein, which is a molecular function.\n",
      "   - Nuclear factor erythroid 2-related factor 2 interacts with Thioredoxin (also known as Thioredoxin, mitochondrial, Histone-lysine N-methyltransferase 2B) via enabling RNA polymerase II-specific DNA-binding transcription factor binding.\n",
      "   - This interaction involves binding to a sequence-specific DNA binding RNA polymerase II transcription factor, which are factors that interact selectively and non-covalently with a specific DNA sequence to modulate transcription. This falls under the aspect of Molecular Function.\n",
      "\n",
      "2. **PARK7 interacts with Cellular tumor antigen p53 via enabling protein binding.**\n",
      "   - This interaction involves the binding of PARK7 to Cellular tumor antigen p53.\n",
      "   - Cellular tumor antigen p53 interacts with Dual specificity protein phosphatase 1 in a process involved in the DNA damage response, specifically signal transduction by the p53 class mediator resulting in the transcription of the p21 class mediator.\n",
      "   - This means that Cellular tumor antigen p53 plays a crucial role in a cascade of processes that are triggered by the detection of DNA damage, leading to the induction of p21 or its equivalents, with Dual specificity protein phosphatase 1 being part of this pathway.\n",
      "\n",
      "3. **PARK7 interacts with Bcl-2-like protein 1 via enabling protein binding.**\n",
      "   - This interaction falls under the aspect of Molecular Function.\n",
      "   - Bcl-2-like protein 1 interacts with Apoptosis regulator BAX as part of the Bcl-2 family protein complex.\n",
      "   - This complex consists of members of the Bcl-2 family of anti- and proapoptotic regulators. Bcl-2 proteins respond to cues from various forms of intracellular stress, such as DNA damage or cytokine deprivation, and interact with opposing family members to determine whether or not the caspase proteolytic cascade should be unleashed.\n",
      "   - Apoptosis regulator BAX interacts with Cytochrome c as part of the Bcl-2 family protein complex.\n",
      "   - Cytochrome c interacts with Caspase-9 as part of the apoptosome, a multisubunit protein complex involved in the signaling phase of the apoptotic process. In mammals, this complex typically consists of seven Apaf-1 subunits bound to cytochrome c and caspase-9, and similar complexes are formed from homologous gene products in other eukaryotic organisms.\n",
      "   - Caspase-9 interacts with Caspase-3 as part of a caspase complex.\n",
      "   - This complex is a protein assembly that includes one or more cysteine-type endopeptidases, known as caspases, which have peptidase activity specific for hydrolyzing aspartyl bonds. These complexes are typically involved in processes such as apoptosis or inflammation.\n",
      "\n",
      "Final Answer: The downstream interactions of PARK7 are as follows:\n",
      "\n",
      "1. PARK7 interacts with Nuclear factor erythroid 2-related factor 2 (Heme-binding protein 1) via enabling protein binding. This interaction involves the binding to a protein, which is a molecular function. Nuclear factor erythroid 2-related factor 2 interacts with Thioredoxin via enabling RNA polymerase II-specific DNA-binding transcription factor binding.\n",
      "\n",
      "2. PARK7 interacts with Cellular tumor antigen p53 via enabling protein binding. Cellular tumor antigen p53 interacts with Dual specificity protein phosphatase 1 in a process involved in the DNA damage response, specifically signal transduction by the p53 class mediator resulting in the transcription of the p21 class mediator.\n",
      "\n",
      "3. PARK7 interacts with Bcl-2-like protein 1 via enabling protein binding. Bcl-2-like protein 1 interacts with Apoptosis regulator BAX as part of the Bcl-2 family protein complex. Apoptosis regulator BAX interacts with Cytochrome c as part of the Bcl-2 family protein complex. Cytochrome c interacts with Caspase-9 as part of the apoptosome. Caspase-9 interacts with Caspase-3 as part of a caspase complex.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'input': 'What are the downstream interactions of PARK7 ? Write all the interactions that your tool provides, do not summarize or omit information',\n",
       " 'output': 'The downstream interactions of PARK7 are as follows:\\n\\n1. PARK7 interacts with Nuclear factor erythroid 2-related factor 2 (Heme-binding protein 1) via enabling protein binding. This interaction involves the binding to a protein, which is a molecular function. Nuclear factor erythroid 2-related factor 2 interacts with Thioredoxin via enabling RNA polymerase II-specific DNA-binding transcription factor binding.\\n\\n2. PARK7 interacts with Cellular tumor antigen p53 via enabling protein binding. Cellular tumor antigen p53 interacts with Dual specificity protein phosphatase 1 in a process involved in the DNA damage response, specifically signal transduction by the p53 class mediator resulting in the transcription of the p21 class mediator.\\n\\n3. PARK7 interacts with Bcl-2-like protein 1 via enabling protein binding. Bcl-2-like protein 1 interacts with Apoptosis regulator BAX as part of the Bcl-2 family protein complex. Apoptosis regulator BAX interacts with Cytochrome c as part of the Bcl-2 family protein complex. Cytochrome c interacts with Caspase-9 as part of the apoptosome. Caspase-9 interacts with Caspase-3 as part of a caspase complex.'}"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "agent('What are the downstream interactions of PARK7 ? Write all the interactions that your tool provides, do not summarize or omit information')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable\n",
    "\n",
    "class CustomAgent(BaseAgent):\n",
    "    \"\"\"\n",
    "    CustomAgent class that inherits from BaseAgent to handle questions related to disease associations \n",
    "    and downstream interactions. It uses OpenAI to classify questions and selects the appropriate tool \n",
    "    (DiseaseAssociation or DownstreamInteraction) to generate responses.\n",
    "\n",
    "    Example usage:\n",
    "    config_path = 'config.yaml'\n",
    "    custom_agent = CustomAgent(uri, user, password, api_key, config_path)\n",
    "    response = custom_agent.ask(\"What are the downstream interactions of gene INS in the pathway Type II diabetes mellitus?\")\n",
    "    print(response)\n",
    "    \"\"\"\n",
    "    def __init__(self, uri: str, user: str, password: str, api_key: str, config_path: str):\n",
    "        \"\"\"\n",
    "        Initialize the CustomAgent with Neo4jGraph, OpenAI API key, and configuration.\n",
    "\n",
    "        Args:\n",
    "            uri (str): URI for the Neo4j database.\n",
    "            user (str): Username for the Neo4j database.\n",
    "            password (str): Password for the Neo4j database.\n",
    "            api_key (str): API key for OpenAI.\n",
    "            config_path (str): Path to the configuration YAML file.\n",
    "        \"\"\"\n",
    "        super().__init__(uri, user, password, api_key, config_path)\n",
    "        self.disease_association_agent = DiseaseAssociation(uri, user, password, api_key,config_path)\n",
    "        self.downstream_interaction_agent = DownstreamInteraction(uri, user, password, api_key,config_path)\n",
    "\n",
    "    def classify_question(self, question: str) -> str:\n",
    "        \"\"\"\n",
    "        Classify the question using OpenAI to determine which tool to use.\n",
    "\n",
    "        Args:\n",
    "            question (str): The question to be classified.\n",
    "\n",
    "        Returns:\n",
    "            str: The classification result.\n",
    "        \"\"\"\n",
    "        classification_prompt = self.config['prompts']['classification_prompt']\n",
    "        prompt = classification_prompt.format(question=question)\n",
    "        messages = [\n",
    "            {\"role\": \"system\", \"content\": prompt},\n",
    "            {\"role\": \"user\", \"content\": question},\n",
    "        ]\n",
    "        client = OpenAI(api_key=self.api_key)\n",
    "        completions = client.chat.completions.create(\n",
    "            model=\"gpt-4o\",\n",
    "            messages=messages,\n",
    "            temperature=0,\n",
    "            max_tokens=10\n",
    "        )\n",
    "        return completions.choices[0].message.content.strip().lower()\n",
    "\n",
    "    def select_tool(self, question: str) -> Callable:\n",
    "        \"\"\"\n",
    "        Select the appropriate tool based on the classified question.\n",
    "\n",
    "        Args:\n",
    "            question (str): The question to be answered.\n",
    "\n",
    "        Returns:\n",
    "            Callable: The selected tool's generate_response method.\n",
    "        \"\"\"\n",
    "        category = self.classify_question(question)\n",
    "        if \"disease\" in category:\n",
    "            return self.disease_association_agent.generate_response\n",
    "        elif \"downstream\" in category:\n",
    "            return self.downstream_interaction_agent.generate_response\n",
    "        else:\n",
    "            raise ValueError(\"No appropriate tool found for the given question.\")\n",
    "\n",
    "    def ask(self, question: str) -> str:\n",
    "        \"\"\"\n",
    "        Generate a response for a given question by selecting the appropriate tool.\n",
    "\n",
    "        Args:\n",
    "            question (str): The question to be answered.\n",
    "\n",
    "        Returns:\n",
    "            str: The generated response.\n",
    "        \"\"\"\n",
    "        tool_func = self.select_tool(question)\n",
    "        return tool_func(question)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "No appropriate tool found for the given question.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[96], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m config_path \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mconfig.yaml\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      2\u001b[0m custom_agent \u001b[38;5;241m=\u001b[39m CustomAgent(uri, user, password, api_key,config_path)\n\u001b[0;32m----> 3\u001b[0m response \u001b[38;5;241m=\u001b[39m custom_agent\u001b[38;5;241m.\u001b[39mask(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mWhat are the benefits of good weather?\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28mprint\u001b[39m(response)\n",
      "Cell \u001b[0;32mIn[89], line 65\u001b[0m, in \u001b[0;36mCustomAgent.ask\u001b[0;34m(self, question)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mask\u001b[39m(\u001b[38;5;28mself\u001b[39m, question: \u001b[38;5;28mstr\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mstr\u001b[39m:\n\u001b[1;32m     56\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;124;03m    Generate a response for a given question by selecting the appropriate tool.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     63\u001b[0m \u001b[38;5;124;03m        str: The generated response.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 65\u001b[0m     tool_func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mselect_tool(question)\n\u001b[1;32m     66\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m tool_func(question)\n",
      "Cell \u001b[0;32mIn[89], line 53\u001b[0m, in \u001b[0;36mCustomAgent.select_tool\u001b[0;34m(self, question)\u001b[0m\n\u001b[1;32m     51\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdownstream_interaction_agent\u001b[38;5;241m.\u001b[39mgenerate_response\n\u001b[1;32m     52\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 53\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNo appropriate tool found for the given question.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: No appropriate tool found for the given question."
     ]
    }
   ],
   "source": [
    "config_path = 'config.yaml'\n",
    "custom_agent = CustomAgent(uri, user, password, api_key,config_path)\n",
    "response = custom_agent.ask(\"What are the benefits of good weather?\")\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "tool_selection_dataset = [\n",
    "    {\"question\": \"What disease is PRKN associated with?\", \"label\": \"disease_association\"},\n",
    "    {\"question\": \"What are the downstream interactions of gene Y?\", \"label\": \"downstream_interaction\"},\n",
    "    {\"question\": \"Is gene GNAI1 associated with Parkinson disease?\", \"label\": \"disease_association\"},\n",
    "    {\"question\": \"What genes interact with gene SNCA downstream?\", \"label\": \"downstream_interaction\"},\n",
    "    {\"question\": \"Who won the Euro Cup in 2024?\", \"label\": \"none\"},\n",
    "    {\"question\": \"Is gene Alpha-synuclein associated with Parkinson's?\", \"label\": \"disease_association\"},\n",
    "    {\"question\": \"What downstream genes are affected by gene Z?\", \"label\": \"downstream_interaction\"},\n",
    "    {\"question\": \"What is the capital of France?\", \"label\": \"none\"},\n",
    "    {\"question\": \"What is the weather like?\", \"label\": \"none\"},\n",
    "    {\"question\": \"What would happend if Insulin is activated in the Type II diabetes mellitus pathway?\", \"label\": \"downstream_interaction\"},\n",
    "    {\"question\": \"Are there any connections between Caspace-3 and Alzheimer Disease?\",\"label\": \"disease_association\"},\n",
    "    {\"question\": \"Which gene is linked to Alzheimer's disease?\", \"label\": \"disease_association\"},\n",
    "    {\"question\": \"What downstream effects does gene BCL2 have?\", \"label\": \"downstream_interaction\"},\n",
    "    {\"question\": \"Is BRCA1 associated with Colorectal cancer?\", \"label\": \"disease_association\"},\n",
    "    {\"question\": \"What pathways are activated by gene TP53?\", \"label\": \"downstream_interaction\"},\n",
    "    {\"question\": \"Is there a link between gene APP and Alzheimer's?\", \"label\": \"disease_association\"},\n",
    "    {\"question\": \"What downstream targets are influenced by gene MFN1?\", \"label\": \"downstream_interaction\"},\n",
    "    {\"question\": \"What environmental factors influence gene mutations?\", \"label\": \"none\"}\n",
    "]\n",
    "\n",
    "with open('evaluation_datasets/tool_selection_dataset.json', 'w') as f:\n",
    "    json.dump(tool_selection_dataset, f, indent=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'question': 'What disease is PRKN associated with?', 'label': 'disease_association'}, {'question': 'What are the downstream interactions of gene Y?', 'label': 'downstream_interaction'}, {'question': 'Is gene GNAI1 associated with Parkinson disease?', 'label': 'disease_association'}, {'question': 'What genes interact with gene SNCA downstream?', 'label': 'downstream_interaction'}, {'question': 'Who won the Euro Cup in 2024?', 'label': 'none'}, {'question': \"Is gene Alpha-synuclein associated with Parkinson's?\", 'label': 'disease_association'}, {'question': 'What downstream genes are affected by gene Z?', 'label': 'downstream_interaction'}, {'question': 'What is the capital of France?', 'label': 'none'}, {'question': 'What is the weather like?', 'label': 'none'}, {'question': 'What would happend if Insulin is activated in the Type II diabetes mellitus pathway?', 'label': 'downstream_interaction'}, {'question': 'Are there any connections between Caspace-3 and Alzheimer Disease?', 'label': 'disease_association'}, {'question': \"Which gene is linked to Alzheimer's disease?\", 'label': 'disease_association'}, {'question': 'What downstream effects does gene BCL2 have?', 'label': 'downstream_interaction'}, {'question': 'Is BRCA1 associated with Colorectal cancer?', 'label': 'disease_association'}, {'question': 'What pathways are activated by gene TP53?', 'label': 'downstream_interaction'}, {'question': \"Is there a link between gene APP and Alzheimer's?\", 'label': 'disease_association'}, {'question': 'What downstream targets are influenced by gene MFN1?', 'label': 'downstream_interaction'}, {'question': 'What environmental factors influence gene mutations?', 'label': 'none'}]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Load the dataset from a JSON file\n",
    "with open('evaluation_datasets/tool_selection_dataset.json', 'r') as f:\n",
    "    loaded_dataset = json.load(f)\n",
    "\n",
    "# Check if the dataset is loaded correctly\n",
    "print(loaded_dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy=0.9444444444444444\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>question</th>\n",
       "      <th>expected_label</th>\n",
       "      <th>predicted_label</th>\n",
       "      <th>is_correct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>What disease is PRKN associated with?</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>What are the downstream interactions of gene Y?</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Is gene GNAI1 associated with Parkinson disease?</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>What genes interact with gene SNCA downstream?</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Who won the Euro Cup in 2024?</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Is gene Alpha-synuclein associated with Parkin...</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>What downstream genes are affected by gene Z?</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>What is the capital of France?</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>What is the weather like?</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>What would happend if Insulin is activated in ...</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Are there any connections between Caspace-3 an...</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Which gene is linked to Alzheimer's disease?</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>What downstream effects does gene BCL2 have?</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Is BRCA1 associated with Colorectal cancer?</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>What pathways are activated by gene TP53?</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Is there a link between gene APP and Alzheimer's?</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>What downstream targets are influenced by gene...</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>downstream_interaction</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>What environmental factors influence gene muta...</td>\n",
       "      <td>none</td>\n",
       "      <td>disease_association</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             question          expected_label  \\\n",
       "0               What disease is PRKN associated with?     disease_association   \n",
       "1     What are the downstream interactions of gene Y?  downstream_interaction   \n",
       "2    Is gene GNAI1 associated with Parkinson disease?     disease_association   \n",
       "3      What genes interact with gene SNCA downstream?  downstream_interaction   \n",
       "4                       Who won the Euro Cup in 2024?                    none   \n",
       "5   Is gene Alpha-synuclein associated with Parkin...     disease_association   \n",
       "6       What downstream genes are affected by gene Z?  downstream_interaction   \n",
       "7                      What is the capital of France?                    none   \n",
       "8                           What is the weather like?                    none   \n",
       "9   What would happend if Insulin is activated in ...  downstream_interaction   \n",
       "10  Are there any connections between Caspace-3 an...     disease_association   \n",
       "11       Which gene is linked to Alzheimer's disease?     disease_association   \n",
       "12       What downstream effects does gene BCL2 have?  downstream_interaction   \n",
       "13        Is BRCA1 associated with Colorectal cancer?     disease_association   \n",
       "14          What pathways are activated by gene TP53?  downstream_interaction   \n",
       "15  Is there a link between gene APP and Alzheimer's?     disease_association   \n",
       "16  What downstream targets are influenced by gene...  downstream_interaction   \n",
       "17  What environmental factors influence gene muta...                    none   \n",
       "\n",
       "           predicted_label  is_correct  \n",
       "0      disease_association        True  \n",
       "1   downstream_interaction        True  \n",
       "2      disease_association        True  \n",
       "3   downstream_interaction        True  \n",
       "4                     none        True  \n",
       "5      disease_association        True  \n",
       "6   downstream_interaction        True  \n",
       "7                     none        True  \n",
       "8                     none        True  \n",
       "9   downstream_interaction        True  \n",
       "10     disease_association        True  \n",
       "11     disease_association        True  \n",
       "12  downstream_interaction        True  \n",
       "13     disease_association        True  \n",
       "14  downstream_interaction        True  \n",
       "15     disease_association        True  \n",
       "16  downstream_interaction        True  \n",
       "17     disease_association       False  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from typing import List, Dict\n",
    "import pandas as pd\n",
    "def evaluate_tool_selection(agent: CustomAgent, dataset: List[Dict[str, str]]) -> float:\n",
    "    \"\"\"\n",
    "    Evaluate the tool selection performance of the CustomAgent.\n",
    "\n",
    "    Args:\n",
    "        agent (CustomAgent): The CustomAgent to be evaluated.\n",
    "        dataset (List[Dict[str, str]]): The labeled dataset containing questions and expected tool labels.\n",
    "\n",
    "    Returns:\n",
    "        float: The accuracy of the tool selection.\n",
    "    \"\"\"\n",
    "    correct_count = 0\n",
    "    results = []\n",
    "\n",
    "    for data in dataset:\n",
    "        question = data[\"question\"]\n",
    "        label = data[\"label\"]\n",
    "        \n",
    "        try:\n",
    "            selected_tool_func = agent.select_tool(question)\n",
    "            selected_tool = \"disease_association\" if selected_tool_func == agent.disease_association_agent.generate_response else \"downstream_interaction\"\n",
    "        except ValueError:\n",
    "            selected_tool = \"none\"\n",
    "        \n",
    "        # Determine if the tool selection was correct\n",
    "        is_correct = (selected_tool == label)\n",
    "        if is_correct:\n",
    "            correct_count += 1\n",
    "\n",
    "        # Append the result to the list\n",
    "        results.append({\n",
    "            \"question\": question,\n",
    "            \"expected_label\": label,\n",
    "            \"predicted_label\": selected_tool,\n",
    "            \"is_correct\": is_correct\n",
    "        })\n",
    "\n",
    "    accuracy = correct_count / len(dataset)\n",
    "    print(f\"Accuracy={accuracy}\")\n",
    "    df = pd.DataFrame(results)\n",
    "\n",
    "    return df\n",
    "\n",
    "# Usage\n",
    "config_path = 'config.yaml'\n",
    "custom_agent = CustomAgent(uri, user, password, api_key, config_path)\n",
    "evaluate_tool_selection(custom_agent, loaded_dataset)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "# Sample questions and their corresponding expected Cypher queries\n",
    "downstream_interaction_examples = [\n",
    "    {\n",
    "        \"question\": \"Can you give me the paths downstream of Growth factor receptor-bound protein 2 in the Colorectal cancer?\",\n",
    "        \"expected_cypher_query\": \"\"\"\n",
    "            MATCH (start: Gene)-[:ASSOCIATED_WITH]->(d:Disease {name: 'Colorectal cancer'})\n",
    "            WHERE \"Growth factor receptor-bound protein 2\" IN start.names\n",
    "            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \n",
    "            YIELD path\n",
    "            WITH path\n",
    "            WHERE NOT EXISTS {\n",
    "                MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\n",
    "                WHERE lastNode = last(nodes(path))\n",
    "            }\n",
    "            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\n",
    "            RETURN collect(relationships) AS interactions\n",
    "        \"\"\"\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What effects should I expect downstream of 'MCH5' in the Alzheimer pathway?\",\n",
    "        \"expected_cypher_query\": \"\"\"\n",
    "            MATCH (start: Gene)-[:ASSOCIATED_WITH]->(d:Disease {name: 'Alzheimer disease'})\n",
    "            WHERE \"MCH5\" IN start.synonyms\n",
    "            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \n",
    "            YIELD path\n",
    "            WITH path\n",
    "            WHERE NOT EXISTS {\n",
    "                MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\n",
    "                WHERE lastNode = last(nodes(path))\n",
    "            }\n",
    "            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\n",
    "            RETURN collect(relationships) AS interactions\n",
    "        \"\"\"\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What will happen if Insulin is activated in the Type II diabetes mellitus pathway?\",\n",
    "        \"expected_cypher_query\": \"\"\"\n",
    "            MATCH (start: Gene)-[:ASSOCIATED_WITH]->(d:Disease {name: 'Type II diabetes mellitus'})\n",
    "            WHERE \"Insulin\" IN start.names\n",
    "            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \n",
    "            YIELD path\n",
    "            WITH path\n",
    "            WHERE NOT EXISTS {\n",
    "                MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\n",
    "                WHERE lastNode = last(nodes(path))\n",
    "            }\n",
    "            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\n",
    "            RETURN collect(relationships) AS interactions\n",
    "        \"\"\"\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What are the downstream interactions of gene PARK7?\",\n",
    "        \"expected_cypher_query\": \"\"\"\n",
    "            MATCH (start: Gene)\n",
    "            WHERE ('PARK7' IN start.names) OR ('PARK7' IN start.synonyms)\n",
    "            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>', minLevel: 1, uniqueness: 'NODE_PATH', bfs: false}) \n",
    "            YIELD path\n",
    "            WITH path\n",
    "            WHERE NOT EXISTS {\n",
    "                MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\n",
    "                WHERE lastNode = last(nodes(path))\n",
    "            }\n",
    "            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\n",
    "            RETURN collect(relationships) AS interactions\n",
    "        \"\"\"\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What downstream genes are affected by Kelch-like ECH-associated protein 1?\",\n",
    "        \"expected_cypher_query\": \"\"\"\n",
    "            MATCH (start: Gene)\n",
    "            WHERE 'Kelch-like ECH-associated protein 1' IN start.names\n",
    "            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>', minLevel: 1, uniqueness: 'NODE_PATH', bfs: false}) \n",
    "            YIELD path\n",
    "            WITH path\n",
    "            WHERE NOT EXISTS {\n",
    "                MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\n",
    "                WHERE lastNode = last(nodes(path))\n",
    "            }\n",
    "            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\n",
    "            RETURN collect(relationships) AS interactions\n",
    "        \"\"\"\n",
    "    }\n",
    "]\n",
    "\n",
    "# Save the dataset to a JSON file\n",
    "with open('downstream_interaction_dataset.json', 'w') as f:\n",
    "    json.dump(downstream_interaction_examples, f, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "def evaluate_run_cypher_query(agent: CustomAgent, dataset: list, context: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Evaluate the run_cypher_query method of the CustomAgent.\n",
    "\n",
    "    Args:\n",
    "        agent (CustomAgent): The CustomAgent to be tested.\n",
    "        dataset (list): List of questions and their expected Cypher queries.\n",
    "        context (str): The prompt context to set for the agent.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: DataFrame containing the questions, generated Cypher queries, expected Cypher queries,\n",
    "                      agent results, manual results, and whether the results match.\n",
    "    \"\"\"\n",
    "    results = []\n",
    "\n",
    "    for item in dataset:\n",
    "        question = item[\"question\"]\n",
    "        expected_cypher_query = item[\"expected_cypher_query\"].strip()\n",
    "\n",
    "        # Set the appropriate context\n",
    "        agent.set_prompt_context(context)\n",
    "\n",
    "        # Generate Cypher query using the agent\n",
    "        generated_cypher_query = agent.generate(agent.current_prompts['initial_prompt'], schema=agent.schema, question=question).strip()\n",
    "\n",
    "        # Run the Cypher query using run_cypher_query method\n",
    "        agent_result = agent.run_cypher_query(generated_cypher_query)\n",
    "\n",
    "        # Run the expected Cypher query manually using kg.query\n",
    "        manual_result = agent.kg.query(expected_cypher_query)\n",
    "\n",
    "        # Compare the results\n",
    "        results.append({\n",
    "            \"question\": question,\n",
    "            \"generated_cypher_query\": generated_cypher_query,\n",
    "            \"expected_cypher_query\": expected_cypher_query,\n",
    "            \"agent_result\": json.dumps(agent_result, indent=4),\n",
    "            \"manual_result\": json.dumps(manual_result, indent=4),\n",
    "            \"match\": agent_result == manual_result\n",
    "        })\n",
    "\n",
    "    # Create a DataFrame from the results\n",
    "    df = pd.DataFrame(results)\n",
    "    return df\n",
    "\n",
    "# Load the downstream interaction dataset from a JSON file\n",
    "with open('evaluation_datasets/downstream_interaction_dataset.json', 'r') as f:\n",
    "    downstream_dataset = json.load(f)\n",
    "\n",
    "# Example usage for downstream interactions\n",
    "config_path = 'config.yaml'\n",
    "custom_agent = CustomAgent(uri, user, password, api_key, config_path)\n",
    "downstream_evaluation_df = evaluate_run_cypher_query(custom_agent, downstream_dataset, 'downstream_interaction')\n",
    "\n",
    "# Set pandas display options to show all content\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.expand_frame_repr', False)\n",
    "\n",
    "# Print the DataFrame\n",
    "print(\"Downstream Interaction Evaluation:\")\n",
    "downstream_evaluation_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>question</th>\n",
       "      <th>generated_cypher_query</th>\n",
       "      <th>expected_cypher_query</th>\n",
       "      <th>agent_result</th>\n",
       "      <th>manual_result</th>\n",
       "      <th>match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Can you give me the paths downstream of Growth factor receptor-bound protein 2 in the Colorectal cancer?</td>\n",
       "      <td>MATCH (start: Gene)-[:ASSOCIATED_WITH]-&gt;(d:Disease {name: 'Colorectal cancer'})\\nWHERE ('Growth factor receptor-bound protein 2' IN start.names) or ('Growth factor receptor-bound protein 2' IN start.synonyms)\\nCALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH&gt;',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\nYIELD path\\nWITH path\\nWHERE NOT EXISTS {\\n    MATCH (lastNode)-[:INTERACTS_WITH]-&gt;(:Gene)\\n    WHERE lastNode = last(nodes(path))\\n}\\nWITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\nRETURN collect(relationships) AS interactions</td>\n",
       "      <td>MATCH (start: Gene)-[:ASSOCIATED_WITH]-&gt;(d:Disease {name: 'Colorectal cancer'})\\n            WHERE \"Growth factor receptor-bound protein 2\" IN start.names\\n            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH&gt;',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\n            YIELD path\\n            WITH path\\n            WHERE NOT EXISTS {\\n                MATCH (lastNode)-[:INTERACTS_WITH]-&gt;(:Gene)\\n                WHERE lastNode = last(nodes(path))\\n            }\\n            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\n            RETURN collect(relationships) AS interactions</td>\n",
       "      <td>[\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_155\",\\n                        \"names\": [\\n                            \"Growth factor receptor-bound protein 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ASH\",\\n                            \"GRB2\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_154\",\\n                        \"names\": [\\n                            \"Hepatocyte growth factor\",\\n                            \"Son of sevenless homolog 1\",\\n                            \"Erythroid transcription factor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"SOS1\",\\n                            \"GF1\",\\n                            \"HGF\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_154\",\\n                        \"names\": [\\n                            \"Hepatocyte growth factor\",\\n                            \"Son of sevenless homolog 1\",\\n                            \"Erythroid transcription factor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"SOS1\",\\n                            \"GF1\",\\n                            \"HGF\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_152\",\\n                        \"names\": [\\n                            \"GTPase HRas\",\\n                            \"GTPase NRas\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"HRAS1\",\\n                            \"HRAS\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_152\",\\n                        \"names\": [\\n                            \"GTPase HRas\",\\n                            \"GTPase NRas\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"HRAS1\",\\n                            \"HRAS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_162\",\\n                        \"names\": [\\n                            \"Serine/threonine-protein kinase A-Raf\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ARAF\",\\n                            \"ARAF1\",\\n                            \"PKS2\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\",\\n                        \"phosphorylation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_162\",\\n                        \"names\": [\\n                            \"Serine/threonine-protein kinase A-Raf\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ARAF\",\\n                            \"ARAF1\",\\n                            \"PKS2\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_161\",\\n                        \"names\": [\\n                            \"Dual specificity mitogen-activated protein kinase kinase 1\",\\n                            \"Ras-related protein Rab-8A\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PRKMK1\",\\n                            \"MAP2K1\",\\n                            \"MEK1\",\\n                            \"MEL\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\",\\n                        \"phosphorylation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_161\",\\n                        \"names\": [\\n                            \"Dual specificity mitogen-activated protein kinase kinase 1\",\\n                            \"Ras-related protein Rab-8A\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PRKMK1\",\\n                            \"MAP2K1\",\\n                            \"MEK1\",\\n                            \"MEL\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_160\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]</td>\n",
       "      <td>[\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_155\",\\n                        \"names\": [\\n                            \"Growth factor receptor-bound protein 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ASH\",\\n                            \"GRB2\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_154\",\\n                        \"names\": [\\n                            \"Hepatocyte growth factor\",\\n                            \"Son of sevenless homolog 1\",\\n                            \"Erythroid transcription factor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"SOS1\",\\n                            \"GF1\",\\n                            \"HGF\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_154\",\\n                        \"names\": [\\n                            \"Hepatocyte growth factor\",\\n                            \"Son of sevenless homolog 1\",\\n                            \"Erythroid transcription factor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"SOS1\",\\n                            \"GF1\",\\n                            \"HGF\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_152\",\\n                        \"names\": [\\n                            \"GTPase HRas\",\\n                            \"GTPase NRas\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"HRAS1\",\\n                            \"HRAS\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_152\",\\n                        \"names\": [\\n                            \"GTPase HRas\",\\n                            \"GTPase NRas\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"HRAS1\",\\n                            \"HRAS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_162\",\\n                        \"names\": [\\n                            \"Serine/threonine-protein kinase A-Raf\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ARAF\",\\n                            \"ARAF1\",\\n                            \"PKS2\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\",\\n                        \"phosphorylation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_162\",\\n                        \"names\": [\\n                            \"Serine/threonine-protein kinase A-Raf\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ARAF\",\\n                            \"ARAF1\",\\n                            \"PKS2\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_161\",\\n                        \"names\": [\\n                            \"Dual specificity mitogen-activated protein kinase kinase 1\",\\n                            \"Ras-related protein Rab-8A\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PRKMK1\",\\n                            \"MAP2K1\",\\n                            \"MEK1\",\\n                            \"MEL\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\",\\n                        \"phosphorylation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_161\",\\n                        \"names\": [\\n                            \"Dual specificity mitogen-activated protein kinase kinase 1\",\\n                            \"Ras-related protein Rab-8A\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PRKMK1\",\\n                            \"MAP2K1\",\\n                            \"MEK1\",\\n                            \"MEL\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_160\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>What effects should I expect downstream of 'MCH5' in the Alzheimer pathway?</td>\n",
       "      <td>MATCH (start: Gene)-[:ASSOCIATED_WITH]-&gt;(d:Disease {name: 'Alzheimer disease'})\\nWHERE 'MCH5' IN start.synonyms\\nCALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH&gt;',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\nYIELD path\\nWITH path\\nWHERE NOT EXISTS {\\n    MATCH (lastNode)-[:INTERACTS_WITH]-&gt;(:Gene)\\n    WHERE lastNode = last(nodes(path))\\n}\\nWITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\nRETURN collect(relationships) AS interactions</td>\n",
       "      <td>MATCH (start: Gene)-[:ASSOCIATED_WITH]-&gt;(d:Disease {name: 'Alzheimer disease'})\\n            WHERE \"MCH5\" IN start.synonyms\\n            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH&gt;',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\n            YIELD path\\n            WITH path\\n            WHERE NOT EXISTS {\\n                MATCH (lastNode)-[:INTERACTS_WITH]-&gt;(:Gene)\\n                WHERE lastNode = last(nodes(path))\\n            }\\n            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\n            RETURN collect(relationships) AS interactions</td>\n",
       "      <td>[\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05010_87\",\\n                        \"names\": [\\n                            \"Caspase-8\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"MCH5\",\\n                            \"CASP8\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05010_88\",\\n                        \"names\": [\\n                            \"BH3-interacting domain death agonist\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BID\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05010_87\",\\n                        \"names\": [\\n                            \"Caspase-8\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"MCH5\",\\n                            \"CASP8\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05010_34\",\\n                        \"names\": [\\n                            \"Caspase-3\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CPP32\",\\n                            \"CASP3\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]</td>\n",
       "      <td>[\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05010_87\",\\n                        \"names\": [\\n                            \"Caspase-8\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"MCH5\",\\n                            \"CASP8\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05010_88\",\\n                        \"names\": [\\n                            \"BH3-interacting domain death agonist\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BID\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05010_87\",\\n                        \"names\": [\\n                            \"Caspase-8\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"MCH5\",\\n                            \"CASP8\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05010_34\",\\n                        \"names\": [\\n                            \"Caspase-3\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CPP32\",\\n                            \"CASP3\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>What will happen if Insulin is activated in the Type II diabetes mellitus pathway?</td>\n",
       "      <td>MATCH (start: Gene)-[:ASSOCIATED_WITH]-&gt;(d:Disease {name: 'Type II diabetes mellitus'})\\nWHERE \"Insulin\" IN start.names\\nCALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH&gt;',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\nYIELD path\\nWITH path\\nWHERE NOT EXISTS {\\n    MATCH (lastNode)-[:INTERACTS_WITH]-&gt;(:Gene)\\n    WHERE lastNode = last(nodes(path))\\n}\\nWITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\nRETURN collect(relationships) AS interactions</td>\n",
       "      <td>MATCH (start: Gene)-[:ASSOCIATED_WITH]-&gt;(d:Disease {name: 'Type II diabetes mellitus'})\\n            WHERE \"Insulin\" IN start.names\\n            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH&gt;',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\n            YIELD path\\n            WITH path\\n            WHERE NOT EXISTS {\\n                MATCH (lastNode)-[:INTERACTS_WITH]-&gt;(:Gene)\\n                WHERE lastNode = last(nodes(path))\\n            }\\n            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\n            RETURN collect(relationships) AS interactions</td>\n",
       "      <td>[\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_15\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"indirect effect\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_23\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"indirect effect\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_23\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_31\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_31\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]</td>\n",
       "      <td>[\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_15\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"indirect effect\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_23\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"indirect effect\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_23\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_31\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_31\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>What are the downstream interactions of gene PARK7?</td>\n",
       "      <td>MATCH (start: Gene)\\nWHERE ('PARK7' IN start.names) OR ('PARK7' IN start.synonyms)\\nCALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH&gt;', minLevel: 1, uniqueness: 'NODE_PATH', bfs: false}) \\nYIELD path\\nWITH path\\nWHERE NOT EXISTS {\\n    MATCH (lastNode)-[:INTERACTS_WITH]-&gt;(:Gene)\\n    WHERE lastNode = last(nodes(path))\\n}\\nWITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\nRETURN collect(relationships) AS interactions</td>\n",
       "      <td>MATCH (start: Gene)\\n            WHERE ('PARK7' IN start.names) OR ('PARK7' IN start.synonyms)\\n            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH&gt;', minLevel: 1, uniqueness: 'NODE_PATH', bfs: false}) \\n            YIELD path\\n            WITH path\\n            WHERE NOT EXISTS {\\n                MATCH (lastNode)-[:INTERACTS_WITH]-&gt;(:Gene)\\n                WHERE lastNode = last(nodes(path))\\n            }\\n            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\n            RETURN collect(relationships) AS interactions</td>\n",
       "      <td>[\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_19\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_362\",\\n                        \"names\": [\\n                            \"Thioredoxin\",\\n                            \"Thioredoxin, mitochondrial\",\\n                            \"Histone-lysine N-methyltransferase 2B\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"TRX2\",\\n                            \"TXN\",\\n                            \"TXN2\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_19\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_338\",\\n                        \"names\": [\\n                            \"Cellular tumor antigen p53\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"P53\",\\n                            \"TP53\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_338\",\\n                        \"names\": [\\n                            \"Cellular tumor antigen p53\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"P53\",\\n                            \"TP53\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_361\",\\n                        \"names\": [\\n                            \"Dual specificity protein phosphatase 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PTPN10\",\\n                            \"CL100\",\\n                            \"MKP1\",\\n                            \"DUSP1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_326\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_327\",\\n                        \"names\": [\\n                            \"Bcl-2-like protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BCL2L\",\\n                            \"BCL2L1\",\\n                            \"BCLX\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_327\",\\n                        \"names\": [\\n                            \"Bcl-2-like protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BCL2L\",\\n                            \"BCL2L1\",\\n                            \"BCLX\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_316\",\\n                        \"names\": [\\n                            \"Apoptosis regulator BAX\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BAX\",\\n                            \"BCL2L4\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_316\",\\n                        \"names\": [\\n                            \"Apoptosis regulator BAX\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BAX\",\\n                            \"BCL2L4\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_25\",\\n                        \"names\": [\\n                            \"Cytochrome c\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CYC\",\\n                            \"CYCS\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_25\",\\n                        \"names\": [\\n                            \"Cytochrome c\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CYC\",\\n                            \"CYCS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_152\",\\n                        \"names\": [\\n                            \"Caspase-9\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CASP9\",\\n                            \"MCH6\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_152\",\\n                        \"names\": [\\n                            \"Caspase-9\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CASP9\",\\n                            \"MCH6\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_153\",\\n                        \"names\": [\\n                            \"Caspase-3\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CPP32\",\\n                            \"CASP3\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]</td>\n",
       "      <td>[\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_19\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_362\",\\n                        \"names\": [\\n                            \"Thioredoxin\",\\n                            \"Thioredoxin, mitochondrial\",\\n                            \"Histone-lysine N-methyltransferase 2B\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"TRX2\",\\n                            \"TXN\",\\n                            \"TXN2\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_19\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_338\",\\n                        \"names\": [\\n                            \"Cellular tumor antigen p53\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"P53\",\\n                            \"TP53\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_338\",\\n                        \"names\": [\\n                            \"Cellular tumor antigen p53\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"P53\",\\n                            \"TP53\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_361\",\\n                        \"names\": [\\n                            \"Dual specificity protein phosphatase 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PTPN10\",\\n                            \"CL100\",\\n                            \"MKP1\",\\n                            \"DUSP1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_326\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_327\",\\n                        \"names\": [\\n                            \"Bcl-2-like protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BCL2L\",\\n                            \"BCL2L1\",\\n                            \"BCLX\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_327\",\\n                        \"names\": [\\n                            \"Bcl-2-like protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BCL2L\",\\n                            \"BCL2L1\",\\n                            \"BCLX\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_316\",\\n                        \"names\": [\\n                            \"Apoptosis regulator BAX\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BAX\",\\n                            \"BCL2L4\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_316\",\\n                        \"names\": [\\n                            \"Apoptosis regulator BAX\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BAX\",\\n                            \"BCL2L4\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_25\",\\n                        \"names\": [\\n                            \"Cytochrome c\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CYC\",\\n                            \"CYCS\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_25\",\\n                        \"names\": [\\n                            \"Cytochrome c\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CYC\",\\n                            \"CYCS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_152\",\\n                        \"names\": [\\n                            \"Caspase-9\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CASP9\",\\n                            \"MCH6\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_152\",\\n                        \"names\": [\\n                            \"Caspase-9\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CASP9\",\\n                            \"MCH6\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_153\",\\n                        \"names\": [\\n                            \"Caspase-3\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CPP32\",\\n                            \"CASP3\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>What downstream genes are affected by Kelch-like ECH-associated protein 1?</td>\n",
       "      <td>MATCH (start: Gene)\\nWHERE ('Kelch-like ECH-associated protein 1' IN start.names) OR ('Kelch-like ECH-associated protein 1' IN start.synonyms)\\nCALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH&gt;', minLevel: 1, uniqueness: 'NODE_PATH', bfs: false}) \\nYIELD path\\nWITH path\\nWHERE NOT EXISTS {\\n    MATCH (lastNode)-[:INTERACTS_WITH]-&gt;(:Gene)\\n    WHERE lastNode = last(nodes(path))\\n}\\nWITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\nRETURN collect(relationships) AS interactions</td>\n",
       "      <td>MATCH (start: Gene)\\n            WHERE 'Kelch-like ECH-associated protein 1' IN start.names\\n            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH&gt;', minLevel: 1, uniqueness: 'NODE_PATH', bfs: false}) \\n            YIELD path\\n            WITH path\\n            WHERE NOT EXISTS {\\n                MATCH (lastNode)-[:INTERACTS_WITH]-&gt;(:Gene)\\n                WHERE lastNode = last(nodes(path))\\n            }\\n            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\n            RETURN collect(relationships) AS interactions</td>\n",
       "      <td>[\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"inhibition\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_714\",\\n                        \"names\": [\\n                            \"Kelch-like ECH-associated protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"KLHL19\",\\n                            \"KEAP1\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_362\",\\n                        \"names\": [\\n                            \"Thioredoxin\",\\n                            \"Thioredoxin, mitochondrial\",\\n                            \"Histone-lysine N-methyltransferase 2B\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"TRX2\",\\n                            \"TXN\",\\n                            \"TXN2\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]</td>\n",
       "      <td>[\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"inhibition\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_714\",\\n                        \"names\": [\\n                            \"Kelch-like ECH-associated protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"KLHL19\",\\n                            \"KEAP1\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_362\",\\n                        \"names\": [\\n                            \"Thioredoxin\",\\n                            \"Thioredoxin, mitochondrial\",\\n                            \"Histone-lysine N-methyltransferase 2B\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"TRX2\",\\n                            \"TXN\",\\n                            \"TXN2\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                   question                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               generated_cypher_query                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      expected_cypher_query                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  agent_result                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 manual_result  match\n",
       "0  Can you give me the paths downstream of Growth factor receptor-bound protein 2 in the Colorectal cancer?  MATCH (start: Gene)-[:ASSOCIATED_WITH]->(d:Disease {name: 'Colorectal cancer'})\\nWHERE ('Growth factor receptor-bound protein 2' IN start.names) or ('Growth factor receptor-bound protein 2' IN start.synonyms)\\nCALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\nYIELD path\\nWITH path\\nWHERE NOT EXISTS {\\n    MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\\n    WHERE lastNode = last(nodes(path))\\n}\\nWITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\nRETURN collect(relationships) AS interactions  MATCH (start: Gene)-[:ASSOCIATED_WITH]->(d:Disease {name: 'Colorectal cancer'})\\n            WHERE \"Growth factor receptor-bound protein 2\" IN start.names\\n            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\n            YIELD path\\n            WITH path\\n            WHERE NOT EXISTS {\\n                MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\\n                WHERE lastNode = last(nodes(path))\\n            }\\n            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\n            RETURN collect(relationships) AS interactions                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         [\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_155\",\\n                        \"names\": [\\n                            \"Growth factor receptor-bound protein 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ASH\",\\n                            \"GRB2\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_154\",\\n                        \"names\": [\\n                            \"Hepatocyte growth factor\",\\n                            \"Son of sevenless homolog 1\",\\n                            \"Erythroid transcription factor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"SOS1\",\\n                            \"GF1\",\\n                            \"HGF\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_154\",\\n                        \"names\": [\\n                            \"Hepatocyte growth factor\",\\n                            \"Son of sevenless homolog 1\",\\n                            \"Erythroid transcription factor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"SOS1\",\\n                            \"GF1\",\\n                            \"HGF\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_152\",\\n                        \"names\": [\\n                            \"GTPase HRas\",\\n                            \"GTPase NRas\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"HRAS1\",\\n                            \"HRAS\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_152\",\\n                        \"names\": [\\n                            \"GTPase HRas\",\\n                            \"GTPase NRas\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"HRAS1\",\\n                            \"HRAS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_162\",\\n                        \"names\": [\\n                            \"Serine/threonine-protein kinase A-Raf\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ARAF\",\\n                            \"ARAF1\",\\n                            \"PKS2\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\",\\n                        \"phosphorylation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_162\",\\n                        \"names\": [\\n                            \"Serine/threonine-protein kinase A-Raf\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ARAF\",\\n                            \"ARAF1\",\\n                            \"PKS2\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_161\",\\n                        \"names\": [\\n                            \"Dual specificity mitogen-activated protein kinase kinase 1\",\\n                            \"Ras-related protein Rab-8A\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PRKMK1\",\\n                            \"MAP2K1\",\\n                            \"MEK1\",\\n                            \"MEL\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\",\\n                        \"phosphorylation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_161\",\\n                        \"names\": [\\n                            \"Dual specificity mitogen-activated protein kinase kinase 1\",\\n                            \"Ras-related protein Rab-8A\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PRKMK1\",\\n                            \"MAP2K1\",\\n                            \"MEK1\",\\n                            \"MEL\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_160\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         [\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_155\",\\n                        \"names\": [\\n                            \"Growth factor receptor-bound protein 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ASH\",\\n                            \"GRB2\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_154\",\\n                        \"names\": [\\n                            \"Hepatocyte growth factor\",\\n                            \"Son of sevenless homolog 1\",\\n                            \"Erythroid transcription factor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"SOS1\",\\n                            \"GF1\",\\n                            \"HGF\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_154\",\\n                        \"names\": [\\n                            \"Hepatocyte growth factor\",\\n                            \"Son of sevenless homolog 1\",\\n                            \"Erythroid transcription factor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"SOS1\",\\n                            \"GF1\",\\n                            \"HGF\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_152\",\\n                        \"names\": [\\n                            \"GTPase HRas\",\\n                            \"GTPase NRas\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"HRAS1\",\\n                            \"HRAS\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_152\",\\n                        \"names\": [\\n                            \"GTPase HRas\",\\n                            \"GTPase NRas\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"HRAS1\",\\n                            \"HRAS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_162\",\\n                        \"names\": [\\n                            \"Serine/threonine-protein kinase A-Raf\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ARAF\",\\n                            \"ARAF1\",\\n                            \"PKS2\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\",\\n                        \"phosphorylation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_162\",\\n                        \"names\": [\\n                            \"Serine/threonine-protein kinase A-Raf\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ARAF\",\\n                            \"ARAF1\",\\n                            \"PKS2\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_161\",\\n                        \"names\": [\\n                            \"Dual specificity mitogen-activated protein kinase kinase 1\",\\n                            \"Ras-related protein Rab-8A\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PRKMK1\",\\n                            \"MAP2K1\",\\n                            \"MEK1\",\\n                            \"MEL\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\",\\n                        \"phosphorylation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05210_161\",\\n                        \"names\": [\\n                            \"Dual specificity mitogen-activated protein kinase kinase 1\",\\n                            \"Ras-related protein Rab-8A\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PRKMK1\",\\n                            \"MAP2K1\",\\n                            \"MEK1\",\\n                            \"MEL\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05210_160\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]   True\n",
       "1                               What effects should I expect downstream of 'MCH5' in the Alzheimer pathway?                                                                                                   MATCH (start: Gene)-[:ASSOCIATED_WITH]->(d:Disease {name: 'Alzheimer disease'})\\nWHERE 'MCH5' IN start.synonyms\\nCALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\nYIELD path\\nWITH path\\nWHERE NOT EXISTS {\\n    MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\\n    WHERE lastNode = last(nodes(path))\\n}\\nWITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\nRETURN collect(relationships) AS interactions                                 MATCH (start: Gene)-[:ASSOCIATED_WITH]->(d:Disease {name: 'Alzheimer disease'})\\n            WHERE \"MCH5\" IN start.synonyms\\n            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\n            YIELD path\\n            WITH path\\n            WHERE NOT EXISTS {\\n                MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\\n                WHERE lastNode = last(nodes(path))\\n            }\\n            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\n            RETURN collect(relationships) AS interactions                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    [\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05010_87\",\\n                        \"names\": [\\n                            \"Caspase-8\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"MCH5\",\\n                            \"CASP8\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05010_88\",\\n                        \"names\": [\\n                            \"BH3-interacting domain death agonist\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BID\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05010_87\",\\n                        \"names\": [\\n                            \"Caspase-8\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"MCH5\",\\n                            \"CASP8\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05010_34\",\\n                        \"names\": [\\n                            \"Caspase-3\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CPP32\",\\n                            \"CASP3\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    [\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05010_87\",\\n                        \"names\": [\\n                            \"Caspase-8\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"MCH5\",\\n                            \"CASP8\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05010_88\",\\n                        \"names\": [\\n                            \"BH3-interacting domain death agonist\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BID\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05010_87\",\\n                        \"names\": [\\n                            \"Caspase-8\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"MCH5\",\\n                            \"CASP8\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05010_34\",\\n                        \"names\": [\\n                            \"Caspase-3\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CPP32\",\\n                            \"CASP3\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]   True\n",
       "2                        What will happen if Insulin is activated in the Type II diabetes mellitus pathway?                                                                                           MATCH (start: Gene)-[:ASSOCIATED_WITH]->(d:Disease {name: 'Type II diabetes mellitus'})\\nWHERE \"Insulin\" IN start.names\\nCALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\nYIELD path\\nWITH path\\nWHERE NOT EXISTS {\\n    MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\\n    WHERE lastNode = last(nodes(path))\\n}\\nWITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\nRETURN collect(relationships) AS interactions                         MATCH (start: Gene)-[:ASSOCIATED_WITH]->(d:Disease {name: 'Type II diabetes mellitus'})\\n            WHERE \"Insulin\" IN start.names\\n            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>',minLevel: 1, uniqueness: 'NODE_PATH',bfs: false}) \\n            YIELD path\\n            WITH path\\n            WHERE NOT EXISTS {\\n                MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\\n                WHERE lastNode = last(nodes(path))\\n            }\\n            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\n            RETURN collect(relationships) AS interactions                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              [\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_15\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"indirect effect\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_23\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"indirect effect\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_23\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_31\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_31\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              [\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_15\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"indirect effect\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_23\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"indirect effect\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_23\",\\n                        \"names\": [\\n                            \"Mitogen-activated protein kinase 1\",\\n                            \"Ephrin type-B receptor 2\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"ERK\",\\n                            \"ERK2\",\\n                            \"MAPK1\",\\n                            \"PRKM2\",\\n                            \"PRKM1\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_31\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_17\",\\n                        \"names\": [\\n                            \"Insulin\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"04930_16\",\\n                        \"names\": [\\n                            \"Insulin receptor\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"INSR\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"04930_31\",\\n                        \"names\": [\\n                            \"Insulin receptor substrate 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"IRS1\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]   True\n",
       "3                                                       What are the downstream interactions of gene PARK7?                                                                                                                              MATCH (start: Gene)\\nWHERE ('PARK7' IN start.names) OR ('PARK7' IN start.synonyms)\\nCALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>', minLevel: 1, uniqueness: 'NODE_PATH', bfs: false}) \\nYIELD path\\nWITH path\\nWHERE NOT EXISTS {\\n    MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\\n    WHERE lastNode = last(nodes(path))\\n}\\nWITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\nRETURN collect(relationships) AS interactions                                                            MATCH (start: Gene)\\n            WHERE ('PARK7' IN start.names) OR ('PARK7' IN start.synonyms)\\n            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>', minLevel: 1, uniqueness: 'NODE_PATH', bfs: false}) \\n            YIELD path\\n            WITH path\\n            WHERE NOT EXISTS {\\n                MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\\n                WHERE lastNode = last(nodes(path))\\n            }\\n            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\n            RETURN collect(relationships) AS interactions  [\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_19\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_362\",\\n                        \"names\": [\\n                            \"Thioredoxin\",\\n                            \"Thioredoxin, mitochondrial\",\\n                            \"Histone-lysine N-methyltransferase 2B\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"TRX2\",\\n                            \"TXN\",\\n                            \"TXN2\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_19\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_338\",\\n                        \"names\": [\\n                            \"Cellular tumor antigen p53\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"P53\",\\n                            \"TP53\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_338\",\\n                        \"names\": [\\n                            \"Cellular tumor antigen p53\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"P53\",\\n                            \"TP53\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_361\",\\n                        \"names\": [\\n                            \"Dual specificity protein phosphatase 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PTPN10\",\\n                            \"CL100\",\\n                            \"MKP1\",\\n                            \"DUSP1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_326\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_327\",\\n                        \"names\": [\\n                            \"Bcl-2-like protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BCL2L\",\\n                            \"BCL2L1\",\\n                            \"BCLX\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_327\",\\n                        \"names\": [\\n                            \"Bcl-2-like protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BCL2L\",\\n                            \"BCL2L1\",\\n                            \"BCLX\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_316\",\\n                        \"names\": [\\n                            \"Apoptosis regulator BAX\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BAX\",\\n                            \"BCL2L4\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_316\",\\n                        \"names\": [\\n                            \"Apoptosis regulator BAX\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BAX\",\\n                            \"BCL2L4\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_25\",\\n                        \"names\": [\\n                            \"Cytochrome c\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CYC\",\\n                            \"CYCS\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_25\",\\n                        \"names\": [\\n                            \"Cytochrome c\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CYC\",\\n                            \"CYCS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_152\",\\n                        \"names\": [\\n                            \"Caspase-9\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CASP9\",\\n                            \"MCH6\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_152\",\\n                        \"names\": [\\n                            \"Caspase-9\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CASP9\",\\n                            \"MCH6\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_153\",\\n                        \"names\": [\\n                            \"Caspase-3\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CPP32\",\\n                            \"CASP3\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]  [\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_19\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_362\",\\n                        \"names\": [\\n                            \"Thioredoxin\",\\n                            \"Thioredoxin, mitochondrial\",\\n                            \"Histone-lysine N-methyltransferase 2B\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"TRX2\",\\n                            \"TXN\",\\n                            \"TXN2\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_19\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_338\",\\n                        \"names\": [\\n                            \"Cellular tumor antigen p53\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"P53\",\\n                            \"TP53\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_338\",\\n                        \"names\": [\\n                            \"Cellular tumor antigen p53\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"P53\",\\n                            \"TP53\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_361\",\\n                        \"names\": [\\n                            \"Dual specificity protein phosphatase 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PTPN10\",\\n                            \"CL100\",\\n                            \"MKP1\",\\n                            \"DUSP1\"\\n                        ]\\n                    }\\n                }\\n            ],\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_326\",\\n                        \"names\": [\\n                            \"Parkinson disease protein 7\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"PARK7\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_327\",\\n                        \"names\": [\\n                            \"Bcl-2-like protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BCL2L\",\\n                            \"BCL2L1\",\\n                            \"BCLX\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"missing interaction\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_327\",\\n                        \"names\": [\\n                            \"Bcl-2-like protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BCL2L\",\\n                            \"BCL2L1\",\\n                            \"BCLX\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_316\",\\n                        \"names\": [\\n                            \"Apoptosis regulator BAX\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BAX\",\\n                            \"BCL2L4\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_316\",\\n                        \"names\": [\\n                            \"Apoptosis regulator BAX\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"BAX\",\\n                            \"BCL2L4\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_25\",\\n                        \"names\": [\\n                            \"Cytochrome c\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CYC\",\\n                            \"CYCS\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_25\",\\n                        \"names\": [\\n                            \"Cytochrome c\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CYC\",\\n                            \"CYCS\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_152\",\\n                        \"names\": [\\n                            \"Caspase-9\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CASP9\",\\n                            \"MCH6\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"activation\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_152\",\\n                        \"names\": [\\n                            \"Caspase-9\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CASP9\",\\n                            \"MCH6\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_153\",\\n                        \"names\": [\\n                            \"Caspase-3\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"CPP32\",\\n                            \"CASP3\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]   True\n",
       "4                                What downstream genes are affected by Kelch-like ECH-associated protein 1?                                                                  MATCH (start: Gene)\\nWHERE ('Kelch-like ECH-associated protein 1' IN start.names) OR ('Kelch-like ECH-associated protein 1' IN start.synonyms)\\nCALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>', minLevel: 1, uniqueness: 'NODE_PATH', bfs: false}) \\nYIELD path\\nWITH path\\nWHERE NOT EXISTS {\\n    MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\\n    WHERE lastNode = last(nodes(path))\\n}\\nWITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\nRETURN collect(relationships) AS interactions                                                               MATCH (start: Gene)\\n            WHERE 'Kelch-like ECH-associated protein 1' IN start.names\\n            CALL apoc.path.expandConfig(start, {relationshipFilter: 'INTERACTS_WITH>', minLevel: 1, uniqueness: 'NODE_PATH', bfs: false}) \\n            YIELD path\\n            WITH path\\n            WHERE NOT EXISTS {\\n                MATCH (lastNode)-[:INTERACTS_WITH]->(:Gene)\\n                WHERE lastNode = last(nodes(path))\\n            }\\n            WITH path, [rel IN relationships(path) | {start: startNode(rel), end: endNode(rel), type: rel.type, subtypes: rel.subtypes}] AS relationships\\n            RETURN collect(relationships) AS interactions                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 [\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"inhibition\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_714\",\\n                        \"names\": [\\n                            \"Kelch-like ECH-associated protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"KLHL19\",\\n                            \"KEAP1\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_362\",\\n                        \"names\": [\\n                            \"Thioredoxin\",\\n                            \"Thioredoxin, mitochondrial\",\\n                            \"Histone-lysine N-methyltransferase 2B\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"TRX2\",\\n                            \"TXN\",\\n                            \"TXN2\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 [\\n    {\\n        \"interactions\": [\\n            [\\n                {\\n                    \"subtypes\": [\\n                        \"inhibition\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_714\",\\n                        \"names\": [\\n                            \"Kelch-like ECH-associated protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"KLHL19\",\\n                            \"KEAP1\"\\n                        ]\\n                    },\\n                    \"type\": \"PPrel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    }\\n                },\\n                {\\n                    \"subtypes\": [\\n                        \"expression\"\\n                    ],\\n                    \"start\": {\\n                        \"unique_id\": \"05012_347\",\\n                        \"names\": [\\n                            \"Nuclear factor erythroid 2-related factor 2\",\\n                            \"Heme-binding protein 1\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"NFE2L2\",\\n                            \"NRF2\",\\n                            \"HEBP1\"\\n                        ]\\n                    },\\n                    \"type\": \"GErel\",\\n                    \"end\": {\\n                        \"unique_id\": \"05012_362\",\\n                        \"names\": [\\n                            \"Thioredoxin\",\\n                            \"Thioredoxin, mitochondrial\",\\n                            \"Histone-lysine N-methyltransferase 2B\"\\n                        ],\\n                        \"synonyms\": [\\n                            \"TRX2\",\\n                            \"TXN\",\\n                            \"TXN2\"\\n                        ]\\n                    }\\n                }\\n            ]\\n        ]\\n    }\\n]   True"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "downstream_evaluation_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "newenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
